// yummies/assert.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const require_chunk = require("./chunk-CZWwpsFl.cjs");
let yummies_type_guard = require("yummies/type-guard");
let yummies_common = require("yummies/common");
//#region src/assert/_exports.ts
/**
* ---header-docs-section---
* # yummies/assert
*
* ## Description
*
* Runtime **assertions** for invariants, narrowing, and exhaustiveness. Helpers throw
* {@link AssertionError} with an optional message (`string` or {@link MaybeFn}) so failing fast stays
* explicit without a heavy assertion library. Use `assert.ok`, `assert.string`, `assert.defined`,
* `assert.fail`, and the rest on the namespace export from the package entry.
*
* ## Usage
*
* ```ts
* import { assert } from "yummies/assert";
*
* assert.ok(user.id, "user id is required");
* assert.defined(maybeName);
* assert.string(raw);
* ```
*/
var _exports_exports = /* @__PURE__ */ require_chunk.__exportAll({
	AssertionError: () => AssertionError,
	array: () => array,
	boolean: () => boolean,
	defined: () => defined,
	fail: () => fail,
	instanceOf: () => instanceOf,
	invariant: () => invariant,
	never: () => unreachable,
	notNull: () => notNull,
	notUndefined: () => notUndefined,
	number: () => number,
	object: () => object,
	ok: () => ok,
	string: () => string,
	symbol: () => symbol,
	unreachable: () => unreachable
});
/**
* Error thrown by assertion helpers in this module.
*/
var AssertionError = class extends Error {
	name = "AssertionError";
};
function resolveAssertMessage(message, fallback) {
	if (message === void 0) return fallback;
	return (0, yummies_common.callFunction)(message);
}
/**
* Throws when `condition` is falsy; narrows the type when it is truthy.
*
* @param condition Value treated as a boolean guard.
* @param message Optional message or lazy message factory.
*
* @example
* ```ts
* assert.ok(user.id, "user id is required");
* ```
*/
function ok(condition, message) {
	if (yummies_type_guard.typeGuard.isFalsy(condition)) throw new AssertionError(resolveAssertMessage(message, "Assertion failed"));
}
/**
* Same as {@link ok}; common alias for internal invariants (e.g. after optional checks).
*/
var invariant = ok;
/**
* Throws when the value is `null` or `undefined`; otherwise narrows away nullish.
*
* @param value Possibly nullish value.
* @param message Optional message or lazy message factory.
*
* @example
* ```ts
* assert.defined(maybeName, "name must be present");
* ```
*/
function defined(value, message) {
	if (!yummies_type_guard.typeGuard.isDefined(value)) throw new AssertionError(resolveAssertMessage(message, "Expected a defined value"));
}
/**
* Throws when the value is `null`; allows `undefined` unless combined with other checks.
*
* @param value Value that may be `null`.
* @param message Optional message or lazy message factory.
*/
function notNull(value, message) {
	if (yummies_type_guard.typeGuard.isNull(value)) throw new AssertionError(resolveAssertMessage(message, "Expected a non-null value"));
}
/**
* Throws when the value is `undefined`.
*
* @param value Value that may be `undefined`.
* @param message Optional message or lazy message factory.
*/
function notUndefined(value, message) {
	if (yummies_type_guard.typeGuard.isUndefined(value)) throw new AssertionError(resolveAssertMessage(message, "Expected a defined value"));
}
/**
* Throws when `value` is not a string (primitive or `String` object); uses `typeGuard.isString`.
*
* @param value Value to narrow.
* @param message Optional message or lazy message factory.
*/
function string(value, message) {
	if (!yummies_type_guard.typeGuard.isString(value)) throw new AssertionError(resolveAssertMessage(message, "Expected a string"));
}
/**
* Throws when `value` is not a object (primitive or `AnyObject` object); uses `typeGuard.isObject`.
*
* @param value Value to narrow.
* @param message Optional message or lazy message factory.
*/
function object(value, message) {
	if (!yummies_type_guard.typeGuard.isObject(value)) throw new AssertionError(resolveAssertMessage(message, "Expected an object"));
}
/**
* Throws when `value` is not a object (primitive or `AnyObject` object); uses `typeGuard.isArray`.
*
* @param value Value to narrow.
* @param message Optional message or lazy message factory.
*/
function array(value, message) {
	if (!yummies_type_guard.typeGuard.isArray(value)) throw new AssertionError(resolveAssertMessage(message, "Expected an array"));
}
/**
* Throws when `value` is not a finite non-NaN number; uses `typeGuard.isNumber`.
*
* @param value Value to narrow.
* @param message Optional message or lazy message factory.
*/
function number(value, message) {
	if (!yummies_type_guard.typeGuard.isNumber(value)) throw new AssertionError(resolveAssertMessage(message, "Expected a finite number"));
}
/**
* Throws when `value` is not a boolean; uses `typeGuard.isBoolean`.
*
* @param value Value to narrow.
* @param message Optional message or lazy message factory.
*/
function boolean(value, message) {
	if (!yummies_type_guard.typeGuard.isBoolean(value)) throw new AssertionError(resolveAssertMessage(message, "Expected a boolean"));
}
/**
* Throws when `value` is not a symbol; uses `typeGuard.isSymbol`.
*
* @param value Value to narrow.
* @param message Optional message or lazy message factory.
*/
function symbol(value, message) {
	if (!yummies_type_guard.typeGuard.isSymbol(value)) throw new AssertionError(resolveAssertMessage(message, "Expected a symbol"));
}
/**
* Always throws; use for branches that must be impossible or as a typed “abort”.
*
* @param message Optional message or lazy message factory.
*
* @example
* ```ts
* function parse(kind: "a" | "b") {
*   if (kind === "a") return 1;
*   if (kind === "b") return 2;
*   assert.fail("unreachable");
* }
* ```
*/
function fail(message) {
	throw new AssertionError(resolveAssertMessage(message, "Unreachable"));
}
/**
* Exhaustiveness helper: call with a `never` value when all union cases are handled.
*
* @param value Should be `never` when the type checker is satisfied.
* @param message Optional message (this path always throws, so a lazy factory is unnecessary).
*/
function unreachable(value, message) {
	throw new AssertionError(resolveAssertMessage(message, `Unexpected value: ${String(value)}`));
}
/**
* Throws when `value` is not an instance of `ctor`.
*
* @param value Value to check.
* @param ctor Constructor used with `instanceof`.
* @param message Optional message or lazy message factory.
*
* @example
* ```ts
* assert.instanceOf(el, HTMLElement, "expected a DOM element");
* ```
*/
function instanceOf(value, ctor, message) {
	if (!(value instanceof ctor)) throw new AssertionError(resolveAssertMessage(message, `Expected instance of ${ctor.name}`));
}
//#endregion
Object.defineProperty(exports, "assert", {
	enumerable: true,
	get: function() {
		return _exports_exports;
	}
});
