// yummies/complex.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
//#region src/complex/counter.ts
/**
* Creates a callable counter object with increment, decrement and reset helpers.
*
* The returned function increments the internal numeric counter when called and
* exposes both the raw counter value and an optionally transformed `value`.
*
* @template TValue Public value type returned by the counter.
* @param processValue Optional mapper that transforms the numeric counter value.
* @param initial Initial numeric counter value.
* @returns Callable counter with state and control methods.
*
* @example
* ```ts
* const counter = createCounter();
* counter.increment(); // 1
* ```
*
* @example
* ```ts
* const idCounter = createCounter((value) => `id-${value}`, 10);
* idCounter(); // 'id-11'
* ```
*/
var createCounter = (processValue, initial = 0) => {
	const update = (counter) => {
		fn.value = processValue?.(counter) ?? counter;
		return fn.value;
	};
	const increment = () => update(++fn.counter);
	const decrement = () => update(--fn.counter);
	const fn = increment;
	fn.increment = increment;
	fn.decrement = decrement;
	fn.reset = () => {
		fn.counter = initial;
		fn.value = processValue?.(initial) ?? initial;
	};
	fn.reset();
	return fn;
};
//#endregion
//#region src/complex/global-point.ts
/**
* Creates a simple storage point that can live either in `globalThis` under a
* provided key or in a local closure when no key is given.
*
* @template TValue Stored value type.
* @param accessSymbol Optional global property name used for storage.
* @returns Getter/setter API for the stored value.
*
* @example
* ```ts
* const point = createGlobalPoint<number>();
* point.set(10);
* ```
*
* @example
* ```ts
* const point = createGlobalPoint<string>('__token__');
* point.get();
* ```
*/
var createGlobalPoint = (accessSymbol) => {
	if (accessSymbol == null) {
		let storedValue;
		return {
			get: () => storedValue,
			unset: () => {
				storedValue = void 0;
			},
			set: (value) => {
				storedValue = value;
				return value;
			}
		};
	}
	const _globalThis = globalThis;
	return {
		get: () => _globalThis[accessSymbol],
		unset: () => {
			delete _globalThis[accessSymbol];
		},
		set: (value) => {
			_globalThis[accessSymbol] = value;
			return value;
		}
	};
};
//#endregion
//#region src/complex/global-config.ts
/**
* Creates or reuses a globally accessible config object.
*
* The config is stored in a global point identified by `accessSymbol`, or in a
* local closure when no symbol is provided.
*
* @template T Config object type.
* @param defaultValue Default value used when no config has been created yet.
* @param accessSymbol Optional global key used to store the config.
* @returns Existing or newly initialized global config object.
*
* @example
* ```ts
* const config = createGlobalConfig({ locale: 'en' }, '__app_config__');
* ```
*
* @example
* ```ts
* const config = createGlobalConfig({ debug: false });
* config.debug;
* ```
*/
var createGlobalConfig = (defaultValue, accessSymbol) => {
	const globalPoint = createGlobalPoint(accessSymbol);
	return globalPoint.get() || globalPoint.set(defaultValue);
};
/**
* Creates a mutable global config manager whose value is produced and updated
* through a custom processor function.
*
* @template T Config object type.
* @param processFn Function that builds the next config state from a partial change and current value.
* @param accessSymbol Optional global key used to store the config.
* @returns API for reading, replacing, resetting and partially updating the config.
*
* @example
* ```ts
* const config = createGlobalDynamicConfig(
*   (change, current) => ({ theme: 'light', ...current, ...change }),
*   '__theme__',
* );
* ```
*
* @example
* ```ts
* const config = createGlobalDynamicConfig((change, current) => ({ ...current, ...change }));
* config.update({ locale: 'ru' });
* ```
*/
var createGlobalDynamicConfig = (processFn, accessSymbol) => {
	const globalPoint = createGlobalPoint(accessSymbol);
	const getValue = () => {
		return globalPoint.get() ?? globalPoint.set(processFn(null, null));
	};
	return {
		get: getValue,
		set: globalPoint.set,
		unset: globalPoint.unset,
		update: (value) => {
			const currentValue = getValue();
			Object.assign(currentValue, processFn(value, currentValue));
		}
	};
};
//#endregion
//#region src/complex/modules-factory.ts
/**
* Universal factory for creating class instances with predefined and per-call
* dependencies.
*
* Works with classes whose constructor accepts a single dependency object.
*
* @template TPredefinedDeps Dependency shape that is always injected by the factory.
*
* @example
* ```
* const factory = new ModulesFactory({
*   factory: (MyClass, deps) => new MyClass(deps),
*   deps: { someDependency: new Dependency() }
* });
*
* const instance = factory.create(MyClass, { extraDependency: new ExtraDependency() });
* ```
*
* @example
* ```ts
* const factory = new ModulesFactory({
*   factory: (Module, deps) => new Module(deps),
* });
* ```
*
* @example
* ```ts
* const service = factory.create(UserService, { api });
* ```
*/
var ModulesFactory = class {
	/**
	* Creates a new module factory.
	*
	* @param config Factory strategy and predefined dependencies.
	*
	* @example
	* ```ts
	* const factory = new ModulesFactory({
	*   factory: (Module, deps) => new Module(deps),
	* });
	* ```
	*
	* @example
	* ```ts
	* const factory = new ModulesFactory({
	*   factory: (Module, deps) => new Module(deps),
	*   deps: { api },
	* });
	* ```
	*/
	constructor(config) {
		this.config = config;
	}
	/**
	* Creates an instance of the provided class by merging predefined and
	* per-call dependencies.
	*
	* @template TInstance Instance type produced by the constructor.
	* @template TDeps Full dependency object expected by the constructor.
	* @param Constructor Class constructor receiving a single dependency object.
	* @param args Additional dependencies merged over predefined ones.
	* @returns Created class instance.
	*
	* @example
	* ```ts
	* const service = factory.create(UserService, { logger });
	* ```
	*
	* @example
	* ```ts
	* const store = factory.create(UserStore);
	* ```
	*/
	create(Constructor, ...args) {
		return this.config.factory(Constructor, {
			...this.config.deps,
			...args[0]
		});
	}
};
//#endregion
//#region src/complex/pub-sub.ts
/**
* Creates a simple publish-subscribe dispatcher that stores the last published
* arguments and allows subscription management.
*
* @template PubArgs Argument tuple delivered to subscribers.
* @returns Callable publisher with subscribe and unsubscribe helpers.
*
* @example
* ```ts
* const pub = createPubSub<[string]>();
* pub('ready');
* ```
*
* @example
* ```ts
* const pub = createPubSub<[number]>();
* const unsub = pub.sub((value) => console.log(value));
* ```
*/
var createPubSub = () => {
	const pubSub = ((...args) => {
		pubSub.lastPub = args;
		pubSub.subs.forEach((sub) => {
			sub(...args);
		});
	});
	pubSub.lastPub = void 0;
	pubSub.subs = [];
	pubSub.unsub = (sub) => {
		pubSub.subs = pubSub.subs.filter((it) => it !== sub);
	};
	pubSub.sub = (sub) => {
		pubSub.subs.push(sub);
		return () => pubSub.unsub(sub);
	};
	return pubSub;
};
//#endregion
exports.ModulesFactory = ModulesFactory;
exports.createCounter = createCounter;
exports.createGlobalConfig = createGlobalConfig;
exports.createGlobalDynamicConfig = createGlobalDynamicConfig;
exports.createPubSub = createPubSub;
