// yummies/data.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
//#region src/data.ts
/**
* Performs a shallow comparison for arrays, plain objects, dates and regular expressions.
*
* @example
* ```ts
* isShallowEqual({ id: 1 }, { id: 1 }); // true
* ```
*/
var isShallowEqual = (a, b) => {
	if (a === b) return true;
	if (typeof a !== "object" || typeof b !== "object" || a === null || b === null) return false;
	if (a.constructor !== b.constructor) return false;
	const isArrayA = Array.isArray(a);
	if (isArrayA !== Array.isArray(b)) return false;
	if (isArrayA) {
		const arrA = a;
		const arrB = b;
		if (arrA.length !== arrB.length) return false;
		for (const [i, element] of arrA.entries()) if (element !== arrB[i]) return false;
		return true;
	}
	if (a instanceof Date) return a.getTime() === b.getTime();
	if (a instanceof RegExp) return a.toString() === b.toString();
	const aKeys = Object.keys(a);
	const bKeys = Object.keys(b);
	if (aKeys.length !== bKeys.length) return false;
	const bObj = b;
	for (const key of aKeys) if (!Object.hasOwn(bObj, key) || a[key] !== bObj[key]) return false;
	return true;
};
/**
* Checks whether an object has at least one enumerable key.
*
* @example
* ```ts
* hasEnumerableKeys({ id: 1 }); // true
* hasEnumerableKeys({}); // false
* ```
*/
var hasEnumerableKeys = (input) => {
	for (const _key in input) return true;
	return false;
};
/**
* Wraps a value in an array when it is not already an array.
*
* @example
* ```ts
* toArray('item'); // ['item']
* ```
*/
var toArray = (value) => {
	return Array.isArray(value) ? value : [value];
};
/**
* Recursively flattens a nested array and maps the collected values.
*
* @example
* ```ts
* flatMapDeep([1, [2, [3]]], (value) => value * 2); // [2, 4, 6]
* ```
*/
var flatMapDeep = (arr, fn) => {
	const source = [];
	const collect = (value) => {
		if (!Array.isArray(value)) {
			source.push(value);
			return;
		}
		for (const item of value) collect(item);
	};
	collect(arr);
	return source.map((value, i) => fn(value, i, source));
};
/**
* Parses JSON safely and returns a fallback value when parsing fails.
*
* @example
* ```ts
* safeJsonParse('{"enabled":true}', {}); // { enabled: true }
* ```
*/
var safeJsonParse = (json, fallback = null) => {
	if (json == null) return fallback;
	try {
		return JSON.parse(json);
	} catch {
		return fallback;
	}
};
var UNSAFE_PROPERTY_KEYS = new Set([
	"__proto__",
	"prototype",
	"constructor"
]);
/**
* Checks whether a property key is unsafe and can lead to prototype pollution.
*
* @example
* isUnsafeProperty('__proto__'); // true
* isUnsafeProperty('name'); // false
*/
var isUnsafeProperty = (key) => UNSAFE_PROPERTY_KEYS.has(key);
//#endregion
exports.flatMapDeep = flatMapDeep;
exports.hasEnumerableKeys = hasEnumerableKeys;
exports.isShallowEqual = isShallowEqual;
exports.isUnsafeProperty = isUnsafeProperty;
exports.safeJsonParse = safeJsonParse;
exports.toArray = toArray;
