// yummies/date-time.js
import { typeGuard } from "yummies/type-guard";
import { format } from "yummies/format";
import { ms, unitsToMs } from "yummies/ms";
import { declension } from "yummies/text";
var SECONDS_A_HOUR = 3600;
var SECONDS_A_DAY = SECONDS_A_HOUR * 24;
var SECONDS_A_WEEK = SECONDS_A_DAY * 7;
var MILLISECONDS_A_SECOND = 1e3;
var MILLISECONDS_A_MINUTE = 60 * MILLISECONDS_A_SECOND;
var MILLISECONDS_A_HOUR = SECONDS_A_HOUR * MILLISECONDS_A_SECOND;
var MILLISECONDS_A_DAY = SECONDS_A_DAY * MILLISECONDS_A_SECOND;
var MILLISECONDS_A_WEEK = SECONDS_A_WEEK * MILLISECONDS_A_SECOND;
var MS = "millisecond";
var S = "second";
var MIN = "minute";
var H = "hour";
var W = "week";
var M = "month";
var Q = "quarter";
var Y = "year";
var DATE = "date";
var INVALID_DATE_STRING = "Invalid Date";
var REGEX_PARSE = /^(\d{4})[-/]?(\d{1,2})?[-/]?(\d{0,2})[Tt\s]*(\d{1,2})?:?(\d{1,2})?:?(\d{1,2})?[.:]?(\d+)?$/;
var REGEX_FORMAT = /\[([^\]]+)]|Y{1,4}|M{1,4}|D{1,2}|d{1,4}|H{1,2}|h{1,2}|a|A|m{1,2}|s{1,2}|Z{1,2}|SSS/g;
//#endregion
//#region node_modules/.pnpm/dayjs@1.11.20/node_modules/dayjs/esm/locale/en.js
var en_default = {
	name: "en",
	weekdays: "Sunday_Monday_Tuesday_Wednesday_Thursday_Friday_Saturday".split("_"),
	months: "January_February_March_April_May_June_July_August_September_October_November_December".split("_"),
	ordinal: function ordinal(n) {
		var s = [
			"th",
			"st",
			"nd",
			"rd"
		];
		var v = n % 100;
		return "[" + n + (s[(v - 20) % 10] || s[v] || s[0]) + "]";
	}
};
//#endregion
//#region node_modules/.pnpm/dayjs@1.11.20/node_modules/dayjs/esm/utils.js
var padStart = function padStart(string, length, pad) {
	var s = String(string);
	if (!s || s.length >= length) return string;
	return "" + Array(length + 1 - s.length).join(pad) + string;
};
var utils_default = {
	s: padStart,
	z: function padZoneStr(instance) {
		var negMinutes = -instance.utcOffset();
		var minutes = Math.abs(negMinutes);
		var hourOffset = Math.floor(minutes / 60);
		var minuteOffset = minutes % 60;
		return (negMinutes <= 0 ? "+" : "-") + padStart(hourOffset, 2, "0") + ":" + padStart(minuteOffset, 2, "0");
	},
	m: function monthDiff(a, b) {
		if (a.date() < b.date()) return -monthDiff(b, a);
		var wholeMonthDiff = (b.year() - a.year()) * 12 + (b.month() - a.month());
		var anchor = a.clone().add(wholeMonthDiff, M);
		var c = b - anchor < 0;
		var anchor2 = a.clone().add(wholeMonthDiff + (c ? -1 : 1), M);
		return +(-(wholeMonthDiff + (b - anchor) / (c ? anchor - anchor2 : anchor2 - anchor)) || 0);
	},
	a: function absFloor(n) {
		return n < 0 ? Math.ceil(n) || 0 : Math.floor(n);
	},
	p: function prettyUnit(u) {
		return {
			M: "month",
			y: "year",
			w: "week",
			d: "day",
			D: "date",
			h: "hour",
			m: "minute",
			s: "second",
			ms: "millisecond",
			Q: "quarter"
		}[u] || String(u || "").toLowerCase().replace(/s$/, "");
	},
	u: function isUndefined(s) {
		return s === void 0;
	}
};
//#endregion
//#region node_modules/.pnpm/dayjs@1.11.20/node_modules/dayjs/esm/index.js
var L = "en";
var Ls = {};
Ls[L] = en_default;
var IS_DAYJS = "$isDayjsObject";
var isDayjs = function isDayjs(d) {
	return d instanceof Dayjs || !!(d && d[IS_DAYJS]);
};
var parseLocale = function parseLocale(preset, object, isLocal) {
	var l;
	if (!preset) return L;
	if (typeof preset === "string") {
		var presetLower = preset.toLowerCase();
		if (Ls[presetLower]) l = presetLower;
		if (object) {
			Ls[presetLower] = object;
			l = presetLower;
		}
		var presetSplit = preset.split("-");
		if (!l && presetSplit.length > 1) return parseLocale(presetSplit[0]);
	} else {
		var name = preset.name;
		Ls[name] = preset;
		l = name;
	}
	if (!isLocal && l) L = l;
	return l || !isLocal && L;
};
var dayjs = function dayjs(date, c) {
	if (isDayjs(date)) return date.clone();
	var cfg = typeof c === "object" ? c : {};
	cfg.date = date;
	cfg.args = arguments;
	return new Dayjs(cfg);
};
var wrapper$1 = function wrapper(date, instance) {
	return dayjs(date, {
		locale: instance.$L,
		utc: instance.$u,
		x: instance.$x,
		$offset: instance.$offset
	});
};
var Utils = utils_default;
Utils.l = parseLocale;
Utils.i = isDayjs;
Utils.w = wrapper$1;
var parseDate = function parseDate(cfg) {
	var date = cfg.date, utc = cfg.utc;
	if (date === null) return /* @__PURE__ */ new Date(NaN);
	if (Utils.u(date)) return /* @__PURE__ */ new Date();
	if (date instanceof Date) return new Date(date);
	if (typeof date === "string" && !/Z$/i.test(date)) {
		var d = date.match(REGEX_PARSE);
		if (d) {
			var m = d[2] - 1 || 0;
			var ms = (d[7] || "0").substring(0, 3);
			if (utc) return new Date(Date.UTC(d[1], m, d[3] || 1, d[4] || 0, d[5] || 0, d[6] || 0, ms));
			return new Date(d[1], m, d[3] || 1, d[4] || 0, d[5] || 0, d[6] || 0, ms);
		}
	}
	return new Date(date);
};
var Dayjs = /* @__PURE__ */ function() {
	function Dayjs(cfg) {
		this.$L = parseLocale(cfg.locale, null, true);
		this.parse(cfg);
		this.$x = this.$x || cfg.x || {};
		this[IS_DAYJS] = true;
	}
	var _proto = Dayjs.prototype;
	_proto.parse = function parse(cfg) {
		this.$d = parseDate(cfg);
		this.init();
	};
	_proto.init = function init() {
		var $d = this.$d;
		this.$y = $d.getFullYear();
		this.$M = $d.getMonth();
		this.$D = $d.getDate();
		this.$W = $d.getDay();
		this.$H = $d.getHours();
		this.$m = $d.getMinutes();
		this.$s = $d.getSeconds();
		this.$ms = $d.getMilliseconds();
	};
	_proto.$utils = function $utils() {
		return Utils;
	};
	_proto.isValid = function isValid() {
		return !(this.$d.toString() === INVALID_DATE_STRING);
	};
	_proto.isSame = function isSame(that, units) {
		var other = dayjs(that);
		return this.startOf(units) <= other && other <= this.endOf(units);
	};
	_proto.isAfter = function isAfter(that, units) {
		return dayjs(that) < this.startOf(units);
	};
	_proto.isBefore = function isBefore(that, units) {
		return this.endOf(units) < dayjs(that);
	};
	_proto.$g = function $g(input, get, set) {
		if (Utils.u(input)) return this[get];
		return this.set(set, input);
	};
	_proto.unix = function unix() {
		return Math.floor(this.valueOf() / 1e3);
	};
	_proto.valueOf = function valueOf() {
		return this.$d.getTime();
	};
	_proto.startOf = function startOf(units, _startOf) {
		var _this = this;
		var isStartOf = !Utils.u(_startOf) ? _startOf : true;
		var unit = Utils.p(units);
		var instanceFactory = function instanceFactory(d, m) {
			var ins = Utils.w(_this.$u ? Date.UTC(_this.$y, m, d) : new Date(_this.$y, m, d), _this);
			return isStartOf ? ins : ins.endOf("day");
		};
		var instanceFactorySet = function instanceFactorySet(method, slice) {
			return Utils.w(_this.toDate()[method].apply(_this.toDate("s"), (isStartOf ? [
				0,
				0,
				0,
				0
			] : [
				23,
				59,
				59,
				999
			]).slice(slice)), _this);
		};
		var $W = this.$W, $M = this.$M, $D = this.$D;
		var utcPad = "set" + (this.$u ? "UTC" : "");
		switch (unit) {
			case Y: return isStartOf ? instanceFactory(1, 0) : instanceFactory(31, 11);
			case M: return isStartOf ? instanceFactory(1, $M) : instanceFactory(0, $M + 1);
			case W:
				var weekStart = this.$locale().weekStart || 0;
				var gap = ($W < weekStart ? $W + 7 : $W) - weekStart;
				return instanceFactory(isStartOf ? $D - gap : $D + (6 - gap), $M);
			case "day":
			case DATE: return instanceFactorySet(utcPad + "Hours", 0);
			case H: return instanceFactorySet(utcPad + "Minutes", 1);
			case MIN: return instanceFactorySet(utcPad + "Seconds", 2);
			case S: return instanceFactorySet(utcPad + "Milliseconds", 3);
			default: return this.clone();
		}
	};
	_proto.endOf = function endOf(arg) {
		return this.startOf(arg, false);
	};
	_proto.$set = function $set(units, _int) {
		var _C$D$C$DATE$C$M$C$Y$C;
		var unit = Utils.p(units);
		var utcPad = "set" + (this.$u ? "UTC" : "");
		var name = (_C$D$C$DATE$C$M$C$Y$C = {}, _C$D$C$DATE$C$M$C$Y$C["day"] = utcPad + "Date", _C$D$C$DATE$C$M$C$Y$C[DATE] = utcPad + "Date", _C$D$C$DATE$C$M$C$Y$C[M] = utcPad + "Month", _C$D$C$DATE$C$M$C$Y$C[Y] = utcPad + "FullYear", _C$D$C$DATE$C$M$C$Y$C[H] = utcPad + "Hours", _C$D$C$DATE$C$M$C$Y$C[MIN] = utcPad + "Minutes", _C$D$C$DATE$C$M$C$Y$C[S] = utcPad + "Seconds", _C$D$C$DATE$C$M$C$Y$C[MS] = utcPad + "Milliseconds", _C$D$C$DATE$C$M$C$Y$C)[unit];
		var arg = unit === "day" ? this.$D + (_int - this.$W) : _int;
		if (unit === "month" || unit === "year") {
			var date = this.clone().set(DATE, 1);
			date.$d[name](arg);
			date.init();
			this.$d = date.set(DATE, Math.min(this.$D, date.daysInMonth())).$d;
		} else if (name) this.$d[name](arg);
		this.init();
		return this;
	};
	_proto.set = function set(string, _int2) {
		return this.clone().$set(string, _int2);
	};
	_proto.get = function get(unit) {
		return this[Utils.p(unit)]();
	};
	_proto.add = function add(number, units) {
		var _this2 = this, _C$MIN$C$H$C$S$unit;
		number = Number(number);
		var unit = Utils.p(units);
		var instanceFactorySet = function instanceFactorySet(n) {
			var d = dayjs(_this2);
			return Utils.w(d.date(d.date() + Math.round(n * number)), _this2);
		};
		if (unit === "month") return this.set(M, this.$M + number);
		if (unit === "year") return this.set(Y, this.$y + number);
		if (unit === "day") return instanceFactorySet(1);
		if (unit === "week") return instanceFactorySet(7);
		var step = (_C$MIN$C$H$C$S$unit = {}, _C$MIN$C$H$C$S$unit["minute"] = 6e4, _C$MIN$C$H$C$S$unit["hour"] = 36e5, _C$MIN$C$H$C$S$unit["second"] = 1e3, _C$MIN$C$H$C$S$unit)[unit] || 1;
		var nextTimeStamp = this.$d.getTime() + number * step;
		return Utils.w(nextTimeStamp, this);
	};
	_proto.subtract = function subtract(number, string) {
		return this.add(number * -1, string);
	};
	_proto.format = function format(formatStr) {
		var _this3 = this;
		var locale = this.$locale();
		if (!this.isValid()) return locale.invalidDate || "Invalid Date";
		var str = formatStr || "YYYY-MM-DDTHH:mm:ssZ";
		var zoneStr = Utils.z(this);
		var $H = this.$H, $m = this.$m, $M = this.$M;
		var weekdays = locale.weekdays, months = locale.months, meridiem = locale.meridiem;
		var getShort = function getShort(arr, index, full, length) {
			return arr && (arr[index] || arr(_this3, str)) || full[index].slice(0, length);
		};
		var get$H = function get$H(num) {
			return Utils.s($H % 12 || 12, num, "0");
		};
		var meridiemFunc = meridiem || function(hour, minute, isLowercase) {
			var m = hour < 12 ? "AM" : "PM";
			return isLowercase ? m.toLowerCase() : m;
		};
		var matches = function matches(match) {
			switch (match) {
				case "YY": return String(_this3.$y).slice(-2);
				case "YYYY": return Utils.s(_this3.$y, 4, "0");
				case "M": return $M + 1;
				case "MM": return Utils.s($M + 1, 2, "0");
				case "MMM": return getShort(locale.monthsShort, $M, months, 3);
				case "MMMM": return getShort(months, $M);
				case "D": return _this3.$D;
				case "DD": return Utils.s(_this3.$D, 2, "0");
				case "d": return String(_this3.$W);
				case "dd": return getShort(locale.weekdaysMin, _this3.$W, weekdays, 2);
				case "ddd": return getShort(locale.weekdaysShort, _this3.$W, weekdays, 3);
				case "dddd": return weekdays[_this3.$W];
				case "H": return String($H);
				case "HH": return Utils.s($H, 2, "0");
				case "h": return get$H(1);
				case "hh": return get$H(2);
				case "a": return meridiemFunc($H, $m, true);
				case "A": return meridiemFunc($H, $m, false);
				case "m": return String($m);
				case "mm": return Utils.s($m, 2, "0");
				case "s": return String(_this3.$s);
				case "ss": return Utils.s(_this3.$s, 2, "0");
				case "SSS": return Utils.s(_this3.$ms, 3, "0");
				case "Z": return zoneStr;
				default: break;
			}
			return null;
		};
		return str.replace(REGEX_FORMAT, function(match, $1) {
			return $1 || matches(match) || zoneStr.replace(":", "");
		});
	};
	_proto.utcOffset = function utcOffset() {
		return -Math.round(this.$d.getTimezoneOffset() / 15) * 15;
	};
	_proto.diff = function diff(input, units, _float) {
		var _this4 = this;
		var unit = Utils.p(units);
		var that = dayjs(input);
		var zoneDelta = (that.utcOffset() - this.utcOffset()) * MILLISECONDS_A_MINUTE;
		var diff = this - that;
		var getMonth = function getMonth() {
			return Utils.m(_this4, that);
		};
		var result;
		switch (unit) {
			case Y:
				result = getMonth() / 12;
				break;
			case M:
				result = getMonth();
				break;
			case Q:
				result = getMonth() / 3;
				break;
			case W:
				result = (diff - zoneDelta) / MILLISECONDS_A_WEEK;
				break;
			case "day":
				result = (diff - zoneDelta) / MILLISECONDS_A_DAY;
				break;
			case H:
				result = diff / MILLISECONDS_A_HOUR;
				break;
			case MIN:
				result = diff / MILLISECONDS_A_MINUTE;
				break;
			case S:
				result = diff / MILLISECONDS_A_SECOND;
				break;
			default:
				result = diff;
				break;
		}
		return _float ? result : Utils.a(result);
	};
	_proto.daysInMonth = function daysInMonth() {
		return this.endOf(M).$D;
	};
	_proto.$locale = function $locale() {
		return Ls[this.$L];
	};
	_proto.locale = function locale(preset, object) {
		if (!preset) return this.$L;
		var that = this.clone();
		var nextLocaleName = parseLocale(preset, object, true);
		if (nextLocaleName) that.$L = nextLocaleName;
		return that;
	};
	_proto.clone = function clone() {
		return Utils.w(this.$d, this);
	};
	_proto.toDate = function toDate() {
		return new Date(this.valueOf());
	};
	_proto.toJSON = function toJSON() {
		return this.isValid() ? this.toISOString() : null;
	};
	_proto.toISOString = function toISOString() {
		return this.$d.toISOString();
	};
	_proto.toString = function toString() {
		return this.$d.toUTCString();
	};
	return Dayjs;
}();
var proto = Dayjs.prototype;
dayjs.prototype = proto;
[
	["$ms", MS],
	["$s", S],
	["$m", MIN],
	["$H", H],
	["$W", "day"],
	["$M", M],
	["$y", Y],
	["$D", DATE]
].forEach(function(g) {
	proto[g[1]] = function(input) {
		return this.$g(input, g[0], g[1]);
	};
});
dayjs.extend = function(plugin, option) {
	if (!plugin.$i) {
		plugin(option, Dayjs, dayjs);
		plugin.$i = true;
	}
	return dayjs;
};
dayjs.locale = parseLocale;
dayjs.isDayjs = isDayjs;
dayjs.unix = function(timestamp) {
	return dayjs(timestamp * 1e3);
};
dayjs.en = Ls[L];
dayjs.Ls = Ls;
dayjs.p = {};
//#endregion
//#region node_modules/.pnpm/dayjs@1.11.20/node_modules/dayjs/esm/plugin/duration/index.js
var MILLISECONDS_A_YEAR = MILLISECONDS_A_DAY * 365;
var MILLISECONDS_A_MONTH = MILLISECONDS_A_YEAR / 12;
var durationRegex = /^(-|\+)?P(?:([-+]?[0-9,.]*)Y)?(?:([-+]?[0-9,.]*)M)?(?:([-+]?[0-9,.]*)W)?(?:([-+]?[0-9,.]*)D)?(?:T(?:([-+]?[0-9,.]*)H)?(?:([-+]?[0-9,.]*)M)?(?:([-+]?[0-9,.]*)S)?)?$/;
var unitToMS = {
	years: MILLISECONDS_A_YEAR,
	months: MILLISECONDS_A_MONTH,
	days: MILLISECONDS_A_DAY,
	hours: MILLISECONDS_A_HOUR,
	minutes: MILLISECONDS_A_MINUTE,
	seconds: MILLISECONDS_A_SECOND,
	milliseconds: 1,
	weeks: MILLISECONDS_A_WEEK
};
var isDuration = function isDuration(d) {
	return d instanceof Duration;
};
var $d;
var $u;
var wrapper = function wrapper(input, instance, unit) {
	return new Duration(input, unit, instance.$l);
};
var prettyUnit = function prettyUnit(unit) {
	return $u.p(unit) + "s";
};
var isNegative = function isNegative(number) {
	return number < 0;
};
var roundNumber = function roundNumber(number) {
	return isNegative(number) ? Math.ceil(number) : Math.floor(number);
};
var absolute = function absolute(number) {
	return Math.abs(number);
};
var getNumberUnitFormat = function getNumberUnitFormat(number, unit) {
	if (!number) return {
		negative: false,
		format: ""
	};
	if (isNegative(number)) return {
		negative: true,
		format: "" + absolute(number) + unit
	};
	return {
		negative: false,
		format: "" + number + unit
	};
};
var Duration = /* @__PURE__ */ function() {
	function Duration(input, unit, locale) {
		var _this = this;
		this.$d = {};
		this.$l = locale;
		if (input === void 0) {
			this.$ms = 0;
			this.parseFromMilliseconds();
		}
		if (unit) return wrapper(input * unitToMS[prettyUnit(unit)], this);
		if (typeof input === "number") {
			this.$ms = input;
			this.parseFromMilliseconds();
			return this;
		}
		if (typeof input === "object") {
			Object.keys(input).forEach(function(k) {
				_this.$d[prettyUnit(k)] = input[k];
			});
			this.calMilliseconds();
			return this;
		}
		if (typeof input === "string") {
			var d = input.match(durationRegex);
			if (d) {
				var numberD = d.slice(2).map(function(value) {
					return value != null ? Number(value) : 0;
				});
				this.$d.years = numberD[0];
				this.$d.months = numberD[1];
				this.$d.weeks = numberD[2];
				this.$d.days = numberD[3];
				this.$d.hours = numberD[4];
				this.$d.minutes = numberD[5];
				this.$d.seconds = numberD[6];
				this.calMilliseconds();
				return this;
			}
		}
		return this;
	}
	var _proto = Duration.prototype;
	_proto.calMilliseconds = function calMilliseconds() {
		var _this2 = this;
		this.$ms = Object.keys(this.$d).reduce(function(total, unit) {
			return total + (_this2.$d[unit] || 0) * unitToMS[unit];
		}, 0);
	};
	_proto.parseFromMilliseconds = function parseFromMilliseconds() {
		var $ms = this.$ms;
		this.$d.years = roundNumber($ms / MILLISECONDS_A_YEAR);
		$ms %= MILLISECONDS_A_YEAR;
		this.$d.months = roundNumber($ms / MILLISECONDS_A_MONTH);
		$ms %= MILLISECONDS_A_MONTH;
		this.$d.days = roundNumber($ms / MILLISECONDS_A_DAY);
		$ms %= MILLISECONDS_A_DAY;
		this.$d.hours = roundNumber($ms / MILLISECONDS_A_HOUR);
		$ms %= MILLISECONDS_A_HOUR;
		this.$d.minutes = roundNumber($ms / MILLISECONDS_A_MINUTE);
		$ms %= MILLISECONDS_A_MINUTE;
		this.$d.seconds = roundNumber($ms / MILLISECONDS_A_SECOND);
		$ms %= MILLISECONDS_A_SECOND;
		this.$d.milliseconds = $ms;
	};
	_proto.toISOString = function toISOString() {
		var Y = getNumberUnitFormat(this.$d.years, "Y");
		var M = getNumberUnitFormat(this.$d.months, "M");
		var days = +this.$d.days || 0;
		if (this.$d.weeks) days += this.$d.weeks * 7;
		var D = getNumberUnitFormat(days, "D");
		var H = getNumberUnitFormat(this.$d.hours, "H");
		var m = getNumberUnitFormat(this.$d.minutes, "M");
		var seconds = this.$d.seconds || 0;
		if (this.$d.milliseconds) {
			seconds += this.$d.milliseconds / 1e3;
			seconds = Math.round(seconds * 1e3) / 1e3;
		}
		var S = getNumberUnitFormat(seconds, "S");
		var negativeMode = Y.negative || M.negative || D.negative || H.negative || m.negative || S.negative;
		var T = H.format || m.format || S.format ? "T" : "";
		var result = (negativeMode ? "-" : "") + "P" + Y.format + M.format + D.format + T + H.format + m.format + S.format;
		return result === "P" || result === "-P" ? "P0D" : result;
	};
	_proto.toJSON = function toJSON() {
		return this.toISOString();
	};
	_proto.format = function format(formatStr) {
		var str = formatStr || "YYYY-MM-DDTHH:mm:ss";
		var matches = {
			Y: this.$d.years,
			YY: $u.s(this.$d.years, 2, "0"),
			YYYY: $u.s(this.$d.years, 4, "0"),
			M: this.$d.months,
			MM: $u.s(this.$d.months, 2, "0"),
			D: this.$d.days,
			DD: $u.s(this.$d.days, 2, "0"),
			H: this.$d.hours,
			HH: $u.s(this.$d.hours, 2, "0"),
			m: this.$d.minutes,
			mm: $u.s(this.$d.minutes, 2, "0"),
			s: this.$d.seconds,
			ss: $u.s(this.$d.seconds, 2, "0"),
			SSS: $u.s(this.$d.milliseconds, 3, "0")
		};
		return str.replace(REGEX_FORMAT, function(match, $1) {
			return $1 || String(matches[match]);
		});
	};
	_proto.as = function as(unit) {
		return this.$ms / unitToMS[prettyUnit(unit)];
	};
	_proto.get = function get(unit) {
		var base = this.$ms;
		var pUnit = prettyUnit(unit);
		if (pUnit === "milliseconds") base %= 1e3;
		else if (pUnit === "weeks") base = roundNumber(base / unitToMS[pUnit]);
		else base = this.$d[pUnit];
		return base || 0;
	};
	_proto.add = function add(input, unit, isSubtract) {
		var another;
		if (unit) another = input * unitToMS[prettyUnit(unit)];
		else if (isDuration(input)) another = input.$ms;
		else another = wrapper(input, this).$ms;
		return wrapper(this.$ms + another * (isSubtract ? -1 : 1), this);
	};
	_proto.subtract = function subtract(input, unit) {
		return this.add(input, unit, true);
	};
	_proto.locale = function locale(l) {
		var that = this.clone();
		that.$l = l;
		return that;
	};
	_proto.clone = function clone() {
		return wrapper(this.$ms, this);
	};
	_proto.humanize = function humanize(withSuffix) {
		return $d().add(this.$ms, "ms").locale(this.$l).fromNow(!withSuffix);
	};
	_proto.valueOf = function valueOf() {
		return this.asMilliseconds();
	};
	_proto.milliseconds = function milliseconds() {
		return this.get("milliseconds");
	};
	_proto.asMilliseconds = function asMilliseconds() {
		return this.as("milliseconds");
	};
	_proto.seconds = function seconds() {
		return this.get("seconds");
	};
	_proto.asSeconds = function asSeconds() {
		return this.as("seconds");
	};
	_proto.minutes = function minutes() {
		return this.get("minutes");
	};
	_proto.asMinutes = function asMinutes() {
		return this.as("minutes");
	};
	_proto.hours = function hours() {
		return this.get("hours");
	};
	_proto.asHours = function asHours() {
		return this.as("hours");
	};
	_proto.days = function days() {
		return this.get("days");
	};
	_proto.asDays = function asDays() {
		return this.as("days");
	};
	_proto.weeks = function weeks() {
		return this.get("weeks");
	};
	_proto.asWeeks = function asWeeks() {
		return this.as("weeks");
	};
	_proto.months = function months() {
		return this.get("months");
	};
	_proto.asMonths = function asMonths() {
		return this.as("months");
	};
	_proto.years = function years() {
		return this.get("years");
	};
	_proto.asYears = function asYears() {
		return this.as("years");
	};
	return Duration;
}();
var manipulateDuration = function manipulateDuration(date, duration, k) {
	return date.add(duration.years() * k, "y").add(duration.months() * k, "M").add(duration.days() * k, "d").add(duration.hours() * k, "h").add(duration.minutes() * k, "m").add(duration.seconds() * k, "s").add(duration.milliseconds() * k, "ms");
};
var duration_default = (function(option, Dayjs, dayjs) {
	$d = dayjs;
	$u = dayjs().$utils();
	dayjs.duration = function(input, unit) {
		return wrapper(input, { $l: dayjs.locale() }, unit);
	};
	dayjs.isDuration = isDuration;
	var oldAdd = Dayjs.prototype.add;
	var oldSubtract = Dayjs.prototype.subtract;
	Dayjs.prototype.add = function(value, unit) {
		if (isDuration(value)) return manipulateDuration(this, value, 1);
		return oldAdd.bind(this)(value, unit);
	};
	Dayjs.prototype.subtract = function(value, unit) {
		if (isDuration(value)) return manipulateDuration(this, value, -1);
		return oldSubtract.bind(this)(value, unit);
	};
});
//#endregion
//#region node_modules/.pnpm/dayjs@1.11.20/node_modules/dayjs/esm/plugin/relativeTime/index.js
var relativeTime_default = (function(o, c, d) {
	o = o || {};
	var proto = c.prototype;
	var relObj = {
		future: "in %s",
		past: "%s ago",
		s: "a few seconds",
		m: "a minute",
		mm: "%d minutes",
		h: "an hour",
		hh: "%d hours",
		d: "a day",
		dd: "%d days",
		M: "a month",
		MM: "%d months",
		y: "a year",
		yy: "%d years"
	};
	d.en.relativeTime = relObj;
	proto.fromToBase = function(input, withoutSuffix, instance, isFrom, postFormat) {
		var loc = instance.$locale().relativeTime || relObj;
		var T = o.thresholds || [
			{
				l: "s",
				r: 44,
				d: "second"
			},
			{
				l: "m",
				r: 89
			},
			{
				l: "mm",
				r: 44,
				d: "minute"
			},
			{
				l: "h",
				r: 89
			},
			{
				l: "hh",
				r: 21,
				d: "hour"
			},
			{
				l: "d",
				r: 35
			},
			{
				l: "dd",
				r: 25,
				d: "day"
			},
			{
				l: "M",
				r: 45
			},
			{
				l: "MM",
				r: 10,
				d: "month"
			},
			{
				l: "y",
				r: 17
			},
			{
				l: "yy",
				d: "year"
			}
		];
		var Tl = T.length;
		var result;
		var out;
		var isFuture;
		for (var i = 0; i < Tl; i += 1) {
			var t = T[i];
			if (t.d) result = isFrom ? d(input).diff(instance, t.d, true) : instance.diff(input, t.d, true);
			var abs = (o.rounding || Math.round)(Math.abs(result));
			isFuture = result > 0;
			if (abs <= t.r || !t.r) {
				if (abs <= 1 && i > 0) t = T[i - 1];
				var format = loc[t.l];
				if (postFormat) abs = postFormat("" + abs);
				if (typeof format === "string") out = format.replace("%d", abs);
				else out = format(abs, withoutSuffix, t.l, isFuture);
				break;
			}
		}
		if (withoutSuffix) return out;
		var pastOrFuture = isFuture ? loc.future : loc.past;
		if (typeof pastOrFuture === "function") return pastOrFuture(out);
		return pastOrFuture.replace("%s", out);
	};
	function fromTo(input, withoutSuffix, instance, isFrom) {
		return proto.fromToBase(input, withoutSuffix, instance, isFrom);
	}
	proto.to = function(input, withoutSuffix) {
		return fromTo(input, withoutSuffix, this, true);
	};
	proto.from = function(input, withoutSuffix) {
		return fromTo(input, withoutSuffix, this);
	};
	var makeNow = function makeNow(thisDay) {
		return thisDay.$u ? d.utc() : d();
	};
	proto.toNow = function(withoutSuffix) {
		return this.to(makeNow(this), withoutSuffix);
	};
	proto.fromNow = function(withoutSuffix) {
		return this.from(makeNow(this), withoutSuffix);
	};
});
//#endregion
//#region node_modules/.pnpm/dayjs@1.11.20/node_modules/dayjs/esm/locale/ru.js
var monthFormat = "января_февраля_марта_апреля_мая_июня_июля_августа_сентября_октября_ноября_декабря".split("_");
var monthStandalone = "январь_февраль_март_апрель_май_июнь_июль_август_сентябрь_октябрь_ноябрь_декабрь".split("_");
var monthShortFormat = "янв._февр._мар._апр._мая_июня_июля_авг._сент._окт._нояб._дек.".split("_");
var monthShortStandalone = "янв._февр._март_апр._май_июнь_июль_авг._сент._окт._нояб._дек.".split("_");
var MONTHS_IN_FORMAT = /D[oD]?(\[[^[\]]*\]|\s)+MMMM?/;
function plural(word, num) {
	var forms = word.split("_");
	return num % 10 === 1 && num % 100 !== 11 ? forms[0] : num % 10 >= 2 && num % 10 <= 4 && (num % 100 < 10 || num % 100 >= 20) ? forms[1] : forms[2];
}
function relativeTimeWithPlural(number, withoutSuffix, key) {
	var format = {
		mm: withoutSuffix ? "минута_минуты_минут" : "минуту_минуты_минут",
		hh: "час_часа_часов",
		dd: "день_дня_дней",
		MM: "месяц_месяца_месяцев",
		yy: "год_года_лет"
	};
	if (key === "m") return withoutSuffix ? "минута" : "минуту";
	return number + " " + plural(format[key], +number);
}
var months = function months(dayjsInstance, format) {
	if (MONTHS_IN_FORMAT.test(format)) return monthFormat[dayjsInstance.month()];
	return monthStandalone[dayjsInstance.month()];
};
months.s = monthStandalone;
months.f = monthFormat;
var monthsShort = function monthsShort(dayjsInstance, format) {
	if (MONTHS_IN_FORMAT.test(format)) return monthShortFormat[dayjsInstance.month()];
	return monthShortStandalone[dayjsInstance.month()];
};
monthsShort.s = monthShortStandalone;
monthsShort.f = monthShortFormat;
var locale = {
	name: "ru",
	weekdays: "воскресенье_понедельник_вторник_среда_четверг_пятница_суббота".split("_"),
	weekdaysShort: "вск_пнд_втр_срд_чтв_птн_сбт".split("_"),
	weekdaysMin: "вс_пн_вт_ср_чт_пт_сб".split("_"),
	months,
	monthsShort,
	weekStart: 1,
	yearStart: 4,
	formats: {
		LT: "H:mm",
		LTS: "H:mm:ss",
		L: "DD.MM.YYYY",
		LL: "D MMMM YYYY г.",
		LLL: "D MMMM YYYY г., H:mm",
		LLLL: "dddd, D MMMM YYYY г., H:mm"
	},
	relativeTime: {
		future: "через %s",
		past: "%s назад",
		s: "несколько секунд",
		m: relativeTimeWithPlural,
		mm: relativeTimeWithPlural,
		h: "час",
		hh: relativeTimeWithPlural,
		d: "день",
		dd: relativeTimeWithPlural,
		M: "месяц",
		MM: relativeTimeWithPlural,
		y: "год",
		yy: relativeTimeWithPlural
	},
	ordinal: function ordinal(n) {
		return n;
	},
	meridiem: function meridiem(hour) {
		if (hour < 4) return "ночи";
		else if (hour < 12) return "утра";
		else if (hour < 17) return "дня";
		return "вечера";
	}
};
dayjs.locale(locale, null, true);
//#endregion
//#region src/date-time.ts
/**
* ---header-docs-section---
* # yummies/date-time
*
* ## Description
*
* Date and time formatting built on **dayjs** with locale plugins, human-readable presets, relative
* phrases, and duration helpers tailored to UI copy. It composes with `yummies/format`, `yummies/ms`,
* and `yummies/text` for declensions so timestamps read naturally in Russian-facing apps.
*
* ## Usage
*
* ```ts
* import { formatDate } from "yummies/date-time";
* ```
*/
dayjs.extend(relativeTime_default);
dayjs.extend(duration_default);
dayjs.locale("ru");
var toLibFormat = (value) => {
	let result = null;
	if (dayjs.isDayjs(value)) result = value;
	else if (value != null) result = dayjs(value);
	if (!result?.isValid()) return null;
	return result;
};
/**
* Formats a date-like value using predefined presets or a custom pattern.
*
* @example
* ```ts
* formatDate('2024-03-15', { format: 'time-short' });
* ```
*/
var formatDate = (value, settings) => {
	const dateFormat = settings?.format;
	const datePattern = settings?.pattern;
	const asTime = settings?.asTime;
	let libDate = null;
	if (asTime && typeGuard.isNumber(value)) libDate = dayjs.duration(value);
	else libDate = toLibFormat(value);
	if (!libDate) return format.NO_VALUE;
	if (datePattern) return libDate.format(datePattern);
	if (dateFormat === "human" || dateFormat === "spent-time") if ("fromNow" in libDate) return libDate.fromNow(dateFormat === "spent-time");
	else return format.NO_VALUE;
	switch (dateFormat) {
		case "full": return libDate.format("DD MMM YYYY HH:mm:ss");
		case "short": return libDate.format("DD MMM HH:mm");
		case "time": return libDate.format("HH:mm:ss");
		case "time-short": return libDate.format("HH:mm");
		case "day": return libDate.format("DD MMM YYYY");
		case "month": return libDate.format("MMMM YYYY");
		default: return libDate.format("DD.MM.YYYY");
	}
};
/**
* Splits a duration in milliseconds into day-based time parts.
*
* @example
* ```ts
* dayTimeDuration(ms(1, 'day') + ms(2, 'hour'));
* ```
*/
var dayTimeDuration = (timeInMs) => {
	let left = Math.max(timeInMs, 0);
	const days = Math.floor(left / unitsToMs.day);
	left = left % unitsToMs.day;
	const hours = Math.floor(left / unitsToMs.hour);
	left = left % unitsToMs.hour;
	const minutes = Math.floor(left / unitsToMs.min);
	left = left % unitsToMs.min;
	const seconds = Math.floor(left / unitsToMs.sec);
	left = left % unitsToMs.sec;
	return {
		days,
		hours,
		minutes,
		seconds,
		milliseconds: Math.floor(left)
	};
};
/**
* Applies one or more date adjustments and returns a new `Date` instance.
*
* @example
* ```ts
* changeDate(new Date(), 1, 'day', 30, 'minute');
* ```
*/
var changeDate = (date, ...args) => {
	let wrappedDate = toLibFormat(date);
	for (let i = 0; i < args.length; i += 2) {
		const amount = args[i];
		const unit = args[i + 1];
		if (unit != null) wrappedDate = wrappedDate.add(amount, unit);
	}
	return wrappedDate.toDate();
};
/**
* Converts milliseconds into hour-based time parts without keeping full days separately.
*
* @example
* ```ts
* timeDuration(ms(1, 'hour') + ms(30, 'min'));
* ```
*/
var timeDuration = (timeInMs) => {
	const { days, hours, milliseconds, minutes, seconds } = dayTimeDuration(timeInMs);
	return {
		hours: hours + unitsToMs.day * days,
		milliseconds,
		minutes,
		seconds
	};
};
var durationFormatLabels = {
	days: {
		compact: "д",
		full: [
			"день",
			"дня",
			"дней"
		]
	},
	hours: {
		compact: "ч",
		full: [
			"час",
			"часа",
			"часов"
		]
	},
	minutes: {
		compact: "мин",
		full: [
			"минута",
			"минуты",
			"минут"
		]
	},
	seconds: {
		compact: "сек",
		full: [
			"секунда",
			"секунды",
			"секунд"
		]
	}
};
function getFormatDuration(...args) {
	let compact = false;
	let diff = 0;
	if (args[0] instanceof Date) {
		const startedDate = dayjs(args[0]);
		diff = dayjs(args[1]).diff(startedDate, "ms");
		compact = args[2] === true;
	} else {
		diff = args[0];
		compact = args[1] === true;
	}
	const { days, hours, minutes, seconds } = dayTimeDuration(diff);
	const formattedParts = [];
	if (days) if (compact) formattedParts.push(`${days} ${durationFormatLabels.days.compact}`);
	else formattedParts.push(`${days} ${declension(days, durationFormatLabels.days.full)}`);
	if (hours) if (compact) formattedParts.push(`${hours} ${durationFormatLabels.hours.compact}`);
	else formattedParts.push(`${hours} ${declension(hours, durationFormatLabels.hours.full)}`);
	if (minutes) if (compact) formattedParts.push(`${minutes} ${durationFormatLabels.minutes.compact}`);
	else formattedParts.push(`${minutes} ${declension(minutes, durationFormatLabels.minutes.full)}`);
	if (seconds) if (compact) formattedParts.push(`${seconds} ${durationFormatLabels.seconds.compact}`);
	else formattedParts.push(`${seconds} ${declension(seconds, durationFormatLabels.seconds.full)}`);
	return formattedParts.join(" ");
}
/**
* Calculates the difference between two dates in minutes, seconds and total units.
*
* @example
* ```ts
* getTimeDiff(new Date('2024-03-15T12:30:00'), new Date('2024-03-15T12:00:00'));
* ```
*/
var getTimeDiff = (dateA, dateB) => {
	const leftDate = toLibFormat(dateA)?.toDate();
	const rightDate = toLibFormat(dateB)?.toDate();
	if (!leftDate || !rightDate) return {
		minutes: 0,
		seconds: 0,
		total: {
			ms: 0,
			hours: 0
		}
	};
	const msDiff = leftDate.getTime() - rightDate.getTime();
	return {
		minutes: Math.max(Math.floor(msDiff / ms(1, "min")), 0),
		seconds: Math.max(Math.floor(msDiff % ms(1, "min") / 1e3), 0),
		total: {
			hours: Math.round(msDiff / ms(1, "hour")),
			ms: msDiff
		}
	};
};
/**
* Returns a new date shifted forward by the provided number of days.
*
* @example
* ```ts
* addDays(new Date('2024-03-15'), 7);
* ```
*/
var addDays = (date, count) => toLibFormat(date)?.add(count, "d").toDate();
/**
* Returns a new date shifted backward by the provided number of days.
*
* @example
* ```ts
* subtractDays(new Date('2024-03-15'), 7);
* ```
*/
var subtractDays = (date, count) => toLibFormat(date)?.subtract(count, "d").toDate();
/**
* Returns a new date shifted forward by the provided number of minutes.
*
* @example
* ```ts
* addMinutes(new Date('2024-03-15T12:00:00'), 15);
* ```
*/
var addMinutes = (date, count) => toLibFormat(date)?.add(count, "m").toDate();
/**
* Returns a new date with the minutes field replaced.
*
* @example
* ```ts
* setMinutes(new Date('2024-03-15T12:00:00'), 45);
* ```
*/
var setMinutes = (date, minutes) => toLibFormat(date)?.set("m", minutes).toDate();
/**
* Returns a new date with the hours field replaced.
*
* @example
* ```ts
* setHours(new Date('2024-03-15T12:00:00'), 9);
* ```
*/
var setHours = (date, hours) => toLibFormat(date)?.set("h", hours).toDate();
//#endregion
export { addDays, addMinutes, changeDate, dayTimeDuration, formatDate, getFormatDuration, getTimeDiff, setHours, setMinutes, subtractDays, timeDuration };
