// yummies/format.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const require_chunk = require("./chunk-CZWwpsFl.cjs");
let yummies_parser = require("yummies/parser");
let yummies_type_guard = require("yummies/type-guard");
//#region src/format/number.ts
/**
* Formats a numeric value with thousands separators, fractional digit control
* and optional postfix text.
*
* Invalid, empty or unsupported values fall back to `emptyText`.
*
* @param rawValue Number or numeric string to format.
* @param userSettings Formatting overrides merged with `number.defaultSettings`.
* @returns Formatted number string or fallback text.
*
* @example
* ```ts
* number(12000); // '12 000'
* ```
*
* @example
* ```ts
* number(12.5, { digits: 1, postfix: '%' }); // '12.5%'
* ```
*/
var number = (rawValue, userSettings) => {
	const settings = {
		...number.defaultSettings,
		...userSettings
	};
	const digits = settings.digits ?? 0;
	const cutZeros = settings?.cutZeros ?? false;
	const delimiter = settings.delimiter ?? " ";
	const postfix = settings.postfix ?? "";
	const emptyText = settings.emptyText ?? "–";
	const digitsOnlyForFloat = settings.digitsOnlyForFloat ?? true;
	let value;
	if (yummies_type_guard.typeGuard.isString(rawValue)) value = yummies_parser.parser.number(value, { fallback: void 0 });
	else value = rawValue;
	if (yummies_type_guard.typeGuard.isNumber(value)) {
		let raw = `${value}`;
		if (digits !== false) if (settings.cropDigitsOnly) {
			const [integerPart, decimalPart] = `${raw}`.split(".");
			const leftPart = integerPart;
			const rightPart = (decimalPart || "").slice(0, digits).padEnd(digits, "0");
			if (rightPart) raw = `${leftPart}.${rightPart}`;
			else raw = leftPart;
		} else raw = value.toFixed(digits);
		if (cutZeros) raw = `${+raw}`;
		const [integerPart, decimalPart] = raw.split(".", 2);
		let formattedIntegerPart = integerPart;
		let formattedDecimalPart = "";
		if (decimalPart && (!digitsOnlyForFloat || !/^0+$/.test(decimalPart))) formattedDecimalPart = `.${decimalPart}`;
		const rgx = /(\d+)(\d{3})/;
		while (rgx.test(formattedIntegerPart) && delimiter) formattedIntegerPart = formattedIntegerPart.replace(rgx, `$1${delimiter}$2`);
		return formattedIntegerPart + formattedDecimalPart + postfix;
	}
	return emptyText;
};
number.defaultSettings = {};
//#endregion
//#region src/format/percent.ts
/**
* Formats a value as a percent string with configurable decimal precision,
* decimal divider and suffix symbol.
*
* @param value Number or numeric string to format.
* @param settings Parser and formatting options.
* @returns Formatted percent string or fallback text for invalid values.
*
* @example
* ```ts
* percent(12.345); // '12.35%'
* ```
*
* @example
* ```ts
* percent(12.345, { divider: ',', symbol: ' pct' }); // '12,35 pct'
* ```
*/
var percent = (value, settings) => {
	const numericValue = yummies_parser.parser.number(value, {
		...settings,
		digits: settings?.digits ?? 2,
		fallback: NaN
	});
	if (yummies_type_guard.typeGuard.isNumber(numericValue)) {
		const divider = settings?.divider ?? ".";
		return `${divider === "." ? numericValue : `${numericValue}`.replace(".", divider)}${settings?.delimiter ?? ""}${settings?.symbol ?? "%"}`;
	} else return settings?.emptyText ?? "–";
};
//#endregion
//#region src/format/skip-spaces.ts
/**
* Removes all whitespace characters from a string.
*
* @param value Source string.
* @returns String without spaces, tabs and line breaks.
*
* @example
* ```ts
* skipSpaces('1 000 000'); // '1000000'
* ```
*
* @example
* ```ts
* skipSpaces('a\tb\nc'); // 'abc'
* ```
*/
var skipSpaces = (value) => value.replaceAll(/\s/g, "");
//#endregion
//#region src/format/_exports.ts
var _exports_exports = /* @__PURE__ */ require_chunk.__exportAll({
	HYPHEN: () => "-",
	INFINITY: () => "∞",
	NO_VALUE: () => "–",
	number: () => number,
	percent: () => percent,
	skipSpaces: () => skipSpaces
});
//#endregion
Object.defineProperty(exports, "format", {
	enumerable: true,
	get: function() {
		return _exports_exports;
	}
});
