// yummies/html.js
import DOMPurify from "dompurify";
import { blobToUrl } from "yummies/media";
//#region src/html.ts
/**
* ---header-docs-section---
* # yummies/html
*
* ## Description
*
* DOM-centric utilities: sanitizing HTML with **DOMPurify**, computed style probes, downloads via
* temporary anchors, and small string helpers for safe markup. Prefer these over `innerHTML` with
* raw user input; keep CSP and server-side validation as the real security boundary.
*
* ## Usage
*
* ```ts
* import { getComputedColor } from "yummies/html";
* ```
*/
/**
* Extracts an RGB value from any valid CSS color.
*
* Not recommended for frequent use because it triggers a reflow.
*/
var getComputedColor = (color) => {
	if (!color) return null;
	const d = document.createElement("div");
	d.style.color = color;
	document.body.append(d);
	const rgbcolor = globalThis.getComputedStyle(d).color;
	const match = /rgba?\((\d+)\s*,\s*(\d+)\s*,\s*(\d+)\s*(,\s*\d+[.d+]*)*\)/g.exec(rgbcolor);
	d.remove();
	if (!match) return null;
	return `${match[1]}, ${match[2]}, ${match[3]}`;
};
/**
* Triggers a file download by creating and clicking a temporary anchor element.
*
* @example
* ```ts
* downloadUsingAnchor('/report.pdf', 'report.pdf');
* ```
*/
var downloadUsingAnchor = (urlOrBlob, fileName) => {
	const url = blobToUrl(urlOrBlob);
	const a = document.createElement("a");
	a.href = url;
	a.download = fileName ?? "file";
	a.target = "_blank";
	document.body.append(a);
	a.click();
	a.remove();
};
/**
* Surrounds string in an anchor tag
*/
function wrapTextToTagLink(link) {
	const descr = String(link).replace(/^(https?:\/{0,2})?(w{3}\.)?/, "www.");
	if (!/^https?:\/{2}/.test(link)) link = `http://${link}`;
	return `<a href=${link} target="_blank">${descr}</a>`;
}
/**
* Collects the cumulative `offsetTop` value through the element parent chain.
*
* @example
* ```ts
* const offsetTop = collectOffsetTop(document.getElementById('section'));
* ```
*/
var collectOffsetTop = (element) => {
	let offsetTop = 0;
	let node = element;
	while (node != null) {
		offsetTop += node.offsetTop;
		node = node.parentElement;
	}
	return offsetTop;
};
/**
* Prevents the default browser action and stops event propagation.
*
* @example
* ```ts
* button.addEventListener('click', (event) => skipEvent(event));
* ```
*/
var skipEvent = (e) => {
	e.preventDefault();
	e.stopPropagation();
	return false;
};
/**
* Scrolls the page vertically to the viewport section containing the target element.
*
* @example
* ```ts
* globalScrollIntoViewForY(document.getElementById('footer')!);
* ```
*/
var globalScrollIntoViewForY = (node) => {
	const scrollContainer = document.body;
	const pageHeight = window.innerHeight;
	const nodeBounding = node.getBoundingClientRect();
	const scrollPagesCount = scrollContainer.scrollHeight / pageHeight;
	const scrollPageNumber = Math.min(Math.max(nodeBounding.top / pageHeight, 1), scrollPagesCount);
	window.scroll({
		top: scrollPageNumber * pageHeight,
		behavior: "smooth"
	});
};
var sanitizeDefaults = {
	ALLOWED_TAGS: [
		"a",
		"article",
		"b",
		"blockquote",
		"br",
		"caption",
		"code",
		"del",
		"details",
		"div",
		"em",
		"h1",
		"h2",
		"h3",
		"h4",
		"h5",
		"h6",
		"hr",
		"i",
		"img",
		"ins",
		"kbd",
		"li",
		"main",
		"ol",
		"p",
		"pre",
		"section",
		"span",
		"strong",
		"sub",
		"summary",
		"sup",
		"table",
		"tbody",
		"td",
		"th",
		"thead",
		"tr",
		"u",
		"ul"
	],
	ALLOWED_ATTR: [
		"href",
		"target",
		"name",
		"src",
		"class"
	]
};
/**
* Sanitizes HTML using the default allowlist merged with custom DOMPurify config.
*
* @example
* ```ts
* sanitizeHtml('<img src=x onerror=alert(1) />');
* ```
*/
var sanitizeHtml = (html, config) => {
	return DOMPurify.sanitize(html || "", {
		...sanitizeDefaults,
		...config
	});
};
/**
* Checks whether the element is nested inside the provided parent element.
*
* @example
* ```ts
* checkElementHasParent(childElement, modalElement);
* ```
*/
var checkElementHasParent = (element, parent) => {
	let node = element;
	if (!parent) return false;
	while (node != null) if (node === parent) return true;
	else node = node.parentElement;
	return false;
};
/**
* Executes a function within a view transition if supported by the browser.
*
* @param {VoidFunction} fn - The function to be executed.
* @returns {ViewTransition} - The result of the executed function.
*
* @see {@link https://developer.mozilla.org/en-US/docs/Web/API/Document/startViewTransition | MDN: Document.startViewTransition}
*/
var startViewTransitionSafety = (fn, params) => {
	if (typeof document !== "undefined" && document.startViewTransition && !params?.disabled) return document.startViewTransition(fn);
	fn();
};
/**
* Calculates the scrollbar width.
*/
var calcScrollbarWidth = (elementToAppend = document.body) => {
	const outer = document.createElement("div");
	outer.style.visibility = "hidden";
	outer.style.width = "100px";
	outer.style.overflow = "scroll";
	elementToAppend.append(outer);
	const inner = document.createElement("div");
	inner.style.width = "100%";
	outer.append(inner);
	const scrollbarWidth = outer.offsetWidth - inner.offsetWidth;
	outer.parentNode?.removeChild(outer);
	return scrollbarWidth;
};
/**
* Calculates the inner height of an HTML element, accounting for padding.
*/
function getElementInnerHeight(element) {
	const { clientHeight } = element;
	const { paddingTop, paddingBottom } = getComputedStyle(element);
	return clientHeight - Number.parseFloat(paddingTop) - Number.parseFloat(paddingBottom);
}
/**
* Calculates the inner width of an HTML element, accounting for padding.
*/
function getElementInnerWidth(el) {
	const { clientWidth } = el;
	const { paddingLeft, paddingRight } = getComputedStyle(el);
	return clientWidth - Number.parseFloat(paddingLeft) - Number.parseFloat(paddingRight);
}
/**
* Checks whether the user prefers a dark color scheme.
*
* @example
* ```ts
* const prefersDark = isPrefersDarkTheme();
* ```
*/
var isPrefersDarkTheme = () => !!globalThis.matchMedia?.("(prefers-color-scheme: dark)")?.matches;
/**
* Checks whether the user prefers a light color scheme.
*
* @example
* ```ts
* const prefersLight = isPrefersLightTheme();
* ```
*/
var isPrefersLightTheme = () => !!globalThis.matchMedia?.("(prefers-color-scheme: light)")?.matches;
//#endregion
export { calcScrollbarWidth, checkElementHasParent, collectOffsetTop, downloadUsingAnchor, getComputedColor, getElementInnerHeight, getElementInnerWidth, globalScrollIntoViewForY, isPrefersDarkTheme, isPrefersLightTheme, sanitizeHtml, skipEvent, startViewTransitionSafety, wrapTextToTagLink };
