// yummies/random.js
//#region src/random.ts
/**
* ---header-docs-section---
* # yummies/random
*
* ## Description
*
* Small RNG helpers for UI demos, games, and sampling: floats, integers, and random choices from
* arrays. They wrap `Math.random` (not cryptographically secure) so prefer platform `crypto` when
* generating secrets, tokens, or lottery outcomes.
*
* ## Usage
*
* ```ts
* import { getRandomInt, getRandomChoice } from "yummies/random";
* ```
*/
/**
* Returns a random floating-point number between `min` and `max`.
*
* @example
* ```ts
* const value = getRandomFloat(1, 10);
* ```
*/
var getRandomFloat = (min = 0, max = 1) => Math.random() * (max - min) + min;
/**
* Returns a random integer between `min` and `max`.
*
* @example
* ```ts
* const value = getRandomInt(1, 10);
* ```
*/
var getRandomInt = (min = 0, max = 1) => min === max ? min : Math.round(getRandomFloat(min, max));
/**
* Picks a random element from the provided array.
*
* @example
* ```ts
* const fruit = getRandomChoice(['apple', 'banana', 'orange']);
* ```
*/
var getRandomChoice = (arr) => arr[getRandomInt(0, arr.length - 1)];
/**
* Creates an array filled with `null` values using a random length.
*
* @example
* ```ts
* const items = getRandomSizeArray(2, 5);
* ```
*/
var getRandomSizeArray = (min = 0, max = 10) => Array.from({ length: getRandomInt(min, max) }).fill(null);
/**
* Returns a uniformly random boolean.
*
* @example
* ```ts
* const value = getRandomBool();
* ```
*/
var getRandomBool = () => getRandomInt(0, 1) === 1;
/**
* Returns `true` more often than `false`.
*
* @example
* ```ts
* const value = getMajorRandomBool();
* ```
*/
var getMajorRandomBool = () => {
	return getRandomInt(0, 10) <= 6;
};
/**
* Returns `true` less often than `false`.
*
* @example
* ```ts
* const value = getMinorRandomBool();
* ```
*/
var getMinorRandomBool = () => {
	return !getMajorRandomBool();
};
/**
* Returns `true` with the provided probability from `0` to `1`.
*
* @example
* ```ts
* const shouldRun = getFrequencyValue(0.25);
* ```
*/
var getFrequencyValue = (frequency) => {
	return Math.random() < frequency;
};
//#endregion
export { getFrequencyValue, getMajorRandomBool, getMinorRandomBool, getRandomBool, getRandomChoice, getRandomFloat, getRandomInt, getRandomSizeArray };
