// yummies/type-guard.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
//#region src/type-guard/_exports.ts
var _exports_exports = /* @__PURE__ */ require("./chunk-CZWwpsFl.cjs").__exportAll({
	isArray: () => isArray,
	isBoolean: () => isBoolean,
	isDefined: () => isDefined,
	isElement: () => isElement,
	isFalsy: () => isFalsy,
	isFunction: () => isFunction,
	isInfinite: () => isInfinite,
	isNaN: () => isNaN,
	isNull: () => isNull,
	isNumber: () => isNumber,
	isObject: () => isObject,
	isRegExp: () => isRegExp,
	isString: () => isString,
	isSymbol: () => isSymbol,
	isTruthy: () => isTruthy,
	isUndefined: () => isUndefined
});
var TYPE = {
	Null: "null",
	Undefined: "undefined",
	NaN: "nan",
	Object: "[object Object]",
	Array: "[object Array]",
	String: "[object String]",
	Number: "[object Number]",
	Boolean: "[object Boolean]",
	Function: "[object Function]",
	AsyncFunction: "[object AsyncFunction]",
	RegExp: "[object RegExp]",
	Symbol: "[object Symbol]",
	Infinite: "infinite",
	Element: "element"
};
function getType(value) {
	if (value === void 0) return TYPE.Undefined;
	if (value === null) return TYPE.Null;
	if (value && (value.nodeType === 1 || value.nodeType === 9)) return TYPE.Element;
	const stringifiedValue = Object.prototype.toString.call(value);
	if (stringifiedValue === TYPE.Number) {
		if (Number.isNaN(value)) return TYPE.NaN;
		if (!Number.isFinite(value)) return TYPE.Infinite;
	}
	return stringifiedValue;
}
var createTypeGuard = (...types) => (value) => types.includes(getType(value));
/**
* Checks that a value is neither `null` nor `undefined`.
*
* @template T Value type without nullish branches.
* @param value Value to test.
* @returns `true` when the value is defined.
*
* @example
* ```ts
* isDefined(0); // true
* ```
*
* @example
* ```ts
* isDefined(null); // false
* ```
*/
var isDefined = (value) => value != null;
/**
* Checks whether a value is truthy (`!!value`).
*
* @param value Value to test.
* @returns `true` when `!!value` is true.
*
* @example
* ```ts
* isTruthy(1); // true
* ```
*
* @example
* ```ts
* isTruthy(0); // false
* ```
*/
var isTruthy = (value) => !!value;
/**
* Checks whether a value is falsy (`!Boolean(value)`).
*
* @param value Value to test.
* @returns `true` when `Boolean(value)` is false.
*
* @example
* ```ts
* isFalsy(null); // true
* ```
*
* @example
* ```ts
* isFalsy("x"); // false
* ```
*/
var isFalsy = (value) => !value;
/**
* Checks whether a value is exactly `null`.
*
* @param value Value to test.
* @returns `true` when the value is `null`.
*
* @example
* ```ts
* isNull(null); // true
* ```
*
* @example
* ```ts
* isNull(undefined); // false
* ```
*/
var isNull = createTypeGuard(TYPE.Null);
/**
* Checks whether a value is exactly `undefined`.
*
* @param value Value to test.
* @returns `true` when the value is `undefined`.
*
* @example
* ```ts
* isUndefined(undefined); // true
* ```
*
* @example
* ```ts
* isUndefined('value'); // false
* ```
*/
var isUndefined = createTypeGuard(TYPE.Undefined);
/**
* Checks whether a value is a plain object.
*
* @param value Value to test.
* @returns `true` when the value matches `[object Object]`.
*
* @example
* ```ts
* isObject({ id: 1 }); // true
* ```
*
* @example
* ```ts
* isObject([]); // false
* ```
*/
var isObject = createTypeGuard(TYPE.Object);
/**
* Checks whether a value is an array.
*
* @param value Value to test.
* @returns `true` when the value is an array.
*
* @example
* ```ts
* isArray([1, 2, 3]); // true
* ```
*
* @example
* ```ts
* isArray({ length: 1 }); // false
* ```
*/
var isArray = createTypeGuard(TYPE.Array);
/**
* Checks whether a value is a string object or primitive string.
*
* @param value Value to test.
* @returns `true` when the value is a string.
*
* @example
* ```ts
* isString('hello'); // true
* ```
*
* @example
* ```ts
* isString(123); // false
* ```
*/
var isString = createTypeGuard(TYPE.String);
/**
* Checks whether a value is a finite number.
*
* Unlike `isNaN` and `isInfinite`, this guard only matches regular numeric values.
*
* @param value Value to test.
* @returns `true` when the value is a non-NaN finite number.
*
* @example
* ```ts
* isNumber(123); // true
* ```
*
* @example
* ```ts
* isNumber(Number.NaN); // false
* ```
*/
var isNumber = createTypeGuard(TYPE.Number);
/**
* Checks whether a value is a boolean.
*
* @param value Value to test.
* @returns `true` when the value is a boolean.
*
* @example
* ```ts
* isBoolean(true); // true
* ```
*
* @example
* ```ts
* isBoolean('true'); // false
* ```
*/
var isBoolean = createTypeGuard(TYPE.Boolean);
/**
* Checks whether a value is a synchronous or asynchronous function.
*
* @param value Value to test.
* @returns `true` when the value is a function.
*
* @example
* ```ts
* isFunction(() => {}); // true
* ```
*
* @example
* ```ts
* isFunction(async () => {}); // true
* ```
*/
var isFunction = createTypeGuard(TYPE.Function, TYPE.AsyncFunction);
/**
* Checks whether a value is a regular expression.
*
* @param value Value to test.
* @returns `true` when the value is a `RegExp`.
*
* @example
* ```ts
* isRegExp(/foo/); // true
* ```
*
* @example
* ```ts
* isRegExp('foo'); // false
* ```
*/
var isRegExp = createTypeGuard(TYPE.RegExp);
/**
* Checks whether a value looks like a DOM element or document node.
*
* @param value Value to test.
* @returns `true` when the value has an element-like node type.
*
* @example
* ```ts
* isElement(document.body); // true
* ```
*
* @example
* ```ts
* isElement({ nodeType: 3 }); // false
* ```
*/
var isElement = createTypeGuard(TYPE.Element);
/**
* Checks whether a value is `NaN`.
*
* @param value Value to test.
* @returns `true` when the value is `NaN`.
*
* @example
* ```ts
* isNaN(Number.NaN); // true
* ```
*
* @example
* ```ts
* isNaN(5); // false
* ```
*/
var isNaN = createTypeGuard(TYPE.NaN);
/**
* Checks whether a value is positive or negative infinity.
*
* @param value Value to test.
* @returns `true` when the value is not finite.
*
* @example
* ```ts
* isInfinite(Infinity); // true
* ```
*
* @example
* ```ts
* isInfinite(10); // false
* ```
*/
var isInfinite = createTypeGuard(TYPE.Infinite);
/**
* Checks whether a value is a symbol.
*
* @param value Value to test.
* @returns `true` when the value is a symbol.
*
* @example
* ```ts
* isSymbol(Symbol('id')); // true
* ```
*
* @example
* ```ts
* isSymbol('id'); // false
* ```
*/
var isSymbol = createTypeGuard(TYPE.Symbol);
//#endregion
Object.defineProperty(exports, "typeGuard", {
	enumerable: true,
	get: function() {
		return _exports_exports;
	}
});
