// swagger-typescript-api/dist/src-UwNroIF-.cjs
//#region \0rolldown/runtime.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __exportAll = (all, no_symbols) => {
	let target = {};
	for (var name in all) __defProp(target, name, {
		get: all[name],
		enumerable: true
	});
	if (!no_symbols) __defProp(target, Symbol.toStringTag, { value: "Module" });
	return target;
};
var __copyProps = (to, from, except, desc) => {
	if (from && typeof from === "object" || typeof from === "function") for (var keys = __getOwnPropNames(from), i = 0, n = keys.length, key; i < n; i++) {
		key = keys[i];
		if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
			get: ((k) => from[k]).bind(null, key),
			enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
		});
	}
	return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
	value: mod,
	enumerable: true
}) : target, mod));
//#endregion
let consola = require("consola");
consola = __toESM(consola, 1);
let es_toolkit = require("es-toolkit");
es_toolkit = __toESM(es_toolkit, 1);
let es_toolkit_compat = require("es-toolkit/compat");
es_toolkit_compat = __toESM(es_toolkit_compat, 1);
let typescript = require("typescript");
typescript = __toESM(typescript, 1);
let node_path = require("node:path");
node_path = __toESM(node_path, 1);
let _biomejs_js_api = require("@biomejs/js-api");
let nanoid = require("nanoid");
nanoid = __toESM(nanoid, 1);
let yummies_type_guard = require("yummies/type-guard");
let yummies_common = require("yummies/common");
let node_crypto = require("node:crypto");
node_crypto = __toESM(node_crypto, 1);
let swagger2openapi = require("swagger2openapi");
swagger2openapi = __toESM(swagger2openapi, 1);
let node_fs = require("node:fs");
node_fs = __toESM(node_fs, 1);
let _apidevtools_swagger_parser = require("@apidevtools/swagger-parser");
_apidevtools_swagger_parser = __toESM(_apidevtools_swagger_parser, 1);
let yaml = require("yaml");
yaml = __toESM(yaml, 1);
let node_module = require("node:module");
node_module = __toESM(node_module, 1);
let node_url = require("node:url");
node_url = __toESM(node_url, 1);
let eta = require("eta");
//#region src/code-formatter.ts
var CodeFormatter = class {
	config;
	constructor(config) {
		this.config = config;
	}
	removeUnusedImports = (content) => {
		const tempFileName = "file.ts";
		const host = new TsLanguageServiceHost(tempFileName, content);
		const fileTextChanges = typescript.createLanguageService(host).organizeImports({
			type: "file",
			fileName: tempFileName
		}, { newLineCharacter: typescript.sys.newLine }, void 0)[0];
		if (fileTextChanges?.textChanges.length) return fileTextChanges.textChanges.reduceRight((content, { span, newText }) => `${content.slice(0, span.start)}${newText}${content.slice(span.start + span.length)}`, content);
		return content;
	};
	format = async (content) => {
		const biome = await _biomejs_js_api.Biome.create({ distribution: _biomejs_js_api.Distribution.NODE });
		const biomeProject = biome.openProject();
		biome.applyConfiguration(biomeProject.projectKey, {
			files: { maxSize: Number.MAX_SAFE_INTEGER },
			formatter: { indentStyle: "space" }
		});
		return biome.formatContent(biomeProject.projectKey, content, { filePath: node_path.format({
			name: nanoid.nanoid(),
			ext: "ts"
		}) }).content;
	};
	formatCode = async (code, { removeUnusedImports = true, format = true } = {}) => {
		if (removeUnusedImports) code = this.removeUnusedImports(code);
		if (format) code = await this.format(code);
		return code;
	};
};
var TsLanguageServiceHost = class {
	fileName;
	content;
	compilerOptions;
	constructor(fileName, content) {
		this.fileName = fileName;
		this.content = content;
		const tsconfig = typescript.findConfigFile(fileName, typescript.sys.fileExists);
		this.compilerOptions = tsconfig ? typescript.convertCompilerOptionsFromJson(typescript.readConfigFile(tsconfig, typescript.sys.readFile).config.compilerOptions, "").options : typescript.getDefaultCompilerOptions();
	}
	getNewLine() {
		return "newLine" in typescript.sys ? typescript.sys.newLine : "\n";
	}
	getScriptFileNames() {
		return [this.fileName];
	}
	getCompilationSettings() {
		return this.compilerOptions;
	}
	getDefaultLibFileName() {
		return typescript.getDefaultLibFileName(this.getCompilationSettings());
	}
	getCurrentDirectory() {
		return process.cwd();
	}
	getScriptVersion() {
		return typescript.version;
	}
	getScriptSnapshot() {
		return typescript.ScriptSnapshot.fromString(this.content);
	}
	readFile(fileName, encoding) {
		if (fileName === this.fileName) return this.content;
		return typescript.sys.readFile(fileName, encoding);
	}
	fileExists(path) {
		return typescript.sys.fileExists(path);
	}
};
//#endregion
//#region src/util/name-resolver.ts
var NameResolver = class {
	reservedNames = [];
	getFallbackName;
	config;
	constructor(config, reservedNames, getFallbackName) {
		this.config = config;
		this.getFallbackName = getFallbackName;
		this.reserve(reservedNames);
	}
	reserve(names) {
		const fixedNames = (0, es_toolkit.uniq)((0, es_toolkit.compact)(names));
		for (const name of fixedNames) if (this.reservedNames.indexOf(name) === -1) this.reservedNames.push(name);
	}
	unreserve(names) {
		this.reservedNames = this.reservedNames.filter((reservedName) => !names.some((name) => name === reservedName));
	}
	isReserved(name) {
		return this.reservedNames.some((reservedName) => reservedName === name);
	}
	resolve(variants, resolver, extras, shouldReserve = true) {
		if (typeof resolver === "function") {
			let usageName = null;
			while (usageName === null) {
				const variant = resolver(variants, extras);
				if (variant === void 0) {
					consola.consola.warn("unable to resolve name. current reserved names: ", ...this.reservedNames);
					return null;
				}
				if (!shouldReserve || !this.isReserved(variant)) usageName = variant;
			}
			shouldReserve && this.reserve([usageName]);
			return usageName;
		}
		if (Array.isArray(variants)) {
			let usageName = null;
			const uniqVariants = (0, es_toolkit.uniq)((0, es_toolkit.compact)(variants));
			for (const variant of uniqVariants) if (!usageName && (!shouldReserve || !this.isReserved(variant))) usageName = variant;
			if (usageName) {
				shouldReserve && this.reserve([usageName]);
				return usageName;
			}
			consola.consola.debug("trying to resolve name with using fallback name generator using variants", ...variants);
			return this.resolve(variants, this.getFallbackName, extras);
		}
		consola.consola.debug("problem with reserving names. current reserved names: ", ...this.reservedNames);
		return null;
	}
};
//#endregion
//#region src/component-type-name-resolver.ts
var ComponentTypeNameResolver = class extends NameResolver {
	counter = 1;
	fallbackNameCounter = 1;
	countersByVariant = /* @__PURE__ */ new Map();
	constructor(config, reservedNames) {
		super(config, reservedNames, (variants) => {
			const baseVariant = (0, es_toolkit.uniq)((0, es_toolkit.compact)(variants)).sort((a, b) => a.localeCompare(b))[0];
			if (baseVariant) {
				if (!this.countersByVariant.has(baseVariant)) this.countersByVariant.set(baseVariant, 0);
				const variantCounter = this.countersByVariant.get(baseVariant) + 1;
				this.countersByVariant.set(baseVariant, variantCounter);
				const dirtyResolvedName = `${baseVariant}${variantCounter}`;
				consola.consola.debug("generated dirty resolved type name for component - ", dirtyResolvedName);
				return dirtyResolvedName;
			}
			const fallbackName = `${this.config.componentTypeNameResolver}${this.fallbackNameCounter++}`;
			consola.consola.debug("generated fallback type name for component - ", fallbackName);
			return fallbackName;
		});
	}
};
//#endregion
//#region package.json
var name = "swagger-typescript-api";
var version = "13.7.2";
var description = "Generate the API client for Fetch or Axios from an OpenAPI Specification";
//#endregion
//#region src/constants.ts
var constants_exports = /* @__PURE__ */ __exportAll({
	DEFAULT_BODY_ARG_NAME: () => DEFAULT_BODY_ARG_NAME,
	FILE_PREFIX: () => FILE_PREFIX,
	HTTP_CLIENT: () => HTTP_CLIENT,
	PROJECT_VERSION: () => PROJECT_VERSION,
	RESERVED_BODY_ARG_NAMES: () => RESERVED_BODY_ARG_NAMES,
	RESERVED_HEADER_ARG_NAMES: () => RESERVED_HEADER_ARG_NAMES,
	RESERVED_PATH_ARG_NAMES: () => RESERVED_PATH_ARG_NAMES,
	RESERVED_QUERY_ARG_NAMES: () => RESERVED_QUERY_ARG_NAMES,
	RESERVED_REQ_PARAMS_ARG_NAMES: () => RESERVED_REQ_PARAMS_ARG_NAMES,
	SCHEMA_TYPES: () => SCHEMA_TYPES$1
});
const DEFAULT_BODY_ARG_NAME = "data";
const FILE_PREFIX = `/* eslint-disable */
/* tslint:disable */
// @ts-nocheck
/*
 * ---------------------------------------------------------------
 * ## THIS FILE WAS GENERATED VIA SWAGGER-TYPESCRIPT-API        ##
 * ##                                                           ##
 * ## AUTHOR: acacode                                           ##
 * ## SOURCE: https://github.com/acacode/swagger-typescript-api ##
 * ---------------------------------------------------------------
 */

`;
const HTTP_CLIENT = {
	FETCH: "fetch",
	AXIOS: "axios"
};
const PROJECT_VERSION = version;
const RESERVED_BODY_ARG_NAMES = [
	"data",
	"body",
	"reqBody"
];
const RESERVED_HEADER_ARG_NAMES = ["headers", "headersParams"];
const RESERVED_PATH_ARG_NAMES = ["path", "pathParams"];
const RESERVED_QUERY_ARG_NAMES = [
	"query",
	"queryParams",
	"queryArg"
];
const RESERVED_REQ_PARAMS_ARG_NAMES = [
	"params",
	"requestParams",
	"reqParams",
	"httpParams"
];
const SCHEMA_TYPES$1 = {
	ARRAY: "array",
	OBJECT: "object",
	ENUM: "enum",
	REF: "$ref",
	PRIMITIVE: "primitive",
	COMPLEX: "complex",
	DISCRIMINATOR: "discriminator",
	COMPLEX_ONE_OF: "oneOf",
	COMPLEX_ANY_OF: "anyOf",
	COMPLEX_ALL_OF: "allOf",
	COMPLEX_NOT: "not",
	COMPLEX_UNKNOWN: "__unknown"
};
//#endregion
//#region src/util/object-assign.ts
const objectAssign = (target, updater) => {
	if (!updater) return;
	const update = typeof updater === "function" ? updater(target) : updater;
	if (!update) return;
	const undefinedKeys = Object.entries(update).filter(([, value]) => value === void 0).map(([key]) => key);
	(0, es_toolkit.merge)(target, update);
	for (const key of undefinedKeys) target[key] = void 0;
};
//#endregion
//#region src/configuration.ts
const TsKeyword = {
	Number: "number",
	String: "string",
	Boolean: "boolean",
	Any: "any",
	Void: "void",
	Unknown: "unknown",
	Null: "null",
	Undefined: "undefined",
	Object: "object",
	File: "File",
	Blob: "Blob",
	Date: "Date",
	Type: "type",
	Enum: "enum",
	Interface: "interface",
	Array: "Array",
	Record: "Record",
	Intersection: "&",
	Union: "|"
};
const TsCodeGenKeyword = { UtilRequiredKeys: "UtilRequiredKeys" };
var CodeGenConfig = class {
	version = PROJECT_VERSION;
	/** CLI flag */
	templates = "";
	/** CLI flag */
	generateResponses = false;
	/** CLI flag */
	defaultResponseAsSuccess = false;
	/** CLI flag */
	generateRouteTypes = false;
	/** CLI flag */
	generateClient = true;
	/** CLI flag */
	generateUnionEnums = false;
	/** CLI flag */
	addReadonly = false;
	enumNamesAsValues = false;
	/** parsed swagger schema from getSwaggerObject() */
	/** parsed swagger schema ref */
	swaggerSchema = null;
	/** original (converted to json) swagger schema ref */
	originalSchema = null;
	/** { "#/components/schemas/Foo": @TypeInfo, ... } */
	componentsMap = {};
	/** flag for catching conversion from swagger 2.0 */
	convertedFromSwagger2 = false;
	/** url index from paths used for merging into modules */
	moduleNameIndex = 0;
	/** use the first tag for the module name */
	moduleNameFirstTag = false;
	extractRequestParams = false;
	extractRequestBody = false;
	extractResponseBody = false;
	extractResponseError = false;
	extractResponses = false;
	extractEnums = false;
	fileNames = {
		dataContracts: "data-contracts",
		routeTypes: "route-types",
		httpClient: "http-client",
		outOfModuleApi: "Common"
	};
	routeNameDuplicatesMap = /* @__PURE__ */ new Map();
	hooks = {
		onPreBuildRoutePath: (_routePath) => void 0,
		onBuildRoutePath: (_routeData) => void 0,
		onInsertPathParam: (_pathParam) => void 0,
		onCreateComponent: (schema) => schema,
		onPreParseSchema: (_originalSchema, _typeName, _schemaType) => void 0,
		onParseSchema: (_originalSchema, parsedSchema) => parsedSchema,
		onCreateRoute: (routeData) => routeData,
		onInit: (config, _codeGenProcess) => config,
		onPrepareConfig: (apiConfig) => apiConfig,
		onCreateRequestParams: (_rawType) => {},
		onCreateRouteName: () => {},
		onFormatTypeName: (_typeName, _rawTypeName, _schemaType) => {},
		onFormatRouteName: (_routeInfo, _templateRouteName) => {}
	};
	resolvedSwaggerSchema;
	defaultResponseType;
	singleHttpClient = false;
	httpClientType = HTTP_CLIENT.FETCH;
	unwrapResponseData = false;
	disableThrowOnError = false;
	disableFormatTypeNames = false;
	sortTypes = false;
	sortRoutes = false;
	templatePaths = {
		/** `templates/base` */
		base: "",
		/** `templates/default` */
		default: "",
		/** `templates/modular` */
		modular: "",
		/** usage path if `--templates` option is not set */
		original: "",
		/** custom path to templates (`--templates`) */
		custom: ""
	};
	/** Record<templateName, templateContent> */
	templatesToRender = {
		api: "",
		dataContracts: "",
		dataContractJsDoc: "",
		interfaceDataContract: "",
		typeDataContract: "",
		enumDataContract: "",
		objectFieldJsDoc: "",
		httpClient: "",
		routeTypes: "",
		routeName: ""
	};
	schemaParsers = {};
	toJS = false;
	silent = false;
	typePrefix = "";
	typeSuffix = "";
	typeNameSeparator = "_";
	enumKeyPrefix = "";
	enumKeySuffix = "";
	patch = false;
	componentTypeNameResolver;
	/** name of the main exported class */
	apiClassName = "Api";
	debug = false;
	anotherArrayType = false;
	internalTemplateOptions = { addUtilRequiredKeysType: false };
	extraTemplates = [];
	input = "";
	modular = false;
	output = "";
	url = "";
	cleanOutput = false;
	spec = null;
	fileName = "Api.ts";
	authorizationToken;
	requestOptions = null;
	jsPrimitiveTypes = [];
	jsEmptyTypes = [];
	fixInvalidTypeNamePrefix = "Type";
	fixInvalidEnumKeyPrefix = "Value";
	enumKeyResolverName = "Value";
	typeNameResolverName = "ComponentType";
	specificArgNameResolverName = "arg";
	successResponseStatusRange = [200, 299];
	extractingOptions = {
		requestBodySuffix: [
			"Payload",
			"Body",
			"Input"
		],
		requestParamsSuffix: ["Params"],
		responseBodySuffix: [
			"Data",
			"Result",
			"Output"
		],
		responseErrorSuffix: [
			"Error",
			"Fail",
			"Fails",
			"ErrorData",
			"HttpError",
			"BadResponse"
		],
		enumSuffix: ["Enum"],
		discriminatorMappingSuffix: [
			"Mapping",
			"Mapper",
			"MapType"
		],
		discriminatorAbstractPrefix: [
			"Base",
			"Abstract",
			"Discriminator",
			"Internal",
			"Polymorph"
		]
	};
	compilerTsConfig = {
		module: typescript.ModuleKind.ESNext,
		noImplicitReturns: true,
		alwaysStrict: true,
		target: typescript.ScriptTarget.ESNext,
		declaration: true,
		noImplicitAny: false,
		sourceMap: false,
		removeComments: false,
		disableSizeLimit: true,
		esModuleInterop: true,
		emitDecoratorMetadata: true,
		skipLibCheck: true
	};
	customTranslator;
	Ts = {
		Keyword: structuredClone(TsKeyword),
		CodeGenKeyword: structuredClone(TsCodeGenKeyword),
		/**
		* $A[] or Array<$A>
		*/
		ArrayType: (content) => {
			if (this.anotherArrayType) return this.Ts.TypeWithGeneric(this.Ts.Keyword.Array, [content]);
			return `${this.Ts.ExpressionGroup(content)}[]`;
		},
		/**
		* "$A"
		*/
		StringValue: (content) => `"${content}"`,
		/**
		* $A
		*/
		BooleanValue: (content) => `${content}`,
		/**
		* $A
		*/
		NumberValue: (content) => `${content}`,
		/**
		* $A
		*/
		NullValue: () => "null",
		/**
		* $A1 | $A2
		*/
		UnionType: (contents) => (0, es_toolkit.uniq)(contents).join(` ${this.Ts.Keyword.Union} `),
		/**
		* ($A1)
		*/
		ExpressionGroup: (content) => content ? `(${content})` : "",
		/**
		* $A1 & $A2
		*/
		IntersectionType: (contents) => (0, es_toolkit.uniq)(contents).join(` ${this.Ts.Keyword.Intersection} `),
		/**
		* Record<$A1, $A2>
		*/
		RecordType: (key, value) => this.Ts.TypeWithGeneric(this.Ts.Keyword.Record, [key, value]),
		/**
		* readonly $key?:$value
		*/
		TypeField: ({ readonly, key, optional, value }) => (0, es_toolkit.compact)([
			readonly && "readonly ",
			key,
			optional && "?",
			": ",
			value
		]).join(""),
		/**
		* [key: $A1]: $A2
		*/
		InterfaceDynamicField: (key, value) => `[key: ${key}]: ${value}`,
		/**
		* EnumName.EnumKey
		*/
		EnumUsageKey: (enumStruct, key) => `${enumStruct}.${key}`,
		/**
		* $A1 = $A2
		*/
		EnumField: (key, value) => `${key} = ${value}`,
		/**
		* /\** description \*\/
		*/
		EnumFieldDescription: (description) => {
			if (description) return `  /** ${description} */`;
			else return "";
		},
		/**
		* /\** $A0.description \*\/
		* $A0.key = $A0.value,
		* /\** $A1.description \*\/
		* $A1.key = $A1.value,
		* /\** $AN.description \*\/
		* $AN.key = $AN.value,
		*/
		EnumFieldsWrapper: (contents) => contents.map(({ key, value, description }) => {
			return (0, es_toolkit.compact)([this.Ts.EnumFieldDescription(description), `  ${this.Ts.EnumField(key, value)}`]).join("\n");
		}).join(",\n"),
		/**
		* {\n $A \n}
		*/
		ObjectWrapper: (content) => `{\n${content}\n}`,
		/**
		* /** $A *\/
		*/
		MultilineComment: (contents, formatFn) => [...contents.length === 1 ? [`/** ${contents[0]} */`] : [
			"/**",
			...contents.map((content) => ` * ${content}`),
			" */"
		]].map((part) => `${formatFn ? formatFn(part) : part}\n`),
		/**
		* $A1<...$A2.join(,)>
		*/
		TypeWithGeneric: (typeName, genericArgs) => {
			return `${typeName}${genericArgs.length ? `<${genericArgs.join(",")}>` : ""}`;
		},
		/**
		* [$A1, $A2, ...$AN]
		*/
		Tuple: (values) => {
			return `[${values.join(", ")}]`;
		}
	};
	/**
	* swagger schema type -> typescript type
	* https://json-schema.org/understanding-json-schema/reference/string.html#dates-and-times
	*/
	primitiveTypes = {
		integer: () => this.Ts.Keyword.Number,
		number: () => this.Ts.Keyword.Number,
		boolean: () => this.Ts.Keyword.Boolean,
		object: () => this.Ts.Keyword.Object,
		file: () => this.Ts.Keyword.File,
		string: {
			$default: this.Ts.Keyword.String,
			/** formats */
			binary: () => this.Ts.Keyword.File,
			byte: () => this.Ts.Keyword.Blob,
			file: () => this.Ts.Keyword.File,
			"date-time": () => this.Ts.Keyword.String,
			time: () => this.Ts.Keyword.String,
			date: () => this.Ts.Keyword.String,
			duration: () => this.Ts.Keyword.String,
			email: () => this.Ts.Keyword.String,
			"idn-email": () => this.Ts.Keyword.String,
			"idn-hostname": () => this.Ts.Keyword.String,
			ipv4: () => this.Ts.Keyword.String,
			ipv6: () => this.Ts.Keyword.String,
			uuid: () => this.Ts.Keyword.String,
			uri: () => this.Ts.Keyword.String,
			"uri-reference": () => this.Ts.Keyword.String,
			"uri-template": () => this.Ts.Keyword.String,
			"json-pointer": () => this.Ts.Keyword.String,
			"relative-json-pointer": () => this.Ts.Keyword.String,
			regex: () => this.Ts.Keyword.String
		}
	};
	templateInfos = [
		{
			name: "api",
			fileName: "api"
		},
		{
			name: "dataContracts",
			fileName: "data-contracts"
		},
		{
			name: "dataContractJsDoc",
			fileName: "data-contract-jsdoc"
		},
		{
			name: "interfaceDataContract",
			fileName: "interface-data-contract"
		},
		{
			name: "typeDataContract",
			fileName: "type-data-contract"
		},
		{
			name: "enumDataContract",
			fileName: "enum-data-contract"
		},
		{
			name: "objectFieldJsDoc",
			fileName: "object-field-jsdoc"
		},
		{
			name: "httpClient",
			fileName: "http-client"
		},
		{
			name: "routeTypes",
			fileName: "route-types"
		},
		{
			name: "routeName",
			fileName: "route-name"
		}
	];
	templateExtensions = [".eta", ".ejs"];
	constructor({ codeGenConstructs, primitiveTypeConstructs, constants, templateInfos, hooks, ...otherConfig }) {
		objectAssign(this.Ts, codeGenConstructs);
		objectAssign(this.primitiveTypes, primitiveTypeConstructs);
		this.defaultResponseType = this.Ts.Keyword.Void;
		this.update({
			...otherConfig,
			hooks: (0, es_toolkit.merge)(this.hooks, hooks || {}),
			constants: {
				...constants_exports,
				...constants
			},
			templateInfos: templateInfos || this.templateInfos
		});
		this.jsPrimitiveTypes = [
			this.Ts.Keyword.Number,
			this.Ts.Keyword.String,
			this.Ts.Keyword.Boolean
		];
		this.jsEmptyTypes = [this.Ts.Keyword.Null, this.Ts.Keyword.Undefined];
		this.componentTypeNameResolver = new ComponentTypeNameResolver(this, []);
	}
	update = (update) => {
		objectAssign(this, update);
		if (this.enumNamesAsValues) this.extractEnums = true;
	};
};
//#endregion
//#region src/util/pascal-case.ts
function pascalCase(value) {
	return (0, es_toolkit_compat.upperFirst)((0, es_toolkit_compat.camelCase)(value));
}
//#endregion
//#region src/schema-components-map.ts
var SchemaComponentsMap = class {
	_data = [];
	constructor(config) {
		this.config = config;
	}
	clear() {
		this._data = [];
	}
	createRef = (paths) => {
		return ["#", ...paths].join("/");
	};
	parseRef = (ref) => {
		return ref.split("/");
	};
	createComponentDraft($ref, rawTypeData) {
		if (yummies_type_guard.typeGuard.isObject(rawTypeData) && rawTypeData.typeName && rawTypeData.rawTypeData && rawTypeData.$ref) return rawTypeData;
		const parsed = this.parseRef($ref);
		const [, rawPointer = ""] = $ref.split("#");
		const pointerParts = (rawPointer.startsWith("/") ? rawPointer : `/${rawPointer}`).split("/").filter(Boolean);
		const typeName = pointerParts.at(-1) || parsed.at(-1) || "Unknown";
		const rawComponentName = pointerParts.at(-2) || parsed[parsed.length - 2] || "schemas";
		return {
			$ref,
			typeName,
			rawTypeData,
			componentName: rawComponentName === "definitions" ? "schemas" : rawComponentName,
			/** result from schema parser */
			typeData: null
		};
	}
	createComponent($ref, rawTypeData, addAtStart) {
		const componentSchema = this.createComponentDraft($ref, rawTypeData);
		const usageComponent = this.config.hooks.onCreateComponent(componentSchema) || componentSchema;
		const refIndex = this._data.findIndex((c) => c.$ref === $ref);
		if (refIndex === -1) if (addAtStart) this._data.unshift(usageComponent);
		else this._data.push(usageComponent);
		else this._data[refIndex] = usageComponent;
		return usageComponent;
	}
	getComponents = () => {
		return this._data;
	};
	filter(...componentNames) {
		return this._data.filter((it) => componentNames.some((componentName) => it.$ref.startsWith(`#/components/${componentName}`)));
	}
	get = ($ref) => {
		const localFound = this._data.find((c) => c.$ref === $ref) || null;
		if (localFound != null) return localFound;
		const { resolvedSwaggerSchema } = this.config;
		if (resolvedSwaggerSchema.isLocalRef($ref)) return null;
		const foundByRef = resolvedSwaggerSchema.getRef($ref);
		const refDetails = resolvedSwaggerSchema.getRefDetails($ref);
		if (foundByRef != null) {
			const componentDraft = this.createComponentDraft($ref, foundByRef);
			componentDraft.typeName = this.config.hooks.onFormatExternalTypeName?.(componentDraft.typeName, refDetails) || componentDraft.typeName;
			if (this._data.some((component) => component.typeName === componentDraft.typeName)) componentDraft.typeName = this.config.hooks.onFixDuplicateExternalTypeName?.(componentDraft.typeName, refDetails, this._data.map((it) => it.typeName)) ?? `${pascalCase(refDetails.externalOpenapiFileName || "External")}${componentDraft.typeName}`;
			return this.createComponent($ref, componentDraft);
		}
		return null;
	};
	enumsFirst() {
		this._data.sort((a, b) => {
			if (Object.keys(a.rawTypeData || {}).includes("enum")) return -1;
			if (Object.keys(b.rawTypeData || {}).includes("enum")) return 1;
			return 0;
		});
	}
	discriminatorsFirst() {
		this._data.sort((a, b) => {
			if (Object.keys(a.rawTypeData || {}).includes("discriminator")) return -1;
			if (Object.keys(b.rawTypeData || {}).includes("discriminator")) return 1;
			return 0;
		});
	}
};
//#endregion
//#region src/schema-parser/schema-formatters.ts
var SchemaFormatters = class {
	config;
	templatesWorker;
	schemaUtils;
	constructor(schemaParser) {
		this.config = schemaParser.config;
		this.schemaUtils = schemaParser.schemaUtils;
		this.templatesWorker = schemaParser.templatesWorker;
	}
	base = {
		[SCHEMA_TYPES$1.ENUM]: (parsedSchema) => {
			if (this.config.generateUnionEnums) return {
				...parsedSchema,
				$content: parsedSchema.content,
				content: this.config.Ts.UnionType(parsedSchema.content.map(({ value }) => value))
			};
			const escapedContent = parsedSchema.content.map((item) => ({
				...item,
				description: item.description ? this.escapeJSDocContent(item.description) : ""
			}));
			return {
				...parsedSchema,
				$content: parsedSchema.content,
				content: this.config.Ts.EnumFieldsWrapper(escapedContent)
			};
		},
		[SCHEMA_TYPES$1.OBJECT]: (parsedSchema) => {
			if (parsedSchema.nullable) return this.inline[SCHEMA_TYPES$1.OBJECT](parsedSchema);
			return {
				...parsedSchema,
				$content: parsedSchema.content,
				content: this.formatObjectContent(parsedSchema.content)
			};
		},
		[SCHEMA_TYPES$1.PRIMITIVE]: (parsedSchema) => {
			return {
				...parsedSchema,
				$content: parsedSchema.content
			};
		}
	};
	inline = {
		[SCHEMA_TYPES$1.ENUM]: (parsedSchema) => {
			return {
				...parsedSchema,
				content: parsedSchema.$ref ? parsedSchema.typeName : this.config.Ts.UnionType((0, es_toolkit.compact)([...parsedSchema.content.map(({ value }) => `${value}`), parsedSchema.nullable && this.config.Ts.Keyword.Null])) || this.config.Ts.Keyword.Any
			};
		},
		[SCHEMA_TYPES$1.OBJECT]: (parsedSchema) => {
			if (typeof parsedSchema.content === "string") return {
				...parsedSchema,
				typeIdentifier: this.config.Ts.Keyword.Type,
				content: this.schemaUtils.safeAddNullToType(parsedSchema.content)
			};
			return {
				...parsedSchema,
				typeIdentifier: this.config.Ts.Keyword.Type,
				content: this.schemaUtils.safeAddNullToType(parsedSchema, parsedSchema.content.length ? this.config.Ts.ObjectWrapper(this.formatObjectContent(parsedSchema.content)) : this.config.Ts.RecordType(this.config.Ts.Keyword.String, this.config.Ts.Keyword.Any))
			};
		}
	};
	formatSchema = (parsedSchema, formatType = "base") => {
		const schemaType = (0, es_toolkit_compat.get)(parsedSchema, ["schemaType"]) || (0, es_toolkit_compat.get)(parsedSchema, ["$parsed", "schemaType"]);
		return (0, es_toolkit_compat.get)(this, [formatType, schemaType])?.(parsedSchema) || parsedSchema;
	};
	escapeJSDocContent = (content) => {
		if (content === void 0) return "";
		return (typeof content === "string" ? content : String(content)).replace(/\*\//g, "*\\/");
	};
	formatDescription = (description, inline) => {
		if (!description) return "";
		const escapedDescription = this.escapeJSDocContent(description);
		if (!escapedDescription.includes("\n")) return escapedDescription;
		if (inline) return (0, es_toolkit.compact)(escapedDescription.split(/\n/g).map((part) => part.trim())).join(" ");
		return escapedDescription.replace(/\n$/g, "");
	};
	formatObjectContent = (content) => {
		const fields = [];
		for (const part of content) {
			const extraSpace = "  ";
			const result = `${extraSpace}${part.field},\n`;
			const routeNameFromTemplate = this.templatesWorker.renderTemplate(this.config.templatesToRender.dataContractJsDoc, { data: part }).split("\n").map((c) => `${extraSpace}${c}`).join("\n");
			if (routeNameFromTemplate) fields.push(`${routeNameFromTemplate}${result}`);
			else fields.push(`${result}`);
		}
		return fields.join("");
	};
};
//#endregion
//#region src/util/sort-by-property.ts
const sortByProperty = (propertyName) => (o1, o2) => {
	if (o1[propertyName] > o2[propertyName]) return 1;
	if (o1[propertyName] < o2[propertyName]) return -1;
	return 0;
};
//#endregion
//#region src/schema-parser/mono-schema-parser.ts
var MonoSchemaParser = class {
	schema;
	typeName;
	schemaPath;
	schemaParser;
	schemaParserFabric;
	typeNameFormatter;
	schemaComponentsMap;
	schemaUtils;
	config;
	schemaFormatters;
	constructor(schemaParser, schema, typeName = null, schemaPath = []) {
		this.schemaParser = schemaParser;
		this.schemaParserFabric = schemaParser.schemaParserFabric;
		this.schema = schema;
		this.typeName = typeName;
		this.typeNameFormatter = schemaParser.typeNameFormatter;
		this.schemaPath = schemaPath;
		this.schemaComponentsMap = this.schemaParser.schemaComponentsMap;
		this.schemaUtils = this.schemaParser.schemaUtils;
		this.config = this.schemaParser.config;
		this.schemaFormatters = this.schemaParser.schemaFormatters;
	}
	parse() {
		throw new Error("not implemented");
	}
	buildTypeNameFromPath = () => {
		return this.schemaUtils.buildTypeNameFromPath(this.schemaPath);
	};
};
//#endregion
//#region src/schema-parser/base-schema-parsers/array.ts
var ArraySchemaParser = class extends MonoSchemaParser {
	parse() {
		let contentType;
		const { type, description, items } = this.schema || {};
		if (Array.isArray(items) && type === SCHEMA_TYPES$1.ARRAY) {
			const tupleContent = [];
			for (const item of items) tupleContent.push(this.schemaParserFabric.createSchemaParser({
				schema: item,
				schemaPath: this.schemaPath
			}).getInlineParseContent());
			contentType = this.config.Ts.Tuple(tupleContent);
		} else {
			const content = this.schemaParserFabric.createSchemaParser({
				schema: items,
				schemaPath: this.schemaPath
			}).getInlineParseContent();
			contentType = this.config.Ts.ArrayType(content);
		}
		return {
			...typeof this.schema === "object" ? this.schema : {},
			$schemaPath: this.schemaPath.slice(),
			$parsedSchema: true,
			schemaType: SCHEMA_TYPES$1.PRIMITIVE,
			type: SCHEMA_TYPES$1.PRIMITIVE,
			typeIdentifier: this.config.Ts.Keyword.Type,
			name: this.typeName,
			description: this.schemaFormatters.formatDescription(description),
			content: this.schemaUtils.safeAddNullToType(this.schema, contentType)
		};
	}
};
//#endregion
//#region src/schema-parser/base-schema-parsers/complex.ts
var ComplexSchemaParser = class extends MonoSchemaParser {
	parse() {
		const complexType = this.schemaUtils.getComplexType(this.schema);
		const simpleSchema = (0, es_toolkit.omit)(this.schema, Object.keys(this.schemaParser._complexSchemaParsers));
		const complexSchemaContent = this.schemaParser._complexSchemaParsers[complexType](this.schema);
		return {
			...typeof this.schema === "object" ? this.schema : {},
			$schemaPath: this.schemaPath.slice(),
			$parsedSchema: true,
			schemaType: SCHEMA_TYPES$1.COMPLEX,
			type: SCHEMA_TYPES$1.PRIMITIVE,
			typeIdentifier: this.config.Ts.Keyword.Type,
			name: this.typeName,
			description: this.schemaFormatters.formatDescription(this.schema.description || (0, es_toolkit.compact)((this.schema[complexType] || []).map((item) => item?.description))[0] || ""),
			content: this.config.Ts.IntersectionType((0, es_toolkit.compact)([this.config.Ts.ExpressionGroup(complexSchemaContent), this.schemaUtils.getInternalSchemaType(simpleSchema) === SCHEMA_TYPES$1.OBJECT && this.config.Ts.ExpressionGroup(this.schemaParserFabric.createSchemaParser({
				schema: simpleSchema,
				schemaPath: this.schemaPath
			}).getInlineParseContent())])) || this.config.Ts.Keyword.Any
		};
	}
};
//#endregion
//#region src/schema-parser/base-schema-parsers/discriminator.ts
var DiscriminatorSchemaParser = class extends MonoSchemaParser {
	parse() {
		const ts = this.config.Ts;
		const { discriminator, ...noDiscriminatorSchema } = this.schema;
		if (!discriminator.mapping) return this.schemaParserFabric.createSchemaParser({
			schema: noDiscriminatorSchema,
			typeName: this.typeName,
			schemaPath: this.schemaPath
		}).parseSchema();
		const skipMappingType = false;
		const abstractSchemaStruct = this.createAbstractSchemaStruct();
		const discriminatorSchemaStruct = this.createDiscriminatorSchema({
			skipMappingType,
			abstractSchemaStruct
		});
		const schemaContent = ts.IntersectionType((0, es_toolkit.compact)([abstractSchemaStruct?.content, discriminatorSchemaStruct?.content]));
		return {
			...typeof this.schema === "object" ? this.schema : {},
			$schemaPath: this.schemaPath.slice(),
			$parsedSchema: true,
			schemaType: SCHEMA_TYPES$1.COMPLEX,
			type: SCHEMA_TYPES$1.PRIMITIVE,
			typeIdentifier: ts.Keyword.Type,
			name: this.typeName,
			description: this.schemaFormatters.formatDescription(this.schema.description),
			content: schemaContent
		};
	}
	createDiscriminatorSchema = ({ skipMappingType, abstractSchemaStruct }) => {
		const ts = this.config.Ts;
		const refPath = this.schemaComponentsMap.createRef([
			"components",
			"schemas",
			this.typeName
		]);
		const { discriminator } = this.schema;
		const mappingEntries = Object.entries(discriminator.mapping || {});
		const ableToCreateMappingType = !skipMappingType && !!(abstractSchemaStruct?.typeName && mappingEntries.length);
		const mappingContents = [];
		let mappingTypeName;
		/** { mapping_key: SchemaEnum.MappingKey, ... } */
		const mappingPropertySchemaEnumKeysMap = this.createMappingPropertySchemaEnumKeys({
			abstractSchemaStruct,
			discPropertyName: discriminator.propertyName
		});
		if (ableToCreateMappingType) {
			const rawTypeName = `${abstractSchemaStruct.typeName}_${discriminator.propertyName}`;
			const generatedTypeName = this.schemaUtils.resolveTypeName(rawTypeName, {
				suffixes: this.config.extractingOptions.discriminatorMappingSuffix,
				resolver: this.config.extractingOptions.discriminatorMappingNameResolver
			});
			const content = ts.IntersectionType([ts.ObjectWrapper(ts.TypeField({
				key: ts.StringValue(discriminator.propertyName),
				value: "Key"
			})), "Type"]);
			const component = this.schemaParserFabric.createParsedComponent({
				typeName: generatedTypeName,
				schema: {
					type: "object",
					properties: {},
					genericArgs: [{ name: "Key" }, { name: "Type" }],
					internal: true
				}
			});
			component.typeData.content = content;
			mappingTypeName = this.typeNameFormatter.format(component.typeName);
		}
		/** returns (GenericType<"mapping_key", MappingType>) or ({ discriminatorProperty: "mapping_key" } & MappingType) */
		const createMappingContent = (mappingSchema, mappingKey) => {
			const content = this.schemaParserFabric.createSchemaParser({
				schema: mappingSchema,
				schemaPath: this.schemaPath
			}).getInlineParseContent();
			const mappingUsageKey = mappingPropertySchemaEnumKeysMap[mappingKey] || ts.StringValue(mappingKey);
			if (ableToCreateMappingType) return ts.TypeWithGeneric(mappingTypeName, [mappingUsageKey, content]);
			return ts.ExpressionGroup(ts.IntersectionType([ts.ObjectWrapper(ts.TypeField({
				key: ts.StringValue(discriminator.propertyName),
				value: mappingUsageKey
			})), content]));
		};
		for (const [mappingKey, schema] of mappingEntries) {
			const mappingSchema = typeof schema === "string" ? { $ref: schema } : schema;
			this.mutateMappingDependentSchema({
				discPropertyName: discriminator.propertyName,
				abstractSchemaStruct,
				mappingSchema,
				refPath,
				mappingPropertySchemaEnumKeysMap
			});
			mappingContents.push(createMappingContent(mappingSchema, mappingKey));
		}
		if (skipMappingType) return null;
		return { content: ts.ExpressionGroup(ts.UnionType(mappingContents)) };
	};
	createMappingPropertySchemaEnumKeys = ({ abstractSchemaStruct, discPropertyName }) => {
		const ts = this.config.Ts;
		const mappingPropertySchemaEnumKeysMap = {};
		let mappingPropertySchema = (0, es_toolkit_compat.get)(abstractSchemaStruct?.component?.rawTypeData, ["properties", discPropertyName]);
		if (this.schemaUtils.isRefSchema(mappingPropertySchema)) mappingPropertySchema = this.schemaUtils.getSchemaRefType(mappingPropertySchema);
		const parsedEnum = mappingPropertySchema?.rawTypeData?.$parsed;
		if (parsedEnum?.type === SCHEMA_TYPES$1.ENUM) {
			const enumEntries = (parsedEnum.enum || []).map((key, index) => [key, index]);
			for (const [key, index] of enumEntries) {
				const enumContent = parsedEnum.content?.[index];
				if (this.config.generateUnionEnums) {
					const literalValue = enumContent?.value ?? (key !== void 0 ? ts.StringValue(key) : void 0);
					if (literalValue !== void 0) mappingPropertySchemaEnumKeysMap[key] = literalValue;
				} else if (parsedEnum.typeName && enumContent?.key) mappingPropertySchemaEnumKeysMap[key] = ts.EnumUsageKey(parsedEnum.typeName, enumContent.key);
			}
		}
		return mappingPropertySchemaEnumKeysMap;
	};
	mutateMappingDependentSchema = ({ discPropertyName, abstractSchemaStruct, mappingSchema, refPath, mappingPropertySchemaEnumKeysMap }) => {
		const complexSchemaKeys = Object.keys(this.schemaParser._complexSchemaParsers);
		if (mappingSchema.$ref && abstractSchemaStruct?.component?.$ref) {
			const mappingRefSchema = this.schemaUtils.getSchemaRefType(mappingSchema)?.rawTypeData;
			if (mappingRefSchema) {
				for (const schemaKey of complexSchemaKeys) if (Array.isArray(mappingRefSchema[schemaKey])) mappingRefSchema[schemaKey] = mappingRefSchema[schemaKey].map((schema) => {
					if (schema.$ref === refPath) return {
						...schema,
						$ref: abstractSchemaStruct.component.$ref
					};
					if (this.schemaUtils.getInternalSchemaType(schema) === SCHEMA_TYPES$1.OBJECT) for (const schemaPropertyName in schema.properties) {
						const schemaProperty = schema.properties[schemaPropertyName];
						if (schemaPropertyName === discPropertyName && this.schemaUtils.getInternalSchemaType(schemaProperty) === SCHEMA_TYPES$1.ENUM && schemaProperty.enum.length === 1 && mappingPropertySchemaEnumKeysMap[schemaProperty.enum[0]]) schema.properties[schemaPropertyName] = this.schemaParserFabric.createSchema({ content: mappingPropertySchemaEnumKeysMap[schemaProperty.enum[0]] });
					}
					return schema;
				});
			}
		}
	};
	createAbstractSchemaStruct = () => {
		const { discriminator, ...noDiscriminatorSchema } = this.schema;
		const complexSchemaKeys = Object.keys(this.schemaParser._complexSchemaParsers);
		const schema = (0, es_toolkit.omit)(structuredClone(noDiscriminatorSchema), complexSchemaKeys);
		const schemaIsAny = this.schemaParserFabric.getInlineParseContent(structuredClone(schema)) === this.config.Ts.Keyword.Any;
		if (!Object.keys(schema).length || schemaIsAny) return null;
		const typeName = this.schemaUtils.resolveTypeName(this.typeName, {
			prefixes: this.config.extractingOptions.discriminatorAbstractPrefix,
			resolver: this.config.extractingOptions.discriminatorAbstractResolver
		});
		const component = this.schemaComponentsMap.createComponent(this.schemaComponentsMap.createRef([
			"components",
			"schemas",
			typeName
		]), {
			...schema,
			internal: true
		});
		return {
			typeName,
			component,
			content: this.schemaParserFabric.createSchemaParser({
				schema: component,
				schemaPath: this.schemaPath
			}).getInlineParseContent()
		};
	};
	createComplexSchemaStruct = () => {
		const ts = this.config.Ts;
		const complexType = this.schemaUtils.getComplexType(this.schema);
		if (complexType === SCHEMA_TYPES$1.COMPLEX_UNKNOWN) return null;
		return { content: ts.ExpressionGroup(this.schemaParser._complexSchemaParsers[complexType](this.schema)) };
	};
};
//#endregion
//#region src/schema-parser/util/enum-key-resolver.ts
var EnumKeyResolver = class extends NameResolver {
	counter = 1;
	constructor(config, reservedNames) {
		super(config, reservedNames, (variants) => {
			const generatedVariant = variants[0] && `${variants[0]}${this.counter++}` || `${this.config.enumKeyResolverName}${this.counter++}`;
			consola.consola.debug("generated fallback type name for enum key - ", generatedVariant);
			return generatedVariant;
		});
	}
};
//#endregion
//#region src/schema-parser/base-schema-parsers/enum.ts
var EnumSchemaParser = class extends MonoSchemaParser {
	enumKeyResolver;
	constructor(...args) {
		super(...args);
		this.enumKeyResolver = new EnumKeyResolver(this.config, []);
	}
	extractEnum = (pathTypeName) => {
		const generatedTypeName = this.schemaUtils.resolveTypeName(pathTypeName, {
			suffixes: this.config.extractingOptions.enumSuffix,
			resolver: this.config.extractingOptions.enumNameResolver
		});
		const customComponent = this.schemaComponentsMap.createComponent(this.schemaComponentsMap.createRef([
			"components",
			"schemas",
			generatedTypeName
		]), { ...this.schema });
		return this.schemaParserFabric.parseSchema(customComponent);
	};
	parse() {
		const pathTypeName = this.buildTypeNameFromPath();
		if (this.config.extractEnums && !this.typeName && pathTypeName != null) return this.extractEnum(pathTypeName);
		const refType = this.schemaUtils.getSchemaRefType(this.schema);
		const $ref = refType?.$ref || null;
		if (Array.isArray(this.schema.enum)) this.schema.enum = this.schema.enum.filter((key) => key != null);
		if (Array.isArray(this.schema.enum) && Array.isArray(this.schema.enum[0])) return this.schemaParserFabric.parseSchema({ oneOf: this.schema.enum.map((enumNames) => ({
			type: "array",
			items: enumNames.map((enumName) => ({
				type: "string",
				enum: [enumName]
			}))
		})) }, this.typeName, this.schemaPath);
		const keyType = this.schemaUtils.getSchemaType(this.schema);
		const enumNames = this.schemaUtils.getEnumNames(this.schema);
		const enumDescriptions = this.schemaUtils.getEnumDescriptions(this.schema);
		let content = null;
		const formatValue = (value) => {
			if (value === null) return this.config.Ts.NullValue(value);
			if (keyType.includes(this.schemaUtils.getSchemaType({ type: "number" }))) {
				const maybeNumber = typeof value === "number" ? value : Number(value);
				if (!Number.isNaN(maybeNumber)) return this.config.Ts.NumberValue(maybeNumber);
			}
			if (keyType.includes(this.schemaUtils.getSchemaType({ type: "boolean" }))) {
				if (typeof value === "boolean") return this.config.Ts.BooleanValue(value);
				if (value === "true" || value === "false") return this.config.Ts.BooleanValue(value === "true");
			}
			switch (typeof value) {
				case "number": return this.config.Ts.NumberValue(value);
				case "boolean": return this.config.Ts.BooleanValue(value);
				default: return this.config.Ts.StringValue(value);
			}
		};
		if (Array.isArray(enumNames) && enumNames.length > 0) content = enumNames.map((enumName, index) => {
			const enumValue = (0, es_toolkit_compat.get)(this.schema.enum, index);
			const formattedKey = this.formatEnumKey({
				key: enumName,
				value: enumValue
			});
			if (this.config.enumNamesAsValues || enumValue === void 0) return {
				key: formattedKey,
				type: this.config.Ts.Keyword.String,
				value: this.config.Ts.StringValue(enumName),
				description: enumDescriptions?.[index]
			};
			return {
				key: formattedKey,
				type: keyType,
				value: formatValue(enumValue),
				description: enumDescriptions?.[index]
			};
		});
		else content = this.schema.enum.map((value, index) => {
			return {
				key: this.formatEnumKey({ value }),
				type: keyType,
				value: formatValue(value),
				description: enumDescriptions?.[index]
			};
		});
		return {
			...typeof this.schema === "object" ? this.schema : {},
			$ref,
			typeName: this.typeName || $ref && refType.typeName || null,
			$parsedSchema: true,
			schemaType: SCHEMA_TYPES$1.ENUM,
			type: SCHEMA_TYPES$1.ENUM,
			keyType,
			typeIdentifier: this.config.generateUnionEnums ? this.config.Ts.Keyword.Type : this.config.Ts.Keyword.Enum,
			name: this.typeName,
			description: this.schemaFormatters.formatDescription(this.schema.description),
			content
		};
	}
	formatEnumKey = ({ key, value }) => {
		let formatted;
		if (key) formatted = this.typeNameFormatter.format(key, { type: "enum-key" });
		if (!formatted) formatted = this.typeNameFormatter.format(`${value}`, { type: "enum-key" });
		return this.enumKeyResolver.resolve([formatted]);
	};
};
//#endregion
//#region src/schema-parser/base-schema-parsers/object.ts
var ObjectSchemaParser = class extends MonoSchemaParser {
	parse() {
		const contentProperties = this.getObjectSchemaContent(this.schema);
		return {
			...typeof this.schema === "object" ? this.schema : {},
			$schemaPath: this.schemaPath.slice(),
			$parsedSchema: true,
			schemaType: SCHEMA_TYPES$1.OBJECT,
			type: SCHEMA_TYPES$1.OBJECT,
			typeIdentifier: this.config.Ts.Keyword.Interface,
			name: this.typeName,
			description: this.schemaFormatters.formatDescription(this.schema.description),
			allFieldsAreOptional: !contentProperties.some((part) => part.isRequired),
			content: contentProperties
		};
	}
	getObjectSchemaContent = (schema) => {
		const { properties, additionalProperties } = schema || {};
		const propertiesContent = [];
		for (const [name, property] of Object.entries(properties || {})) {
			const required = this.schemaUtils.isPropertyRequired(name, property, schema);
			const rawTypeData = (0, es_toolkit_compat.get)(this.schemaUtils.getSchemaRefType(property), "rawTypeData", {});
			const nullable = !!(rawTypeData.nullable || property.nullable);
			const fieldName = this.typeNameFormatter.isValidName(name) ? name : this.config.Ts.StringValue(name);
			const rawFieldValue = this.schemaParserFabric.createSchemaParser({
				schema: property,
				schemaPath: [...this.schemaPath, name]
			}).getInlineParseContent();
			const fieldValue = nullable ? this.schemaUtils.safeAddNullToType(property, rawFieldValue) : rawFieldValue;
			const readOnly = property.readOnly;
			const complexType = this.schemaUtils.getComplexType(property);
			const rawDataComplexType = this.schemaUtils.getComplexType(rawTypeData);
			propertiesContent.push({
				...property,
				$$raw: property,
				title: property.title,
				description: property.description || (0, es_toolkit.compact)((property[complexType] || []).map((item) => item?.description))[0] || rawTypeData.description || (0, es_toolkit.compact)((rawTypeData[rawDataComplexType] || []).map((item) => item?.description))[0] || "",
				isRequired: required,
				isNullable: nullable,
				name: fieldName,
				value: fieldValue,
				field: this.config.Ts.TypeField({
					readonly: readOnly && this.config.addReadonly,
					optional: !required,
					key: fieldName,
					value: fieldValue
				})
			});
		}
		if (additionalProperties) {
			const propertyNamesSchema = this.schemaUtils.getSchemaPropertyNamesSchema(schema);
			let interfaceKeysContent;
			if (propertyNamesSchema) interfaceKeysContent = this.schemaParserFabric.createSchemaParser({
				schema: propertyNamesSchema,
				schemaPath: this.schemaPath
			}).getInlineParseContent();
			else interfaceKeysContent = this.config.Ts.Keyword.String;
			propertiesContent.push({
				$$raw: { additionalProperties },
				description: "",
				isRequired: false,
				field: this.config.Ts.InterfaceDynamicField(interfaceKeysContent, this.config.Ts.Keyword.Any)
			});
		}
		return propertiesContent;
	};
};
//#endregion
//#region src/schema-parser/base-schema-parsers/primitive.ts
var PrimitiveSchemaParser = class extends MonoSchemaParser {
	parse() {
		let contentType = null;
		const { additionalProperties, type, description, items } = this.schema || {};
		if (type === this.config.Ts.Keyword.Object && additionalProperties) {
			const propertyNamesSchema = this.schemaUtils.getSchemaPropertyNamesSchema(this.schema);
			let recordKeysContent;
			let recordValuesContent;
			if (propertyNamesSchema) recordKeysContent = this.schemaParserFabric.createSchemaParser({
				schema: propertyNamesSchema,
				schemaPath: this.schemaPath
			}).getInlineParseContent();
			else recordKeysContent = this.config.Ts.Keyword.String;
			if (typeof additionalProperties === "object") recordValuesContent = this.schemaParserFabric.createSchemaParser({
				schema: additionalProperties,
				schemaPath: this.schemaPath
			}).getInlineParseContent();
			else recordValuesContent = this.config.Ts.Keyword.Any;
			const recordType = this.config.Ts.RecordType(recordKeysContent, recordValuesContent);
			contentType = propertyNamesSchema ? this.config.Ts.TypeWithGeneric("Partial", [recordType]) : recordType;
		}
		if (Array.isArray(type) && type.length) contentType = this.schemaParser._complexSchemaParsers.oneOf({
			...typeof this.schema === "object" ? this.schema : {},
			oneOf: type.map((t) => {
				const branch = { type: t };
				Object.assign(branch, this.schema);
				branch.type = t;
				return branch;
			})
		});
		if (Array.isArray(items) && type === SCHEMA_TYPES$1.ARRAY) contentType = this.config.Ts.Tuple(items.map((item) => this.schemaParserFabric.createSchemaParser({
			schema: item,
			schemaPath: this.schemaPath
		}).getInlineParseContent()));
		return {
			...typeof this.schema === "object" ? this.schema : {},
			$schemaPath: this.schemaPath.slice(),
			$parsedSchema: true,
			schemaType: SCHEMA_TYPES$1.PRIMITIVE,
			type: SCHEMA_TYPES$1.PRIMITIVE,
			typeIdentifier: this.config.Ts.Keyword.Type,
			name: this.typeName,
			description: this.schemaFormatters.formatDescription(description),
			content: type === this.config.Ts.Keyword.Null ? type : contentType || this.schemaUtils.getSchemaType(this.schema)
		};
	}
};
//#endregion
//#region src/schema-parser/complex-schema-parsers/all-of.ts
var AllOfSchemaParser = class extends MonoSchemaParser {
	parse() {
		const ignoreTypes = [this.config.Ts.Keyword.Any];
		const combined = this.schema.allOf.map((childSchema) => this.schemaParserFabric.getInlineParseContent(this.schemaUtils.makeAddRequiredToChildSchema(this.schema, childSchema), null, this.schemaPath));
		const filtered = this.schemaUtils.filterSchemaContents(combined, (content) => !ignoreTypes.includes(content));
		const type = this.config.Ts.IntersectionType(filtered);
		return this.schemaUtils.safeAddNullToType(this.schema, type);
	}
};
//#endregion
//#region src/schema-parser/complex-schema-parsers/any-of.ts
var AnyOfSchemaParser = class extends MonoSchemaParser {
	parse() {
		const ignoreTypes = [this.config.Ts.Keyword.Any];
		const combined = this.schema.anyOf.map((childSchema) => this.schemaParserFabric.getInlineParseContent(this.schemaUtils.makeAddRequiredToChildSchema(this.schema, childSchema), null, this.schemaPath));
		const filtered = this.schemaUtils.filterSchemaContents(combined, (content) => !ignoreTypes.includes(content));
		const type = this.config.Ts.UnionType(filtered);
		return this.schemaUtils.safeAddNullToType(this.schema, type);
	}
};
//#endregion
//#region src/schema-parser/complex-schema-parsers/not.ts
var NotSchemaParser = class extends MonoSchemaParser {
	parse() {
		return this.config.Ts.Keyword.Any;
	}
};
//#endregion
//#region src/schema-parser/complex-schema-parsers/one-of.ts
var OneOfSchemaParser = class extends MonoSchemaParser {
	parse() {
		const ignoreTypes = [this.config.Ts.Keyword.Any];
		const combined = this.schema.oneOf.map((childSchema) => this.schemaParserFabric.getInlineParseContent(this.schemaUtils.makeAddRequiredToChildSchema(this.schema, childSchema), null, this.schemaPath));
		const filtered = this.schemaUtils.filterSchemaContents(combined, (content) => !ignoreTypes.includes(content));
		const type = this.config.Ts.UnionType(filtered);
		return this.schemaUtils.safeAddNullToType(this.schema, type);
	}
};
//#endregion
//#region src/schema-parser/schema-parser.ts
var SchemaParser = class {
	schemaParserFabric;
	config;
	schemaComponentsMap;
	typeNameFormatter;
	schemaFormatters;
	schemaUtils;
	templatesWorker;
	typeName;
	schema;
	schemaPath;
	constructor(schemaParserFabric, { typeName, schema, schemaPath } = {}) {
		this.schemaParserFabric = schemaParserFabric;
		this.config = schemaParserFabric.config;
		this.templatesWorker = schemaParserFabric.templatesWorker;
		this.schemaComponentsMap = schemaParserFabric.schemaComponentsMap;
		this.typeNameFormatter = schemaParserFabric.typeNameFormatter;
		this.schemaFormatters = schemaParserFabric.schemaFormatters;
		this.schemaUtils = schemaParserFabric.schemaUtils;
		this.typeName = typeName || null;
		this.schema = schema;
		this.schemaPath = [...schemaPath || []];
	}
	_complexSchemaParsers = {
		[SCHEMA_TYPES$1.COMPLEX_ONE_OF]: (schema) => {
			return new (this.config.schemaParsers.complexOneOf || OneOfSchemaParser)(this, schema, null, this.schemaPath).parse();
		},
		[SCHEMA_TYPES$1.COMPLEX_ALL_OF]: (schema) => {
			return new (this.config.schemaParsers.complexAllOf || AllOfSchemaParser)(this, schema, null, this.schemaPath).parse();
		},
		[SCHEMA_TYPES$1.COMPLEX_ANY_OF]: (schema) => {
			return new (this.config.schemaParsers.complexAnyOf || AnyOfSchemaParser)(this, schema, null, this.schemaPath).parse();
		},
		[SCHEMA_TYPES$1.COMPLEX_NOT]: (schema) => {
			return new (this.config.schemaParsers.complexNot || NotSchemaParser)(this, schema, null, this.schemaPath).parse();
		}
	};
	_baseSchemaParsers = {
		[SCHEMA_TYPES$1.ENUM]: (schema, typeName) => {
			return new (this.config.schemaParsers.enum || EnumSchemaParser)(this, schema, typeName, this.schemaPath).parse();
		},
		[SCHEMA_TYPES$1.OBJECT]: (schema, typeName) => {
			return new (this.config.schemaParsers.object || ObjectSchemaParser)(this, schema, typeName, this.schemaPath).parse();
		},
		[SCHEMA_TYPES$1.COMPLEX]: (schema, typeName) => {
			return new (this.config.schemaParsers.complex || ComplexSchemaParser)(this, schema, typeName, this.schemaPath).parse();
		},
		[SCHEMA_TYPES$1.PRIMITIVE]: (schema, typeName) => {
			return new (this.config.schemaParsers.primitive || PrimitiveSchemaParser)(this, schema, typeName, this.schemaPath).parse();
		},
		[SCHEMA_TYPES$1.DISCRIMINATOR]: (schema, typeName) => {
			return new (this.config.schemaParsers.discriminator || DiscriminatorSchemaParser)(this, schema, typeName, this.schemaPath).parse();
		},
		[SCHEMA_TYPES$1.ARRAY]: (schema, typeName) => {
			return new (this.config.schemaParsers.array || ArraySchemaParser)(this, schema, typeName, this.schemaPath).parse();
		}
	};
	parseSchema = () => {
		if (!this.schema) return this._baseSchemaParsers[SCHEMA_TYPES$1.PRIMITIVE](null, this.typeName);
		let schemaType = null;
		let parsedSchema = null;
		if (typeof this.schema === "string") return this.schema;
		if (!this.schema.$parsed) {
			if (!this.typeName && this.schemaUtils.isRefSchema(this.schema)) this.typeName = this.schemaUtils.getSchemaType(this.schema);
			if (this.schema.items && !Array.isArray(this.schema.items) && !this.schema.type) this.schema.type = SCHEMA_TYPES$1.ARRAY;
			if (Array.isArray(this.schema.enum) && this.schema.enum.length === 1 && this.schema.enum[0] == null) {
				consola.consola.debug("invalid enum schema", this.schema);
				this.schema = { type: this.config.Ts.Keyword.Null };
			}
			if ("content" in this.schema && typeof this.schema.content === "object") {
				const schema = this.extractSchemaFromResponseStruct(this.schema);
				const schemaParser = this.schemaParserFabric.createSchemaParser({
					schema,
					typeName: this.typeName,
					schemaPath: this.schemaPath
				});
				this.schema.$parsed = schemaParser.parseSchema();
				return this.schema.$parsed;
			}
			schemaType = this.schemaUtils.getInternalSchemaType(this.schema);
			this.schemaPath.push(this.typeName);
			(0, es_toolkit.merge)(this.schema, this.config.hooks.onPreParseSchema(this.schema, this.typeName, schemaType) || {});
			parsedSchema = this._baseSchemaParsers[schemaType](this.schema, this.typeName);
			this.schema.$parsed = this.config.hooks.onParseSchema(this.schema, parsedSchema) || parsedSchema;
			if (this.config.sortTypes && Array.isArray(this.schema.$parsed?.content)) this.schema.$parsed.content = this.schema.$parsed.content.sort(sortByProperty("name"));
		}
		this.schemaPath.pop();
		return this.schema.$parsed;
	};
	getInlineParseContent = () => {
		const parsedSchema = this.parseSchema();
		return this.schemaFormatters.formatSchema(parsedSchema, "inline").content;
	};
	getParseContent = () => {
		const parsedSchema = this.parseSchema();
		return this.schemaFormatters.formatSchema(parsedSchema, "base").content;
	};
	extractSchemaFromResponseStruct = (responseStruct) => {
		const { content, ...extras } = responseStruct;
		const firstResponse = Object.values(content || {})[0];
		const firstSchema = (0, es_toolkit_compat.get)(firstResponse, "schema");
		if (!firstSchema) return;
		const { schema: _, ...restResponse } = firstResponse || {};
		return {
			...extras,
			...restResponse,
			...firstSchema
		};
	};
};
//#endregion
//#region src/schema-parser/schema-utils.ts
var SchemaUtils = class {
	constructor(config, schemaComponentsMap, typeNameFormatter) {
		this.config = config;
		this.schemaComponentsMap = schemaComponentsMap;
		this.typeNameFormatter = typeNameFormatter;
	}
	isBinaryLikeMimeType = (contentMediaType) => {
		if (typeof contentMediaType !== "string" || !contentMediaType) return false;
		const mediaType = contentMediaType.split(";")[0]?.trim().toLowerCase();
		if (!mediaType) return false;
		/**
		* `contentMediaType` comes from JSON Schema. In practice it is often used to
		* signal "this string is a file/blob", but it may also be used for textual
		* payloads (json/xml/etc). We treat only binary-ish media types as `File`.
		*/
		if (mediaType.startsWith("text/")) return false;
		if (mediaType.includes("json") || mediaType.includes("+json")) return false;
		/** application/vnd.* binary types first: names like "openxmlformats" / "spreadsheetml" contain "xml" but are binary. */
		if (mediaType.startsWith("application/vnd.")) return mediaType.endsWith(".blob") || mediaType.includes("spreadsheetml.sheet") || mediaType.startsWith("application/vnd.ms-excel") || mediaType.startsWith("application/vnd.openxmlformats-officedocument.") || mediaType === "application/vnd.rar" || mediaType.startsWith("application/vnd.oasis.opendocument.") || mediaType.startsWith("application/vnd.ms-powerpoint") || mediaType.startsWith("application/vnd.ms-fontobject") || mediaType === "application/vnd.visio" || mediaType === "application/vnd.amazon.ebook";
		if (mediaType.includes("xml") || mediaType.includes("+xml")) return false;
		if (mediaType === "application/x-www-form-urlencoded") return false;
		if (mediaType === "application/javascript" || mediaType === "application/ecmascript" || mediaType === "application/graphql" || mediaType === "application/yaml" || mediaType === "application/x-yaml" || mediaType === "application/jwt") return false;
		if (mediaType.startsWith("application/")) return mediaType === "application/octet-stream" || mediaType.startsWith("application/pdf") || mediaType === "application/zip" || mediaType.startsWith("application/x-zip") || mediaType === "application/gzip" || mediaType.startsWith("application/x-gzip") || mediaType.startsWith("application/x-bzip") || mediaType === "application/x-bzip2" || mediaType.startsWith("application/x-tar") || mediaType.startsWith("application/x-rar") || mediaType.startsWith("application/x-7z") || mediaType === "application/x-binary" || mediaType === "application/java-archive" || mediaType === "application/epub+zip" || mediaType === "application/msword" || mediaType === "application/rtf" || mediaType === "application/x-abiword" || mediaType === "application/x-freearc";
		return mediaType.startsWith("image/") || mediaType.startsWith("audio/") || mediaType.startsWith("video/") || mediaType.startsWith("font/") || mediaType.startsWith("model/");
	};
	getRequiredProperties = (schema) => {
		return (0, es_toolkit.uniq)(schema && Array.isArray(schema.required) && schema.required || []);
	};
	isRefSchema = (schema) => {
		return !!schema?.$ref;
	};
	getEnumNames = (schema) => {
		return schema["x-enumNames"] || schema.xEnumNames || schema["x-enumnames"] || schema["x-enum-varnames"];
	};
	getEnumDescriptions = (schema) => {
		return schema["x-enumDescriptions"] || schema.xEnumDescriptions || schema["x-enumdescriptions"] || schema["x-enum-descriptions"];
	};
	getSchemaPropertyNamesSchema = (schema) => {
		if (!schema) return null;
		return schema.propertyNames || schema["x-propertyNames"] || null;
	};
	getSchemaRefType = (schema) => {
		if (!this.isRefSchema(schema)) return null;
		return this.schemaComponentsMap.get(schema.$ref);
	};
	isPropertyRequired = (name, propertySchema, rootSchema) => {
		if (propertySchema["x-omitempty"] === false) return true;
		const isRequired = typeof propertySchema.required === "boolean" ? !!propertySchema.required : Array.isArray(rootSchema.required) ? rootSchema.required.includes(name) : !!rootSchema.required;
		if (this.config.convertedFromSwagger2) return typeof propertySchema.nullable === this.config.Ts.Keyword.Undefined ? isRequired : !propertySchema.nullable;
		return isRequired;
	};
	isNullMissingInType = (schema, type) => {
		const { nullable, type: schemaType } = schema || {};
		if (!(nullable || !!(0, es_toolkit_compat.get)(schema, "x-nullable") || schemaType === this.config.Ts.Keyword.Null) || typeof type !== "string") return false;
		const nullKeyword = this.config.Ts.Keyword.Null;
		const lastLine = type.trimEnd().split("\n").pop() ?? type;
		return !lastLine.includes(` ${nullKeyword}`) && !lastLine.includes(`${nullKeyword} `);
	};
	safeAddNullToType = (schema, type) => {
		if (this.isNullMissingInType(schema, type)) return this.config.Ts.UnionType([type, this.config.Ts.Keyword.Null]);
		return type;
	};
	getSchemaPrimitiveType = (rawSchema) => {
		const schema = rawSchema || {};
		if (schema.type) return (0, es_toolkit_compat.camelCase)(schema.type);
		if (schema.enum) {
			const enumFieldType = typeof schema.enum[0];
			if (enumFieldType === this.config.Ts.Keyword.Undefined) return;
			return (0, es_toolkit_compat.camelCase)(enumFieldType);
		}
		if (Object.keys(schema.properties || {}).length) return SCHEMA_TYPES$1.OBJECT;
		if (schema.items) return SCHEMA_TYPES$1.ARRAY;
		return null;
	};
	checkAndAddRequiredKeys = (schema, resultType) => {
		if ("$$requiredKeys" in schema && schema.$$requiredKeys.length) {
			this.config.update({ internalTemplateOptions: { addUtilRequiredKeysType: true } });
			return this.config.Ts.TypeWithGeneric(this.config.Ts.CodeGenKeyword.UtilRequiredKeys, [resultType, this.config.Ts.UnionType(schema.$$requiredKeys.map(this.config.Ts.StringValue))]);
		}
		return resultType;
	};
	makeAddRequiredToChildSchema = (parentSchema, childSchema) => {
		if (!childSchema) return childSchema;
		const required = (0, es_toolkit.uniq)([...this.getRequiredProperties(parentSchema), ...this.getRequiredProperties(childSchema)]);
		const refData = this.getSchemaRefType(childSchema);
		if (refData) {
			const existedRequiredKeys = Object.keys(refData.rawTypeData?.properties || {}).filter((key) => required.includes(key));
			if (!existedRequiredKeys.length) return childSchema;
			return {
				...childSchema,
				$$requiredKeys: existedRequiredKeys
			};
		}
		if (childSchema.properties) {
			const existedRequiredKeys = Object.keys(childSchema.properties).filter((key) => required.includes(key));
			if (!existedRequiredKeys.length) return childSchema;
			return {
				required: (0, es_toolkit.uniq)([...this.getRequiredProperties(childSchema), ...existedRequiredKeys]),
				...childSchema
			};
		}
		return childSchema;
	};
	filterSchemaContents = (contents, filterFn) => {
		return (0, es_toolkit.uniq)(contents.filter((type) => filterFn(type)));
	};
	resolveTypeName = (typeName, { suffixes, resolver, prefixes, shouldReserve = true }) => {
		if (resolver) return this.config.componentTypeNameResolver.resolve([], (reserved) => {
			return resolver(pascalCase(typeName), reserved);
		});
		return this.config.componentTypeNameResolver.resolve([...(prefixes || []).map((prefix) => pascalCase(`${prefix} ${typeName}`)), ...(suffixes || []).map((suffix) => pascalCase(`${typeName} ${suffix}`))], shouldReserve);
	};
	getComplexType = (schema) => {
		if (schema.oneOf) return SCHEMA_TYPES$1.COMPLEX_ONE_OF;
		if (schema.allOf) return SCHEMA_TYPES$1.COMPLEX_ALL_OF;
		if (schema.anyOf) return SCHEMA_TYPES$1.COMPLEX_ANY_OF;
		if (schema.not) return SCHEMA_TYPES$1.COMPLEX_NOT;
		return SCHEMA_TYPES$1.COMPLEX_UNKNOWN;
	};
	getInternalSchemaType = (schema) => {
		if (schema.enum && schema.enum.length > 0 || this.getEnumNames(schema) && this.getEnumNames(schema).length > 0) return SCHEMA_TYPES$1.ENUM;
		if (schema.discriminator) return SCHEMA_TYPES$1.DISCRIMINATOR;
		if (schema.allOf || schema.oneOf || schema.anyOf || schema.not) return SCHEMA_TYPES$1.COMPLEX;
		if (schema.properties && Object.keys(schema.properties).length > 0) return SCHEMA_TYPES$1.OBJECT;
		if (schema.type === SCHEMA_TYPES$1.ARRAY) return SCHEMA_TYPES$1.ARRAY;
		return SCHEMA_TYPES$1.PRIMITIVE;
	};
	getSchemaType = (schema) => {
		if (!schema) return this.config.Ts.Keyword.Any;
		const refTypeInfo = this.getSchemaRefType(schema);
		if (refTypeInfo) return this.checkAndAddRequiredKeys(schema, this.safeAddNullToType(schema, this.typeNameFormatter.format(refTypeInfo.typeName)));
		let resultType;
		if (this.isConstantSchema(schema) && !schema.enum) resultType = this.formatJsValue(schema.const);
		else {
			const primitiveType = this.getSchemaPrimitiveType(schema);
			if (primitiveType == null) return this.config.Ts.Keyword.Any;
			if (primitiveType === this.config.Ts.Keyword.String && !schema.format && this.isBinaryLikeMimeType(schema.contentMediaType)) resultType = this.config.Ts.UnionType([this.config.Ts.Keyword.File, this.config.Ts.Keyword.Blob]);
			else {
				const typeAlias = (0, es_toolkit_compat.get)(this.config.primitiveTypes, [primitiveType, schema.format]) || (0, es_toolkit_compat.get)(this.config.primitiveTypes, [primitiveType, "$default"]) || this.config.primitiveTypes[primitiveType];
				if (typeof typeAlias === "function") resultType = typeAlias(schema, this);
				else resultType = typeAlias || primitiveType;
			}
		}
		if (!resultType) return this.config.Ts.Keyword.Any;
		return this.checkAndAddRequiredKeys(schema, this.safeAddNullToType(schema, resultType));
	};
	buildTypeNameFromPath = (schemaPath) => {
		schemaPath = (0, es_toolkit.uniq)((0, es_toolkit.compact)(schemaPath || []));
		if (!schemaPath || !schemaPath[0]) return null;
		return pascalCase((0, es_toolkit.uniq)([schemaPath[0], schemaPath[schemaPath.length - 1]]).join("_"));
	};
	isConstantSchema(schema) {
		return "const" in schema;
	}
	formatJsValue = (value) => {
		switch (typeof value) {
			case "string": return this.config.Ts.StringValue(value);
			case "boolean": return this.config.Ts.BooleanValue(value);
			case "number": return this.config.Ts.NumberValue(value);
			default:
				if (value === null) return this.config.Ts.NullValue(value);
				return this.config.Ts.Keyword.Any;
		}
	};
};
//#endregion
//#region src/schema-parser/schema-parser-fabric.ts
var SchemaParserFabric = class {
	config;
	schemaComponentsMap;
	typeNameFormatter;
	schemaFormatters;
	templatesWorker;
	schemaUtils;
	constructor(config, templatesWorker, schemaComponentsMap, typeNameFormatter) {
		this.config = config;
		this.schemaComponentsMap = schemaComponentsMap;
		this.typeNameFormatter = typeNameFormatter;
		this.templatesWorker = templatesWorker;
		this.schemaUtils = new SchemaUtils(this.config, this.schemaComponentsMap, this.typeNameFormatter);
		this.schemaFormatters = new SchemaFormatters(this);
	}
	createSchemaParser = ({ schema, typeName, schemaPath }) => {
		return new SchemaParser(this, {
			schema,
			typeName,
			schemaPath
		});
	};
	createSchema = ({ content, linkedSchema = {}, linkedComponent, schemaPath, ...otherSchemaProps }) => {
		const parser = this.createSchemaParser({
			schema: linkedComponent || linkedSchema,
			schemaPath
		});
		const parsed = parser.parseSchema();
		parsed.content = content;
		Object.assign(parsed, otherSchemaProps);
		if (linkedComponent) linkedComponent.typeData = parsed;
		return parser.schema;
	};
	createParsedComponent = ({ typeName, schema, schemaPath }) => {
		const schemaCopy = structuredClone(schema);
		const customComponent = this.schemaComponentsMap.createComponent(this.schemaComponentsMap.createRef([
			"components",
			"schemas",
			typeName
		]), schemaCopy);
		const parsed = this.parseSchema(schemaCopy, null, schemaPath);
		parsed.name = typeName;
		customComponent.typeData = parsed;
		return customComponent;
	};
	parseSchema = (schema, typeName = null, schemaPath = []) => {
		return this.createSchemaParser({
			schema,
			typeName,
			schemaPath
		}).parseSchema();
	};
	getInlineParseContent = (schema, typeName, schemaPath) => {
		return this.createSchemaParser({
			schema,
			typeName,
			schemaPath
		}).getInlineParseContent();
	};
	getParseContent = (schema, typeName, schemaPath) => {
		return this.createSchemaParser({
			schema,
			typeName,
			schemaPath
		}).getParseContent();
	};
};
//#endregion
//#region src/util/id.ts
const generateId = () => node_crypto.randomUUID();
//#endregion
//#region src/schema-routes/util/specific-arg-name-resolver.ts
var SpecificArgNameResolver = class extends NameResolver {
	counter = 1;
	constructor(config, reservedNames) {
		super(config, reservedNames, (variants) => {
			const generatedVariant = variants[0] && `${variants[0]}${this.counter++}` || `${this.config.specificArgNameResolverName}${this.counter++}`;
			consola.consola.debug("generated fallback type name for specific arg - ", generatedVariant);
			return generatedVariant;
		});
	}
};
//#endregion
//#region src/schema-routes/schema-routes.ts
const CONTENT_KIND = {
	JSON: "JSON",
	JSON_API: "JSON_API",
	URL_ENCODED: "URL_ENCODED",
	FORM_DATA: "FORM_DATA",
	IMAGE: "IMAGE",
	OTHER: "OTHER",
	TEXT: "TEXT"
};
/**
* When a colliding extract name is repeatedly resolved, cap iterations so a resolver
* bug cannot loop forever. In practice 1–2 attempts are enough (suffix list is short).
*/
const MAX_EXTRACT_SCHEMA_KEY_COLLISION_ATTEMPTS = 32;
var SchemaRoutes = class {
	schemaUtils;
	FORM_DATA_TYPES = [];
	routes = [];
	hasSecurityRoutes = false;
	hasQueryRoutes = false;
	hasFormDataRoutes = false;
	constructor(config, schemaParserFabric, schemaComponentsMap, templatesWorker, typeNameFormatter) {
		this.config = config;
		this.schemaParserFabric = schemaParserFabric;
		this.schemaComponentsMap = schemaComponentsMap;
		this.templatesWorker = templatesWorker;
		this.typeNameFormatter = typeNameFormatter;
		this.schemaUtils = this.schemaParserFabric.schemaUtils;
		this.FORM_DATA_TYPES = (0, es_toolkit.uniq)([this.schemaUtils.getSchemaType({
			type: "string",
			format: "file"
		}), this.schemaUtils.getSchemaType({
			type: "string",
			format: "binary"
		})]);
	}
	/**
	* `extractResponseBody` / `extractResponseError` call `createParsedComponent`, which
	* registers `#/components/schemas/<typeName>`. If that key already exists (e.g.
	* `MergeFluffyData` in definitions), the map entry would be overwritten unless we
	* pick another name via `resolveTypeName` after reserving the colliding one.
	*
	* `getComponents` may be missing in narrow unit tests that pass a stub map.
	*
	* `resolveTypeName` ends in `NameResolver.resolve`, which **reserves** the chosen
	* string when `shouldReserve` is true (default). So after a colliding first pick,
	* the next `resolveTypeName` skips that variant. The extra `reserve([typeName])`
	* is still needed when callers pass `shouldReserve: false` — then the first pick
	* is not auto-reserved and we must block it before retrying.
	*/
	extractTypeNameWithoutSchemaKeyCollision = (routeNameUsage, options) => {
		const refFor = (name) => this.schemaComponentsMap.createRef([
			"components",
			"schemas",
			name
		]);
		const existingComponents = (0, yummies_common.callFunction)(this.schemaComponentsMap.getComponents) ?? [];
		const collides = (name) => !!name && existingComponents.some((c) => c.$ref === refFor(name));
		let typeName = this.schemaUtils.resolveTypeName(routeNameUsage, options);
		for (let attempt = 0; attempt < MAX_EXTRACT_SCHEMA_KEY_COLLISION_ATTEMPTS; attempt++) {
			if (!collides(typeName)) break;
			this.config.componentTypeNameResolver.reserve([typeName]);
			typeName = this.schemaUtils.resolveTypeName(routeNameUsage, options);
		}
		return typeName;
	};
	createRequestsMap = (resolvedSwaggerSchema, routesByMethod) => {
		const parameters = (0, es_toolkit_compat.get)(routesByMethod, "parameters");
		const result = {};
		for (const [method, requestInfo] of Object.entries(routesByMethod)) {
			if (method.startsWith("x-") || ["parameters"].includes(method)) continue;
			if (method === "$ref") {
				const refData = resolvedSwaggerSchema.getRef(requestInfo);
				if (yummies_type_guard.typeGuard.isObject(refData)) Object.assign(result, this.createRequestsMap(resolvedSwaggerSchema, refData));
				continue;
			}
			result[method] = {
				...requestInfo,
				parameters: (0, es_toolkit.compact)([...parameters || [], ...requestInfo.parameters || []])
			};
		}
		return result;
	};
	parseRouteName = (rawRoute) => {
		const routeName = this.config.hooks.onPreBuildRoutePath(rawRoute) || rawRoute;
		const pathParamMatches = (routeName || "").match(/({[\w[\\\]^`][-_.\w]*})|(:[\w[\\\]^`][-_.\w]*:?)/g);
		const pathParams = [];
		for (const match of pathParamMatches || []) {
			const paramName = match.replace(/\{|\}|:/g, "");
			if (!paramName) continue;
			if (paramName.includes("-")) consola.consola.warn("wrong path param name", paramName);
			pathParams.push({
				$match: match,
				name: (0, es_toolkit_compat.camelCase)(paramName),
				required: true,
				type: "string",
				description: "",
				schema: { type: "string" },
				in: "path"
			});
		}
		let fixedRoute = pathParams.reduce((fixedRoute, pathParam, i, arr) => {
			const insertion = this.config.hooks.onInsertPathParam(pathParam.name, i, arr, fixedRoute) || pathParam.name;
			return fixedRoute.replace(pathParam.$match, `\${${insertion}}`);
		}, routeName || "");
		const queryParamMatches = fixedRoute.match(/(\{\?.*\})/g);
		const queryParams = [];
		if (queryParamMatches?.length) {
			for (const match of queryParamMatches) fixedRoute = fixedRoute.replace(match, "");
			const paramNames = (0, es_toolkit.uniq)(queryParamMatches.join(",").replace(/(\{\?)|(\})|\s/g, "").split(","));
			for (const paramName of paramNames) {
				if (typeof paramName === "string" && paramName.includes("-")) consola.consola.warn("wrong query param name", paramName);
				queryParams.push({
					$match: paramName,
					name: typeof paramName === "string" ? (0, es_toolkit_compat.camelCase)(paramName) : (0, es_toolkit_compat.camelCase)(String(paramName)),
					required: true,
					type: "string",
					description: "",
					schema: { type: "string" },
					in: "query"
				});
			}
		}
		const result = {
			originalRoute: rawRoute || "",
			route: fixedRoute,
			pathParams,
			queryParams
		};
		return this.config.hooks.onBuildRoutePath(result) || result;
	};
	getRouteParams = (routeInfo, pathParamsFromRouteName, queryParamsFromRouteName) => {
		const { parameters } = routeInfo;
		const routeParams = {
			path: [],
			header: [],
			body: [],
			query: [],
			formData: [],
			cookie: []
		};
		for (const parameter of parameters || []) {
			const refTypeInfo = this.schemaParserFabric.schemaUtils.getSchemaRefType(parameter);
			let routeParam = null;
			if (!!refTypeInfo?.rawTypeData && typeof refTypeInfo === "object" && refTypeInfo?.rawTypeData.in) {
				if (!routeParams[refTypeInfo.rawTypeData.in]) routeParams[refTypeInfo.rawTypeData.in] = [];
				routeParam = {
					...refTypeInfo.rawTypeData,
					...refTypeInfo.rawTypeData.schema || {}
				};
				if (parameter?.required && !routeParam.required) routeParam.required = parameter.required;
			} else {
				if (!parameter.in) continue;
				if (!routeParams[parameter.in]) routeParams[parameter.in] = [];
				routeParam = {
					...parameter,
					...parameter.schema || {}
				};
			}
			if (routeParam.in === "path") {
				if (!routeParam.name) continue;
				routeParam.name = (0, es_toolkit_compat.camelCase)(routeParam.name);
			}
			routeParams[routeParam.in].push(routeParam);
		}
		for (const pathParam of pathParamsFromRouteName) if (!routeParams.path.some((parameter) => parameter.name === pathParam.name)) routeParams.path.push(pathParam);
		for (const queryParam of queryParamsFromRouteName) if (!routeParams.query.some((parameter) => parameter.name === queryParam.name)) routeParams.query.push(queryParam);
		return routeParams;
	};
	getContentTypes = (requestInfo, extraContentTypes) => {
		const requestInfoArray = Array.isArray(requestInfo) ? requestInfo : Object.values(requestInfo || {});
		return (0, es_toolkit.uniq)((0, es_toolkit.compact)([...extraContentTypes || [], ...(0, es_toolkit.flattenDeep)(requestInfoArray.map((requestInfoData) => requestInfoData && Object.keys(requestInfoData?.content || {})))]));
	};
	getContentKind = (contentTypes) => {
		if (contentTypes.includes("application/vnd.api+json")) return CONTENT_KIND.JSON_API;
		if (contentTypes.some((contentType) => contentType.startsWith("application/json")) || contentTypes.some((contentType) => contentType.endsWith("+json"))) return CONTENT_KIND.JSON;
		if (contentTypes.includes("application/x-www-form-urlencoded")) return CONTENT_KIND.URL_ENCODED;
		if (contentTypes.includes("multipart/form-data")) return CONTENT_KIND.FORM_DATA;
		if (contentTypes.some((contentType) => contentType.includes("image/"))) return CONTENT_KIND.IMAGE;
		if (contentTypes.some((contentType) => contentType.startsWith("text/"))) return CONTENT_KIND.TEXT;
		return CONTENT_KIND.OTHER;
	};
	/** True when response produces only binary media types (e.g. file download). */
	isBinaryOnlyContentTypes = (contentTypes) => !!contentTypes?.length && contentTypes.every((ct) => this.schemaUtils.isBinaryLikeMimeType(ct));
	isSuccessStatus = (status) => this.config.defaultResponseAsSuccess && status === "default" || +status >= this.config.successResponseStatusRange[0] && +status <= this.config.successResponseStatusRange[1] || status === "2xx";
	getSchemaFromRequestType = (requestInfo) => {
		const content = (0, es_toolkit_compat.get)(requestInfo, "content");
		if (!content) return null;
		for (const dataType in content) if (content[dataType]?.schema) return {
			...content[dataType].schema,
			dataType
		};
		return null;
	};
	getTypeFromRequestInfo = ({ requestInfo, parsedSchemas, operationId, defaultType, typeName }) => {
		const schema = this.getSchemaFromRequestType(requestInfo);
		const refTypeInfo = this.schemaParserFabric.schemaUtils.getSchemaRefType(requestInfo);
		if (schema) {
			const content = this.schemaParserFabric.getInlineParseContent(schema, typeName, [operationId]);
			const foundedSchemaByName = parsedSchemas.find((parsedSchema) => this.typeNameFormatter.format(parsedSchema.name) === content);
			const foundSchemaByContent = parsedSchemas.find((parsedSchema) => (0, es_toolkit.isEqual)(parsedSchema.content, content));
			const foundSchema = foundedSchemaByName || foundSchemaByContent;
			return foundSchema ? this.typeNameFormatter.format(foundSchema.name) : content;
		}
		if (refTypeInfo) {
			const typeNameWithoutOpId = refTypeInfo.typeName.replace(operationId, "");
			if (parsedSchemas.find((schema) => schema.name === typeNameWithoutOpId)) return this.typeNameFormatter.format(typeNameWithoutOpId);
			switch (refTypeInfo.componentName) {
				case "schemas": return this.typeNameFormatter.format(refTypeInfo.typeName);
				case "responses":
				case "requestBodies": return this.schemaParserFabric.getInlineParseContent(this.getSchemaFromRequestType(refTypeInfo.rawTypeData), refTypeInfo.typeName || null, [operationId]);
				default: return this.schemaParserFabric.getInlineParseContent(refTypeInfo.rawTypeData, refTypeInfo.typeName || null, [operationId]);
			}
		}
		return defaultType || this.config.Ts.Keyword.Any;
	};
	getRequestInfoTypes = ({ requestInfos, parsedSchemas, operationId, defaultType, resolvedSwaggerSchema }) => {
		const result = [];
		for (const [status, requestInfo] of Object.entries(requestInfos || {})) {
			const contentTypes = this.getContentTypes([requestInfo]);
			const links = this.getRouteLinksFromResponse(resolvedSwaggerSchema, requestInfo, status);
			result.push({
				...requestInfo || {},
				contentTypes,
				contentKind: this.getContentKind(contentTypes),
				type: this.schemaParserFabric.schemaUtils.safeAddNullToType(requestInfo, this.getTypeFromRequestInfo({
					requestInfo,
					parsedSchemas,
					operationId,
					defaultType
				})),
				description: this.schemaParserFabric.schemaFormatters.formatDescription(requestInfo.description || "", true),
				links,
				status: Number.isNaN(+status) ? status : +status,
				isSuccess: this.isSuccessStatus(status)
			});
		}
		return result;
	};
	getRouteLinksFromResponse = (resolvedSwaggerSchema, responseInfo, status) => {
		const links = (0, es_toolkit_compat.get)(responseInfo, "links");
		if (!yummies_type_guard.typeGuard.isObject(links)) return [];
		return (0, es_toolkit_compat.reduce)(links, (acc, linkInfo, linkName) => {
			if (!yummies_type_guard.typeGuard.isObject(linkInfo)) return acc;
			let normalizedLinkInfo = linkInfo;
			if (typeof linkInfo.$ref === "string") {
				const refData = resolvedSwaggerSchema.getRef(linkInfo.$ref);
				if (yummies_type_guard.typeGuard.isObject(refData)) normalizedLinkInfo = refData;
			}
			const operationId = typeof normalizedLinkInfo.operationId === "string" ? normalizedLinkInfo.operationId : void 0;
			const operationRef = typeof normalizedLinkInfo.operationRef === "string" ? normalizedLinkInfo.operationRef : typeof linkInfo.$ref === "string" ? linkInfo.$ref : void 0;
			if (!operationId && !operationRef) return acc;
			const parameters = yummies_type_guard.typeGuard.isObject(normalizedLinkInfo.parameters) ? (0, es_toolkit.mapValues)(normalizedLinkInfo.parameters, (value) => String(value)) : void 0;
			acc.push({
				status: Number.isNaN(+status) ? status : +status,
				name: String(linkName),
				operationId,
				operationRef,
				parameters
			});
			return acc;
		}, []);
	};
	getResponseBodyInfo = (routeInfo, parsedSchemas, resolvedSwaggerSchema, pathName, method) => {
		const { produces, operationId, responses } = routeInfo;
		const contentTypes = this.getContentTypes(responses, [...produces || [], routeInfo["x-accepts"]]);
		const successStatus = Object.keys(responses || {}).find((s) => this.isSuccessStatus(s));
		const successResponseContent = successStatus && responses?.[successStatus];
		const successContentTypes = successResponseContent?.content && typeof successResponseContent.content === "object" ? Object.keys(successResponseContent.content) : null;
		const originalProduces = pathName && method ? resolvedSwaggerSchema.getOriginalProduces(pathName, method) ?? (0, es_toolkit_compat.get)(resolvedSwaggerSchema.originalSchema, [
			"paths",
			pathName,
			method,
			"produces"
		]) : void 0;
		const responseInfos = this.getRequestInfoTypes({
			requestInfos: responses,
			parsedSchemas,
			operationId,
			defaultType: this.config.defaultResponseType,
			resolvedSwaggerSchema
		});
		const links = responseInfos.flatMap((responseInfo) => responseInfo.links || []);
		const successResponse = responseInfos.find((response) => response.isSuccess);
		const errorResponses = responseInfos.filter((response) => !response.isSuccess && response.type !== this.config.Ts.Keyword.Any);
		const handleResponseHeaders = (src) => {
			if (!src) return "headers: {},";
			const headerTypes = Object.fromEntries(Object.entries(src).map(([k, v]) => {
				return [k, this.schemaUtils.getSchemaType(v)];
			}));
			return `headers: { ${Object.entries(headerTypes).map(([k, v]) => `"${k}": ${v}`).join(",")} },`;
		};
		const typesToCheck = (Array.isArray(originalProduces) && originalProduces.length > 0 ? originalProduces : null) ?? (produces?.length ? produces : null) ?? (successContentTypes?.length ? successContentTypes : null) ?? contentTypes;
		const isBinarySuccessType = this.isBinaryOnlyContentTypes(typesToCheck);
		return {
			contentTypes,
			responses: responseInfos,
			links,
			typesToCheck,
			success: {
				isBinary: isBinarySuccessType,
				schema: successResponse,
				type: isBinarySuccessType ? this.config.Ts.Keyword.Blob : successResponse?.type || this.config.Ts.Keyword.Any
			},
			error: {
				schemas: errorResponses,
				type: this.config.Ts.UnionType(errorResponses.map((response) => response.type)) || this.config.Ts.Keyword.Any
			},
			full: { types: this.config.Ts.UnionType(responseInfos.map((response) => `{
      data: ${response.type}, status: ${response.status}, statusCode: ${response.status}, statusText: "${response.description}", ${handleResponseHeaders(response.headers)} config: {} }`)) || this.config.Ts.Keyword.Any }
		};
	};
	convertRouteParamsIntoObject = (params) => {
		return params.reduce((objectSchema, schemaPart) => {
			if (!schemaPart || !schemaPart.name) return objectSchema;
			return {
				...objectSchema,
				properties: {
					...objectSchema.properties,
					[schemaPart.name]: {
						...schemaPart,
						...schemaPart.schema || {}
					}
				}
			};
		}, {
			properties: {},
			type: "object"
		});
	};
	getRequestBodyInfo = (routeInfo, routeParams, parsedSchemas, routeName) => {
		const { requestBody, consumes, requestBodyName, operationId } = routeInfo;
		let schema = null;
		let content = null;
		const contentTypes = this.getContentTypes([requestBody], [...consumes || [], routeInfo["x-contentType"]]);
		let contentKind = this.getContentKind(contentTypes);
		let typeName = null;
		if (this.config.extractRequestBody) typeName = this.schemaUtils.resolveTypeName(routeName.usage, {
			suffixes: this.config.extractingOptions.requestBodySuffix,
			resolver: this.config.extractingOptions.requestBodyNameResolver
		});
		if (routeParams.formData.length) {
			contentKind = CONTENT_KIND.FORM_DATA;
			schema = this.convertRouteParamsIntoObject(routeParams.formData);
			content = this.schemaParserFabric.getInlineParseContent(schema, typeName, [operationId]);
		} else if (contentKind === CONTENT_KIND.FORM_DATA) {
			schema = this.getSchemaFromRequestType(requestBody);
			content = this.schemaParserFabric.getInlineParseContent(schema, typeName, [operationId]);
		} else if (requestBody) {
			schema = this.getSchemaFromRequestType(requestBody);
			content = this.schemaParserFabric.schemaUtils.safeAddNullToType(requestBody, this.getTypeFromRequestInfo({
				requestInfo: requestBody,
				parsedSchemas,
				operationId,
				typeName
			}));
			if (this.FORM_DATA_TYPES.some((dataType) => content.includes(`: ${dataType}`))) contentKind = CONTENT_KIND.FORM_DATA;
		}
		if (schema && !schema.$ref && this.config.extractRequestBody) {
			schema = this.schemaParserFabric.createParsedComponent({
				schema,
				typeName,
				schemaPath: [operationId]
			});
			if (schema?.typeData) schema.typeData.isExtractedRequestBody = true;
			content = this.schemaParserFabric.getInlineParseContent({ $ref: schema.$ref });
		}
		if (schema && schema.typeData && !schema.typeData.description && requestBody?.description) schema.typeData.description = requestBody.description;
		return {
			...requestBody || {},
			paramName: requestBodyName || requestBody?.name || "data",
			contentTypes,
			contentKind,
			schema,
			type: content,
			required: requestBody && (typeof requestBody.required === "undefined" || !!requestBody.required)
		};
	};
	createRequestParamsSchema = ({ queryParams, queryObjectSchema, pathArgsSchemas, extractRequestParams, routeName }) => {
		if ((!queryParams || !queryParams.length) && (!pathArgsSchemas || !pathArgsSchemas.length)) return null;
		const pathParams = pathArgsSchemas.reduce((acc, pathArgSchema) => {
			if (pathArgSchema.name) acc[pathArgSchema.name] = {
				...pathArgSchema,
				in: "path"
			};
			return acc;
		}, {});
		const fixedQueryParams = {};
		const queryObjectProperties = (0, es_toolkit_compat.get)(queryObjectSchema, "properties") || {};
		for (const [name, property] of Object.entries(queryObjectProperties)) if (name && typeof property === "object") fixedQueryParams[name] = {
			...property,
			in: "query"
		};
		const schema = {
			...queryObjectSchema,
			properties: {
				...fixedQueryParams,
				...pathParams
			}
		};
		const fixedSchema = this.config.hooks.onCreateRequestParams(schema);
		if (fixedSchema) return fixedSchema;
		if (extractRequestParams) {
			const generatedTypeName = this.schemaUtils.resolveTypeName(routeName.usage, {
				suffixes: this.config.extractingOptions.requestParamsSuffix,
				resolver: this.config.extractingOptions.requestParamsNameResolver
			});
			const component = this.schemaParserFabric.createParsedComponent({
				typeName: generatedTypeName,
				schema
			});
			if (component.typeData) component.typeData.isExtractedRequestParams = true;
			return component;
		}
		return schema;
	};
	extractResponseBodyIfItNeeded = (routeInfo, responseBodyInfo, routeName) => {
		if (responseBodyInfo.responses.length && responseBodyInfo.success && responseBodyInfo.success.schema) {
			const typeName = this.extractTypeNameWithoutSchemaKeyCollision(routeName.usage, {
				suffixes: this.config.extractingOptions.responseBodySuffix,
				resolver: this.config.extractingOptions.responseBodyNameResolver
			});
			const idx = responseBodyInfo.responses.indexOf(responseBodyInfo.success.schema);
			const successResponse = responseBodyInfo.success;
			const contentKind = successResponse.schema?.contentKind;
			const actualSchema = this.getSchemaFromRequestType(successResponse.schema);
			if (actualSchema && !actualSchema.$ref) {
				successResponse.schema = this.schemaParserFabric.createParsedComponent({
					schema: actualSchema,
					typeName,
					schemaPath: [routeInfo.operationId]
				});
				successResponse.schema.contentKind = contentKind;
				if (successResponse.schema.typeData) successResponse.schema.typeData.isExtractedResponseBody = true;
				successResponse.type = this.schemaParserFabric.getInlineParseContent({ $ref: successResponse.schema.$ref });
				if (idx > -1) Object.assign(responseBodyInfo.responses[idx], {
					...successResponse.schema,
					type: successResponse.type
				});
			} else if (responseBodyInfo.success.isBinary) {
				successResponse.schema = this.schemaParserFabric.createParsedComponent({
					schema: {
						type: "string",
						format: "byte"
					},
					typeName,
					schemaPath: [routeInfo.operationId]
				});
				successResponse.schema.contentKind = contentKind;
				if (successResponse.schema.typeData) successResponse.schema.typeData.isExtractedResponseBody = true;
				successResponse.type = this.config.Ts.Keyword.Blob;
				if (idx > -1) Object.assign(responseBodyInfo.responses[idx], {
					...successResponse.schema,
					type: successResponse.type
				});
			} else if (actualSchema?.$ref) {
				successResponse.schema = this.schemaParserFabric.createParsedComponent({
					schema: actualSchema,
					typeName,
					schemaPath: [routeInfo.operationId]
				});
				successResponse.schema.contentKind = contentKind;
				if (successResponse.schema.typeData) successResponse.schema.typeData.isExtractedResponseBody = true;
				successResponse.type = this.schemaParserFabric.getInlineParseContent({ $ref: successResponse.schema.$ref });
				if (idx > -1) Object.assign(responseBodyInfo.responses[idx], {
					...successResponse.schema,
					type: successResponse.type
				});
			} else if (successResponse.schema && actualSchema === null && (responseBodyInfo.success.type === this.config.Ts.Keyword.Any || responseBodyInfo.success.type === this.config.defaultResponseType)) {
				successResponse.schema = this.schemaParserFabric.createParsedComponent({
					schema: {},
					typeName,
					schemaPath: [routeInfo.operationId]
				});
				successResponse.schema.contentKind = contentKind;
				if (successResponse.schema.typeData) successResponse.schema.typeData.isExtractedResponseBody = true;
				successResponse.type = this.schemaParserFabric.getInlineParseContent({ $ref: successResponse.schema.$ref });
				if (idx > -1) Object.assign(responseBodyInfo.responses[idx], {
					...successResponse.schema,
					type: successResponse.type
				});
			}
		}
	};
	extractResponseErrorIfItNeeded = (routeInfo, responseBodyInfo, routeName) => {
		if (responseBodyInfo.responses.length && responseBodyInfo.error.schemas && responseBodyInfo.error.schemas.length) {
			const typeName = this.extractTypeNameWithoutSchemaKeyCollision(routeName.usage, {
				suffixes: this.config.extractingOptions.responseErrorSuffix,
				resolver: this.config.extractingOptions.responseErrorNameResolver
			});
			const errorSchemas = (0, es_toolkit.compact)(responseBodyInfo.error.schemas.map(this.getSchemaFromRequestType));
			if (!errorSchemas.length) return;
			const schema = this.schemaParserFabric.parseSchema({
				oneOf: errorSchemas,
				title: (0, es_toolkit.compact)(errorSchemas.map((schema) => schema.title)).join(" "),
				description: (0, es_toolkit.compact)(errorSchemas.map((schema) => schema.description)).join("\n")
			}, null, [routeInfo.operationId]);
			const component = this.schemaComponentsMap.createComponent(this.schemaComponentsMap.createRef([
				"components",
				"schemas",
				typeName
			]), schema);
			responseBodyInfo.error.schemas = [component];
			if (component.typeData) component.typeData.isExtractedResponseError = true;
			responseBodyInfo.error.type = this.typeNameFormatter.format(component.typeName);
		}
	};
	getRouteName = (rawRouteInfo) => {
		const { moduleName } = rawRouteInfo;
		const { routeNameDuplicatesMap, templatesToRender } = this.config;
		const routeNameTemplate = templatesToRender.routeName;
		const routeNameFromTemplate = this.templatesWorker.renderTemplate(routeNameTemplate, { routeInfo: rawRouteInfo });
		const routeName = this.config.hooks.onFormatRouteName?.(rawRouteInfo, routeNameFromTemplate) || routeNameFromTemplate;
		const duplicateIdentifier = `${moduleName}|${routeName}`;
		if (routeNameDuplicatesMap.has(duplicateIdentifier)) {
			routeNameDuplicatesMap.set(duplicateIdentifier, routeNameDuplicatesMap.get(duplicateIdentifier) + 1);
			consola.consola.warn(`Module "${moduleName}" already has method "${routeName}()".`, `This method has been renamed to "${routeName + routeNameDuplicatesMap.get(duplicateIdentifier)}()" to solve conflict names.`);
		} else routeNameDuplicatesMap.set(duplicateIdentifier, 1);
		const duplicates = routeNameDuplicatesMap.get(duplicateIdentifier);
		const routeNameInfo = {
			usage: routeName + (duplicates > 1 ? duplicates : ""),
			original: routeName,
			duplicate: duplicates > 1
		};
		return this.config.hooks.onCreateRouteName?.(routeNameInfo, rawRouteInfo) || routeNameInfo;
	};
	parseRouteInfo = (rawRouteName, routeInfo, method, resolvedSwaggerSchema, parsedSchemas, routeServers) => {
		const { security: globalSecurity } = resolvedSwaggerSchema.usageSchema;
		const { moduleNameIndex, moduleNameFirstTag, extractRequestParams } = this.config;
		const { operationId, requestBody, security, parameters, summary, description, tags, responses, requestBodyName, produces, consumes } = routeInfo;
		routeInfo.servers = routeInfo.servers ?? routeServers;
		const { route, pathParams: pathParamsFromRouteName, queryParams: queryParamsFromRouteName } = this.parseRouteName(rawRouteName);
		const routeId = generateId();
		const firstTag = tags && tags.length > 0 ? tags[0] : null;
		const moduleName = moduleNameFirstTag && firstTag ? (0, es_toolkit_compat.camelCase)(firstTag) : (0, es_toolkit_compat.camelCase)((0, es_toolkit.compact)(route.split("/"))[moduleNameIndex] || "");
		let hasSecurity = !!globalSecurity?.length;
		if (security) hasSecurity = security.length > 0;
		const routeParams = this.getRouteParams(routeInfo, pathParamsFromRouteName, queryParamsFromRouteName);
		const pathArgs = routeParams.path.map((pathArgSchema) => ({
			name: pathArgSchema.name,
			optional: !pathArgSchema.required,
			type: this.config.Ts.Keyword.Any,
			description: pathArgSchema.description
		}));
		const pathArgsNames = pathArgs.map((arg) => arg.name);
		const responseBodyInfo = this.getResponseBodyInfo(routeInfo, parsedSchemas, resolvedSwaggerSchema, rawRouteName, method);
		const rawRouteInfo = {
			...routeInfo,
			pathArgs,
			operationId,
			method,
			route: rawRouteName,
			moduleName,
			responsesTypes: responseBodyInfo.responses,
			links: responseBodyInfo.links,
			description,
			tags,
			summary,
			responses,
			produces,
			requestBody,
			consumes,
			security
		};
		const queryObjectSchema = this.convertRouteParamsIntoObject(routeParams.query);
		const pathObjectSchema = this.convertRouteParamsIntoObject(routeParams.path);
		const headersObjectSchema = this.convertRouteParamsIntoObject(routeParams.header);
		const routeName = this.getRouteName(rawRouteInfo);
		const requestBodyInfo = this.getRequestBodyInfo(routeInfo, routeParams, parsedSchemas, routeName);
		const requestParamsSchema = this.createRequestParamsSchema({
			queryParams: routeParams.query,
			pathArgsSchemas: routeParams.path,
			queryObjectSchema,
			extractRequestParams,
			routeName
		});
		if (this.config.extractResponseBody) this.extractResponseBodyIfItNeeded(routeInfo, responseBodyInfo, routeName);
		if (this.config.extractResponseError) this.extractResponseErrorIfItNeeded(routeInfo, responseBodyInfo, routeName);
		const typeName = this.schemaUtils.resolveTypeName(routeName.usage, {
			suffixes: this.config.extractingOptions.requestParamsSuffix,
			resolver: this.config.extractingOptions.requestParamsNameResolver,
			shouldReserve: false
		});
		const queryType = routeParams.query.length ? this.schemaParserFabric.getInlineParseContent(queryObjectSchema, null, [typeName]) : null;
		const pathType = routeParams.path.length ? this.schemaParserFabric.getInlineParseContent(pathObjectSchema, null, [typeName]) : null;
		const headersType = routeParams.header.length ? this.schemaParserFabric.getInlineParseContent(headersObjectSchema, null, [typeName]) : null;
		const nameResolver = new SpecificArgNameResolver(this.config, pathArgsNames);
		const specificArgs = {
			query: queryType ? {
				name: nameResolver.resolve(RESERVED_QUERY_ARG_NAMES),
				optional: this.schemaParserFabric.parseSchema(queryObjectSchema, null, [routeName.usage]).allFieldsAreOptional,
				type: queryType
			} : void 0,
			body: requestBodyInfo.type ? {
				...requestBodyInfo,
				name: nameResolver.resolve([requestBodyInfo.paramName, ...RESERVED_BODY_ARG_NAMES]),
				optional: !requestBodyInfo.required,
				type: requestBodyInfo.type
			} : void 0,
			pathParams: pathType ? {
				name: nameResolver.resolve(RESERVED_PATH_ARG_NAMES),
				optional: this.schemaParserFabric.parseSchema(pathObjectSchema, null, [routeName.usage]).allFieldsAreOptional,
				type: pathType
			} : void 0,
			headers: headersType ? {
				name: nameResolver.resolve(RESERVED_HEADER_ARG_NAMES),
				optional: this.schemaParserFabric.parseSchema(headersObjectSchema, null, [routeName.usage]).allFieldsAreOptional,
				type: headersType
			} : void 0
		};
		pathArgs.forEach((pathArg, i) => {
			pathArg.type = this.schemaParserFabric.getInlineParseContent(routeParams.path[i].schema, null, [typeName]);
		});
		return {
			id: routeId,
			namespace: moduleName.replace(/^(\d)/, "v$1"),
			routeName,
			routeParams,
			requestBodyInfo,
			responseBodyInfo,
			specificArgs,
			queryObjectSchema,
			pathObjectSchema,
			headersObjectSchema,
			responseBodySchema: responseBodyInfo.success.schema,
			requestBodySchema: requestBodyInfo.schema,
			specificArgNameResolver: nameResolver,
			request: {
				contentTypes: requestBodyInfo.contentTypes,
				parameters: pathArgs,
				path: route,
				formData: requestBodyInfo.contentKind === CONTENT_KIND.FORM_DATA,
				isQueryBody: requestBodyInfo.contentKind === CONTENT_KIND.URL_ENCODED,
				security: hasSecurity,
				method,
				requestParams: requestParamsSchema,
				payload: specificArgs.body,
				query: specificArgs.query,
				pathParams: specificArgs.pathParams,
				headers: specificArgs.headers
			},
			response: {
				contentTypes: responseBodyInfo.contentTypes,
				type: responseBodyInfo.success.type,
				errorType: responseBodyInfo.error.type,
				fullTypes: responseBodyInfo.full.types
			},
			raw: rawRouteInfo
		};
	};
	attachSchema = (resolvedSwaggerSchema, parsedSchemas) => {
		this.config.routeNameDuplicatesMap.clear();
		const pathsEntries = Object.entries(resolvedSwaggerSchema.usageSchema.paths || {});
		for (const [rawRouteName, routeInfoByMethodsMap] of pathsEntries) {
			const routeInfosMap = this.createRequestsMap(resolvedSwaggerSchema, routeInfoByMethodsMap);
			const routeServers = routeInfosMap["servers"];
			for (const [method, routeInfo] of Object.entries(routeInfosMap)) {
				if (method === "servers") continue;
				const parsedRouteInfo = this.parseRouteInfo(rawRouteName, routeInfo, method, resolvedSwaggerSchema, parsedSchemas, routeServers);
				const processedRouteInfo = this.config.hooks.onCreateRoute(parsedRouteInfo);
				if (processedRouteInfo !== false) {
					const route = processedRouteInfo || parsedRouteInfo;
					if (!this.hasSecurityRoutes && route.security) this.hasSecurityRoutes = route.security;
					if (!this.hasQueryRoutes && route.hasQuery) this.hasQueryRoutes = route.hasQuery;
					if (!this.hasFormDataRoutes && route.hasFormDataParams) this.hasFormDataRoutes = route.hasFormDataParams;
					this.routes.push(route);
				}
			}
		}
	};
	getGroupedRoutes = () => {
		const groupedRoutes = this.routes.reduce((modules, route) => {
			if (route.namespace) {
				if (!modules[route.namespace]) modules[route.namespace] = [];
				modules[route.namespace].push(route);
			} else modules.$outOfModule.push(route);
			return modules;
		}, { $outOfModule: [] });
		const routeGroups = {
			outOfModule: void 0,
			combined: void 0
		};
		for (const [moduleName, routesGroup] of Object.entries(groupedRoutes)) if (moduleName === "$outOfModule") routeGroups.outOfModule = routesGroup;
		else {
			if (!routeGroups.combined) routeGroups.combined = [];
			routeGroups.combined.push({
				moduleName,
				routes: routesGroup.map((route) => {
					const { original: originalName, usage: usageName } = route.routeName;
					if (routesGroup.length > 1 && usageName !== originalName && !routesGroup.some(({ routeName, id }) => id !== route.id && originalName === routeName.original)) return {
						...route,
						routeName: {
							...route.routeName,
							usage: originalName
						}
					};
					return route;
				})
			});
		}
		if (this.config.sortRoutes) {
			if (routeGroups.outOfModule) routeGroups.outOfModule = this.sortRoutes(routeGroups.outOfModule);
			if (routeGroups.combined) for (const routeGroup of routeGroups.combined) routeGroup.routes = this.sortRoutes(routeGroup.routes);
		}
		return routeGroups;
	};
	sortRoutes = (routes) => {
		return routes.slice().sort((routeA, routeB) => routeA.routeName.usage.localeCompare(routeB.routeName.usage));
	};
};
//#endregion
//#region src/util/parse-schema-content.ts
/**
* Some OpenAPI YAML files use a double-quoted value that spans a newline without
* a trailing `\` (invalid YAML 1.2; see acacode/swagger-typescript-api#1433):
*
*   description: "first line
*   tail"
*
* `yaml` rejects that with "Missing closing quote". We fold it into one line using
* an escaped newline inside the scalar: `"first line\ntail"`.
*
* If the first line ends with `\`, it is a normal YAML line continuation and is
* left for {@link normalizeYamlEscapedLineBreaks}.
*/
function mergeBrokenDoubleQuotedMapLines(content) {
	const lines = content.split(/\r?\n|\r/);
	const result = [];
	for (let i = 0; i < lines.length; i += 1) {
		const current = lines[i];
		if (current === void 0) break;
		const following = lines[i + 1];
		const repaired = following === void 0 ? null : tryRepairSplitDoubleQuotedMapValue(current, following);
		if (repaired) {
			result.push(repaired);
			i += 1;
			continue;
		}
		result.push(current);
	}
	return result.join("\n");
}
function tryRepairSplitDoubleQuotedMapValue(line, nextLine) {
	const unterminated = matchMapKeyOpeningDoubleQuotedScalar(line);
	if (!unterminated) return null;
	const closing = matchClosingQuoteLine(nextLine);
	if (!closing) return null;
	const { keyPrefixAndQuote, valueBeforeBreak } = unterminated;
	const { lineBeforeClosingQuote } = closing;
	if (valueBeforeBreak.endsWith("\\")) return null;
	return `${keyPrefixAndQuote}${valueBeforeBreak}\\n${lineBeforeClosingQuote}"`;
}
/** `  foo: "text` with no `"` before end of line (value may be empty). */
function matchMapKeyOpeningDoubleQuotedScalar(line) {
	const match = /^(\s+[A-Za-z_][\w-]*:\s*")([^"]*)$/.exec(line);
	if (!match?.[1] || match[2] === void 0) return null;
	return {
		keyPrefixAndQuote: match[1],
		valueBeforeBreak: match[2]
	};
}
/** `      any text here"` — indented continuation that closes the quote. */
function matchClosingQuoteLine(line) {
	const match = /^(\s+)([^"]+?)"\s*$/.exec(line);
	if (!match?.[2]) return null;
	return { lineBeforeClosingQuote: match[2].trim() };
}
function normalizeYamlEscapedLineBreaks(content) {
	let normalized = "";
	let inDoubleQuotedScalar = false;
	let escaped = false;
	for (let index = 0; index < content.length; index += 1) {
		const currentChar = content[index];
		if (!inDoubleQuotedScalar) {
			if (currentChar === "\"") inDoubleQuotedScalar = true;
			normalized += currentChar;
			continue;
		}
		if (escaped) {
			normalized += currentChar;
			escaped = false;
			continue;
		}
		if (currentChar === "\\") {
			const nextChar = content[index + 1];
			if (nextChar === "\n" || nextChar === "\r") {
				index += 1;
				if (nextChar === "\r" && content[index + 1] === "\n") index += 1;
				while (content[index + 1] === " " || content[index + 1] === "	") index += 1;
				continue;
			}
			normalized += currentChar;
			escaped = true;
			continue;
		}
		if (currentChar === "\"") inDoubleQuotedScalar = false;
		normalized += currentChar;
	}
	return normalized;
}
function parseSchemaContent(content) {
	try {
		return JSON.parse(content);
	} catch {
		const merged = mergeBrokenDoubleQuotedMapLines(content);
		return yaml.parse(normalizeYamlEscapedLineBreaks(merged));
	}
}
//#endregion
//#region src/resolved-swagger-schema.ts
var ResolvedSwaggerSchema = class ResolvedSwaggerSchema {
	parsedRefsCache = /* @__PURE__ */ new Map();
	externalSchemaCache = /* @__PURE__ */ new Map();
	httpMethodSegments = new Set([
		"get",
		"put",
		"post",
		"delete",
		"patch",
		"options",
		"head",
		"trace",
		"parameters"
	]);
	normalizeRef(ref) {
		let normalizedRef = ref;
		normalizedRef = normalizedRef.replace(/\/#(?=\/)/, "#");
		normalizedRef = normalizedRef.replace(/#(?!\/)/, "#/");
		return normalizedRef;
	}
	createEscapedPathsRefVariant(ref) {
		const [prefix = "", rawPointer = ""] = this.normalizeRef(ref).split("#");
		const pointer = rawPointer.startsWith("/") ? rawPointer : `/${rawPointer}`;
		if (!pointer.startsWith("/paths/") || pointer.startsWith("/paths/~1")) return null;
		const rest = pointer.slice(7);
		if (!rest) return null;
		const parts = rest.split("/");
		const last = parts.at(-1)?.toLowerCase() || "";
		const hasTailSegment = this.httpMethodSegments.has(last);
		const pathParts = hasTailSegment ? parts.slice(0, -1) : parts;
		const tail = hasTailSegment ? `/${parts.at(-1)}` : "";
		if (!pathParts.length) return null;
		return `${prefix}#/paths/${`~1${pathParts.join("/").replace(/\//g, "~1")}`}${tail}`;
	}
	isHttpUrl(value) {
		return /^https?:\/\//i.test(value);
	}
	getRemoteRequestHeaders() {
		return Object.assign({}, this.config.authorizationToken ? { Authorization: this.config.authorizationToken } : {}, this.config.requestOptions?.headers || {});
	}
	stripHash(urlOrPath) {
		return urlOrPath.split("#")[0] || urlOrPath;
	}
	extractRefsFromSchema(schema) {
		const refs = /* @__PURE__ */ new Set();
		const walk = (node) => {
			if (!node || typeof node !== "object") return;
			if (Array.isArray(node)) {
				for (const item of node) walk(item);
				return;
			}
			const recordNode = node;
			if (typeof recordNode.$ref === "string") refs.add(recordNode.$ref);
			for (const value of Object.values(recordNode)) walk(value);
		};
		walk(schema);
		return [...refs];
	}
	async fetchRemoteSchemaDocument(url) {
		try {
			const response = await fetch(url, { headers: this.getRemoteRequestHeaders() });
			if (!response.ok) return null;
			const content = await response.text();
			try {
				const parsed = parseSchemaContent(content);
				if (parsed && typeof parsed === "object") return parsed;
			} catch {
				return null;
			}
		} catch (e) {
			consola.default.debug(e);
		}
		return null;
	}
	async warmUpRemoteSchemasCache() {
		if (typeof this.config.url !== "string" || !this.isHttpUrl(this.config.url)) return;
		const visited = /* @__PURE__ */ new Set();
		const queue = [this.stripHash(this.config.url)];
		while (queue.length > 0) {
			const currentUrl = queue.shift();
			if (!currentUrl || visited.has(currentUrl)) continue;
			visited.add(currentUrl);
			if (this.externalSchemaCache.has(currentUrl)) continue;
			const schema = await this.fetchRemoteSchemaDocument(currentUrl);
			if (!schema) continue;
			this.externalSchemaCache.set(currentUrl, schema);
			for (const ref of this.extractRefsFromSchema(schema)) {
				const normalizedRef = this.normalizeRef(ref);
				if (normalizedRef.startsWith("#")) continue;
				const [externalPath = ""] = normalizedRef.split("#");
				if (!externalPath) continue;
				let absoluteUrl = "";
				try {
					absoluteUrl = this.isHttpUrl(externalPath) ? this.stripHash(externalPath) : this.stripHash(new URL(externalPath, currentUrl).toString());
				} catch (e) {
					consola.default.debug(e);
				}
				if (absoluteUrl && !visited.has(absoluteUrl)) queue.push(absoluteUrl);
			}
		}
	}
	collectRemoteAbsoluteRefCandidates(ref) {
		if (!ref || this.isHttpUrl(ref) || ref.startsWith("#")) return [];
		const [relativePath = "", rawPointer = ""] = this.normalizeRef(ref).split("#");
		if (!relativePath) return [];
		const pointer = rawPointer ? rawPointer.startsWith("/") ? rawPointer : `/${rawPointer}` : "";
		const bases = /* @__PURE__ */ new Set();
		if (typeof this.config.url === "string" && this.isHttpUrl(this.config.url)) bases.add(this.config.url);
		for (const cachedUrl of this.externalSchemaCache.keys()) if (this.isHttpUrl(cachedUrl)) bases.add(cachedUrl);
		for (const resolver of this.resolvers) try {
			const resolverPaths = typeof resolver.paths === "function" ? resolver.paths() : [];
			for (const resolverPath of resolverPaths) if (typeof resolverPath === "string" && this.isHttpUrl(resolverPath)) bases.add(resolverPath);
		} catch (e) {
			consola.default.debug(e);
		}
		const results = /* @__PURE__ */ new Set();
		for (const base of bases) try {
			const absolutePath = new URL(relativePath, base).toString();
			results.add(pointer ? `${absolutePath}#${pointer}` : absolutePath);
		} catch (e) {
			consola.default.debug(e);
		}
		return [...results];
	}
	resolveFromRemoteSchemaCache(absoluteRef) {
		const [externalPath = "", rawPointer = ""] = this.normalizeRef(absoluteRef).split("#");
		if (!externalPath || !this.isHttpUrl(externalPath)) return null;
		const schema = this.externalSchemaCache.get(this.stripHash(externalPath));
		if (!schema) return null;
		const pointer = rawPointer ? rawPointer.startsWith("/") ? rawPointer : `/${rawPointer}` : "/";
		const resolved = this.resolveJsonPointer(schema, pointer);
		if (resolved == null) return null;
		return this.absolutizeLocalRefs(resolved, this.stripHash(externalPath));
	}
	originalProducesByRoute = Object.create(null);
	constructor(config, usageSchema, originalSchema, resolvers, originalProducesByRoute) {
		this.config = config;
		this.usageSchema = usageSchema;
		this.originalSchema = originalSchema;
		this.resolvers = resolvers;
		this.usageSchema = usageSchema;
		this.originalSchema = originalSchema;
		if (originalProducesByRoute) this.originalProducesByRoute = originalProducesByRoute;
	}
	getOriginalProduces(pathName, method) {
		return this.originalProducesByRoute[pathName]?.[method];
	}
	getRefDetails(ref) {
		const normalizedRef = this.normalizeRef(ref);
		if (!this.parsedRefsCache.has(ref)) {
			const isLocal = normalizedRef.startsWith("#");
			if (isLocal) this.parsedRefsCache.set(ref, {
				ref: normalizedRef,
				isLocal,
				externalUrlOrPath: null
			});
			else {
				const externalUrlOrPath = normalizedRef.split("#")[0] || "";
				let externalOpenapiFileName = externalUrlOrPath.replace(/\/$/, "").split("/").at(-1) || "";
				if (externalOpenapiFileName.endsWith(".json") || externalOpenapiFileName.endsWith(".yaml")) externalOpenapiFileName = externalOpenapiFileName.slice(0, -5);
				else if (externalOpenapiFileName.endsWith(".yml")) externalOpenapiFileName = externalOpenapiFileName.slice(0, -4);
				this.parsedRefsCache.set(ref, {
					ref: normalizedRef,
					isLocal,
					externalUrlOrPath,
					externalOpenapiFileName
				});
			}
		}
		const cachedRef = this.parsedRefsCache.get(ref);
		if (cachedRef) return cachedRef;
		if (normalizedRef.startsWith("#")) return {
			ref: normalizedRef,
			isLocal: true,
			externalUrlOrPath: null
		};
		return {
			ref: normalizedRef,
			isLocal: false,
			externalUrlOrPath: normalizedRef.split("#")[0] || null,
			externalOpenapiFileName: ""
		};
	}
	isLocalRef(ref) {
		return this.getRefDetails(ref).isLocal;
	}
	getRef(ref) {
		if (!ref) return null;
		const normalizedRef = this.normalizeRef(ref);
		const escapedPathsRefVariant = this.createEscapedPathsRefVariant(ref);
		if (normalizedRef !== ref) {
			const resolvedByNormalizedRef = this.tryToResolveRef(normalizedRef);
			if (resolvedByNormalizedRef) return this.normalizeResolvedExternalSchemaRef(normalizedRef, resolvedByNormalizedRef);
		}
		if (escapedPathsRefVariant) {
			const resolvedByEscapedPathsRef = this.tryToResolveRef(escapedPathsRefVariant);
			if (resolvedByEscapedPathsRef) return this.normalizeResolvedExternalSchemaRef(escapedPathsRefVariant, resolvedByEscapedPathsRef);
		}
		const remoteAbsoluteCandidates = this.collectRemoteAbsoluteRefCandidates(ref);
		for (const remoteAbsoluteRef of remoteAbsoluteCandidates) {
			const resolvedFromRemoteCache = this.resolveFromRemoteSchemaCache(remoteAbsoluteRef);
			if (resolvedFromRemoteCache) return resolvedFromRemoteCache;
			const resolvedByRemoteAbsoluteRef = this.tryToResolveRef(remoteAbsoluteRef);
			if (resolvedByRemoteAbsoluteRef) return this.normalizeResolvedExternalSchemaRef(remoteAbsoluteRef, resolvedByRemoteAbsoluteRef);
		}
		const resolvedByOrigRef = this.tryToResolveRef(ref);
		if (resolvedByOrigRef) return this.normalizeResolvedExternalSchemaRef(ref, resolvedByOrigRef);
		if (/#[a-z]/.test(ref)) {
			const fixedRef = ref.replace(/#[a-z]/, (match) => {
				const [hashtag, char] = match.split("");
				return `${hashtag}/${char}`;
			});
			const resolvedByFixedRef = this.tryToResolveRef(fixedRef);
			if (resolvedByFixedRef) return this.normalizeResolvedExternalSchemaRef(fixedRef, resolvedByFixedRef);
		}
		return this.tryToResolveRefFromFile(normalizedRef);
	}
	tryToResolveRef(ref) {
		if (!this.resolvers || !ref) return null;
		for (const resolver of this.resolvers) try {
			return resolver.get(ref);
		} catch (e) {
			consola.default.debug(e);
		}
		return null;
	}
	readExternalSchema(filePath) {
		if (this.externalSchemaCache.has(filePath)) return this.externalSchemaCache.get(filePath) || null;
		if (!node_fs.existsSync(filePath)) return null;
		try {
			const parsed = parseSchemaContent(node_fs.readFileSync(filePath, "utf8"));
			this.externalSchemaCache.set(filePath, parsed);
			return parsed;
		} catch {
			consola.default.debug("Failed to parse external schema", filePath);
		}
		return null;
	}
	resolveJsonPointer(source, pointer) {
		if (!source || typeof source !== "object") return null;
		if (!pointer || pointer === "/") return source;
		const tokens = pointer.replace(/^\/+/, "").split("/").filter(Boolean).map((part) => decodeURIComponent(part.replace(/~1/g, "/").replace(/~0/g, "~")));
		let current = source;
		for (const token of tokens) {
			if (!current || typeof current !== "object") return null;
			current = current[token];
		}
		return current ?? null;
	}
	absolutizeLocalRefs(value, externalPath) {
		if (value == null || typeof value !== "object") return value;
		const cloneValue = structuredClone(value);
		const walk = (node) => {
			if (!node || typeof node !== "object") return;
			if (Array.isArray(node)) {
				for (const item of node) walk(item);
				return;
			}
			const recordNode = node;
			if (typeof recordNode.$ref === "string" && recordNode.$ref.startsWith("#")) recordNode.$ref = `${externalPath}${this.normalizeRef(recordNode.$ref)}`;
			for (const nested of Object.values(recordNode)) walk(nested);
		};
		walk(cloneValue);
		return cloneValue;
	}
	normalizeResolvedExternalSchemaRef(ref, resolved) {
		const normalizedRef = this.normalizeRef(ref);
		if (normalizedRef.startsWith("#")) return resolved;
		const externalPath = normalizedRef.split("#")[0] || "";
		if (!externalPath) return resolved;
		return this.absolutizeLocalRefs(resolved, externalPath);
	}
	collectExternalRefCandidates(externalPath) {
		const candidates = /* @__PURE__ */ new Set();
		const isRemote = /^https?:\/\//i.test(externalPath);
		const normalizedExternalPath = externalPath.startsWith("/") ? externalPath.replace(/^\/+/, "") : externalPath;
		if (isRemote) return [];
		if (node_path.default.isAbsolute(externalPath)) candidates.add(externalPath);
		const inputPath = this.config.input;
		if (typeof inputPath === "string" && inputPath) candidates.add(node_path.default.resolve(node_path.default.dirname(inputPath), normalizedExternalPath));
		for (const resolver of this.resolvers) try {
			const resolverPaths = typeof resolver.paths === "function" ? resolver.paths() : [];
			for (const resolverPath of resolverPaths) {
				if (typeof resolverPath !== "string") continue;
				if (/^https?:\/\//i.test(resolverPath)) continue;
				candidates.add(node_path.default.resolve(node_path.default.dirname(resolverPath), normalizedExternalPath));
			}
		} catch (e) {
			consola.default.debug(e);
		}
		return [...candidates];
	}
	tryToResolveRefFromFile(ref) {
		if (!ref || ref.startsWith("#")) return null;
		const [externalPath = "", rawPointer = ""] = ref.split("#");
		if (!externalPath) return null;
		const pointer = rawPointer ? rawPointer.startsWith("/") ? rawPointer : `/${rawPointer}` : "/";
		const candidates = this.collectExternalRefCandidates(externalPath);
		for (const candidate of candidates) {
			const schema = this.readExternalSchema(candidate);
			const resolved = this.resolveJsonPointer(schema, pointer);
			if (resolved != null) return this.absolutizeLocalRefs(resolved, externalPath);
		}
		return null;
	}
	static async create(config, usageSchema, originalSchema, originalProducesByRoute) {
		const resolvers = [];
		const options = {
			continueOnError: true,
			mutateInputSchema: true,
			dereference: {},
			validate: {
				schema: false,
				spec: false
			},
			resolve: {
				external: true,
				http: {
					...config.requestOptions,
					headers: Object.assign({}, config.authorizationToken ? { Authorization: config.authorizationToken } : {}, config.requestOptions?.headers ?? {})
				}
			}
		};
		try {
			resolvers.push(await _apidevtools_swagger_parser.default.resolve(originalSchema, options));
		} catch (e) {
			consola.default.debug(e);
		}
		try {
			resolvers.push(await _apidevtools_swagger_parser.default.resolve(usageSchema, options));
		} catch (e) {
			consola.default.debug(e);
		}
		try {
			resolvers.push(await _apidevtools_swagger_parser.default.resolve(config.url || config.input || config.spec, options));
		} catch (e) {
			consola.default.debug(e);
		}
		const resolvedSwaggerSchema = new ResolvedSwaggerSchema(config, usageSchema, originalSchema, resolvers, originalProducesByRoute);
		await resolvedSwaggerSchema.warmUpRemoteSchemasCache();
		return resolvedSwaggerSchema;
	}
};
//#endregion
//#region src/util/request.ts
var Request = class {
	config;
	constructor(config) {
		this.config = config;
	}
	async download({ url, authToken, ...options }) {
		const requestOptions = {};
		if (authToken) requestOptions.headers = { Authorization: authToken };
		(0, es_toolkit.merge)((0, es_toolkit.merge)(requestOptions, options), this.config.requestOptions || {});
		try {
			return await (await fetch(url, requestOptions)).text();
		} catch (error) {
			const message = `error while fetching data from URL "${url}"`;
			consola.consola.error(message, error);
			return message;
		}
	}
};
//#endregion
//#region src/swagger-schema-resolver.ts
var SwaggerSchemaResolver = class {
	config;
	fileSystem;
	request;
	constructor(config, fileSystem) {
		this.config = config;
		this.fileSystem = fileSystem;
		this.request = new Request(config);
	}
	async create() {
		const { spec, patch, input, url, authorizationToken } = this.config;
		let swaggerSchemas;
		if (spec) swaggerSchemas = await this.convertSwaggerObject(spec, { patch });
		else {
			const swaggerSchemaFile = await this.fetchSwaggerSchemaFile(input, url, authorizationToken);
			const swaggerSchemaObject = this.processSwaggerSchemaFile(swaggerSchemaFile);
			swaggerSchemas = await this.convertSwaggerObject(swaggerSchemaObject, { patch });
		}
		const originalProducesByRoute = this.fixSwaggerSchemas(swaggerSchemas);
		return await ResolvedSwaggerSchema.create(this.config, swaggerSchemas.usageSchema, swaggerSchemas.originalSchema, originalProducesByRoute);
	}
	convertSwaggerObject(swaggerSchema, converterOptions) {
		return new Promise((resolve) => {
			const result = structuredClone(swaggerSchema);
			const originalSchemaForProduces = structuredClone(swaggerSchema);
			result.info = (0, es_toolkit.merge)({
				title: "No title",
				version: ""
			}, result.info || {});
			if (!Object.hasOwn(result, "openapi")) {
				result.paths = (0, es_toolkit.merge)({}, result.paths || {});
				swagger2openapi.convertObj(result, {
					...converterOptions,
					resolveInternal: true,
					warnOnly: true,
					refSiblings: "preserve",
					rbname: "requestBodyName"
				}, (err, options) => {
					const parsedSwaggerSchema = (0, es_toolkit_compat.get)(err, "options.openapi") ?? (0, es_toolkit_compat.get)(options, "openapi");
					if (!parsedSwaggerSchema && err) throw err;
					this.config.update({ convertedFromSwagger2: true });
					resolve({
						usageSchema: parsedSwaggerSchema,
						originalSchema: result,
						originalSchemaForProduces
					});
				});
			} else resolve({
				usageSchema: result,
				originalSchema: structuredClone(result),
				originalSchemaForProduces
			});
		});
	}
	getSwaggerSchemaByPath = (pathToSwagger) => {
		consola.consola.info(`try to get swagger by path "${pathToSwagger}"`);
		return this.fileSystem.getFileContent(pathToSwagger);
	};
	async fetchSwaggerSchemaFile(pathToSwagger, urlToSwagger, authToken) {
		if (this.fileSystem.pathIsExist(pathToSwagger)) return this.getSwaggerSchemaByPath(pathToSwagger);
		consola.consola.info(`try to get swagger by URL "${urlToSwagger}"`);
		return await this.request.download({
			url: urlToSwagger,
			authToken
		});
	}
	processSwaggerSchemaFile(file) {
		if (typeof file !== "string") return file;
		return parseSchemaContent(file);
	}
	normalizeRefValue(ref) {
		const refWithoutSlashBeforeHash = ref.split("/#/").join("#/");
		const hashIndex = refWithoutSlashBeforeHash.indexOf("#");
		if (hashIndex === -1) return refWithoutSlashBeforeHash;
		if (refWithoutSlashBeforeHash[hashIndex + 1] === "/") return refWithoutSlashBeforeHash;
		return `${refWithoutSlashBeforeHash.slice(0, hashIndex + 1)}/${refWithoutSlashBeforeHash.slice(hashIndex + 1)}`;
	}
	normalizeRefsInSchema(schema) {
		if (!schema || typeof schema !== "object") return;
		if (Array.isArray(schema)) {
			for (const value of schema) this.normalizeRefsInSchema(value);
			return;
		}
		const objectSchema = schema;
		if (typeof objectSchema.$ref === "string") objectSchema.$ref = this.normalizeRefValue(objectSchema.$ref);
		for (const value of Object.values(objectSchema)) this.normalizeRefsInSchema(value);
	}
	fixSwaggerSchemas({ usageSchema, originalSchema, originalSchemaForProduces }) {
		this.normalizeRefsInSchema(usageSchema);
		this.normalizeRefsInSchema(originalSchema);
		const usagePaths = (0, es_toolkit_compat.get)(usageSchema, "paths") || {};
		const schemaForProduces = originalSchemaForProduces ?? originalSchema;
		const originalPaths = (0, es_toolkit_compat.get)(schemaForProduces, "paths") || {};
		const basePath = schemaForProduces.basePath?.replace(/\/$/, "") || "";
		const originalProducesByRoute = Object.create(null);
		for (const [route, usagePathObject] of Object.entries(usagePaths)) {
			const routeWithoutBase = basePath && route.startsWith(basePath) ? route.slice(basePath.length) || "/" : route;
			const routeWithBase = basePath && !route.startsWith(basePath) ? `${basePath}${route.startsWith("/") ? route : `/${route}`}` : route;
			const originalPathObject = (0, es_toolkit_compat.get)(originalPaths, route) || (0, es_toolkit_compat.get)(originalPaths, routeWithoutBase.startsWith("/") ? routeWithoutBase : `/${routeWithoutBase}`) || (0, es_toolkit_compat.get)(originalPaths, routeWithBase) || {};
			for (const [methodName, usageRouteInfo] of Object.entries(usagePathObject)) {
				const originalRouteInfo = (0, es_toolkit_compat.get)(originalPathObject, methodName) || {};
				const usageRouteParams = (0, es_toolkit_compat.get)(usageRouteInfo, "parameters") || [];
				const originalRouteParams = (0, es_toolkit_compat.get)(originalRouteInfo, "parameters") || [];
				const usageAsOpenapiv2 = usageRouteInfo;
				if (typeof usageRouteInfo === "object") {
					usageAsOpenapiv2.consumes = (0, es_toolkit.uniq)((0, es_toolkit.compact)([...usageAsOpenapiv2.consumes || [], ...originalRouteInfo.consumes || []]));
					let mergedProduces = (0, es_toolkit.uniq)((0, es_toolkit.compact)([...usageAsOpenapiv2.produces || [], ...originalRouteInfo.produces || []]));
					if (mergedProduces.length === 0) mergedProduces = (0, es_toolkit.uniq)((0, es_toolkit.compact)((0, es_toolkit_compat.get)(schemaForProduces, "produces") || []));
					usageAsOpenapiv2.produces = mergedProduces;
					if (mergedProduces.length > 0) {
						if (!originalProducesByRoute[route]) originalProducesByRoute[route] = Object.create(null);
						originalProducesByRoute[route][methodName] = mergedProduces;
					}
				}
				for (const originalRouteParam of originalRouteParams) if (!usageRouteParams.find((param) => originalRouteParam.in === param.in && originalRouteParam.name === param.name)) usageRouteParams.push(originalRouteParam);
			}
		}
		return originalProducesByRoute;
	}
};
//#endregion
//#region src/templates-worker.ts
const require$1 = node_module.createRequire(require("url").pathToFileURL(__filename).href);
const eta$1 = new eta.Eta({ functionHeader: "const includeFile = options.includeFile;" });
var TemplatesWorker = class {
	config;
	fileSystem;
	getRenderTemplateData;
	constructor(config, fileSystem, getRenderTemplateData) {
		this.config = config;
		this.fileSystem = fileSystem;
		this.getRenderTemplateData = getRenderTemplateData;
		if (this.config.debug) consola.consola.level = Number.MAX_SAFE_INTEGER;
		if (this.config.silent) consola.consola.level = 0;
	}
	getTemplatePaths = (config) => {
		const __dirname = node_path.dirname(node_url.fileURLToPath(require("url").pathToFileURL(__filename).href));
		const baseTemplatesPath = node_path.resolve(__dirname, "../templates/base");
		const defaultTemplatesPath = node_path.resolve(__dirname, "../templates/default");
		const modularTemplatesPath = node_path.resolve(__dirname, "../templates/modular");
		return {
			/** `templates/base` */
			base: baseTemplatesPath,
			/** `templates/default` */
			default: defaultTemplatesPath,
			/** `templates/modular` */
			modular: modularTemplatesPath,
			/** usage path if `--templates` option is not set */
			original: config.modular ? modularTemplatesPath : defaultTemplatesPath,
			/** custom path to templates (`--templates`) */
			custom: config.templates && node_path.resolve(process.cwd(), config.templates)
		};
	};
	cropExtension = (path) => this.config.templateExtensions.reduce((path, ext) => path.endsWith(ext) ? path.replace(ext, "") : path, path);
	getTemplateFullPath = (path_, fileName) => {
		const raw = node_path.resolve(path_, "./", this.cropExtension(fileName));
		return this.config.templateExtensions.map((extension) => `${raw}${extension}`).find((variant) => !!this.fileSystem.pathIsExist(variant));
	};
	requireFnFromTemplate = (packageOrPath) => {
		if (packageOrPath.startsWith("./") || packageOrPath.startsWith("../")) return require$1(node_path.resolve(this.config.templatePaths.custom || this.config.templatePaths.original, packageOrPath));
		return require$1(packageOrPath);
	};
	getTemplate = (name, fileName, path) => {
		const { templatePaths } = this.config;
		if (path) return this.fileSystem.getFileContent(path);
		if (!fileName) return "";
		const customFullPath = templatePaths.custom && this.getTemplateFullPath(templatePaths.custom, fileName);
		let fileContent = customFullPath && this.fileSystem.getFileContent(customFullPath);
		if (fileContent) {
			consola.consola.info(`"${name.toLowerCase()}" template found in "${templatePaths.custom}"`);
			return fileContent;
		}
		const baseFullPath = this.getTemplateFullPath(templatePaths.base, fileName);
		if (baseFullPath) fileContent = this.fileSystem.getFileContent(baseFullPath);
		else if (templatePaths.custom) consola.consola.warn("Code generator will use the default template:", `"${name.toLowerCase()}"`, "template not found in", `"${templatePaths.custom}"`);
		else consola.consola.info(`Code generator will use the default template for "${name.toLowerCase()}"`);
		const originalFullPath = this.getTemplateFullPath(templatePaths.original, fileName);
		if (originalFullPath) fileContent = this.fileSystem.getFileContent(originalFullPath);
		return fileContent;
	};
	getTemplates = ({ templatePaths }) => {
		if (templatePaths.custom) consola.consola.info(`try to read templates from directory "${templatePaths.custom}"`);
		return this.config.templateInfos.reduce((acc, { name, fileName }) => ({
			...acc,
			[name]: this.getTemplate(name, fileName)
		}), {});
	};
	findTemplateWithExt = (path) => {
		const raw = this.cropExtension(path);
		return this.config.templateExtensions.map((extension) => `${raw}${extension}`).find((variant) => this.fileSystem.pathIsExist(variant));
	};
	getTemplateContent = (path_) => {
		const foundTemplatePathKey = Object.keys(this.config.templatePaths).find((key) => path_.startsWith(`@${key}`));
		if (foundTemplatePathKey) {
			const rawPath = node_path.resolve(path_.replace(`@${foundTemplatePathKey}`, (0, es_toolkit_compat.get)(this.config.templatePaths, foundTemplatePathKey)));
			const fixedPath = this.findTemplateWithExt(rawPath);
			if (fixedPath) return this.fileSystem.getFileContent(fixedPath);
		}
		const customPath = this.config.templatePaths.custom && this.findTemplateWithExt(node_path.resolve(this.config.templatePaths.custom, path_));
		if (customPath) return this.fileSystem.getFileContent(customPath);
		const originalPath = this.findTemplateWithExt(node_path.resolve(this.config.templatePaths.original, path_));
		if (originalPath) return this.fileSystem.getFileContent(originalPath);
		return "";
	};
	renderTemplate = (template, configuration) => {
		if (!template) return "";
		return eta$1.render(eta$1.compile(template, { async: false }), {
			...this.getRenderTemplateData(),
			...configuration
		}, { includeFile: (path, configuration) => this.renderTemplate(this.getTemplateContent(path), configuration) });
	};
};
//#endregion
//#region src/translators/translator.ts
var Translator = class {
	config;
	codeFormatter;
	constructor(config, codeFormatter) {
		this.config = config;
		this.codeFormatter = codeFormatter;
	}
	translate(_input) {
		throw new Error("not implemented");
	}
};
//#endregion
//#region src/translators/javascript.ts
var JavascriptTranslator = class extends Translator {
	compileTSCode = (input) => {
		const fileNameFull = `${input.fileName}${input.fileExtension}`;
		const output = {};
		const host = typescript.createCompilerHost(this.config.compilerTsConfig, true);
		const fileNames = [fileNameFull];
		const originalSourceFileGet = host.getSourceFile.bind(host);
		host.getSourceFile = (sourceFileName, languageVersion, onError, shouldCreateNewSourceFile) => {
			if (sourceFileName !== fileNameFull) return originalSourceFileGet(sourceFileName, languageVersion, onError, shouldCreateNewSourceFile);
			return typescript.createSourceFile(sourceFileName, input.fileContent, languageVersion, true, typescript.ScriptKind.TS);
		};
		host.writeFile = (fileName, contents) => {
			output[fileName] = contents;
		};
		typescript.createProgram(fileNames, this.config.compilerTsConfig, host).emit();
		return output;
	};
	translate = async (input) => {
		const compiled = this.compileTSCode(input);
		const jsFileName = `${input.fileName}${typescript.Extension.Js}`;
		const dtsFileName = `${input.fileName}${typescript.Extension.Dts}`;
		const sourceContent = compiled[jsFileName];
		const tsImportRows = input.fileContent.split("\n").filter((line) => line.startsWith("import "));
		const declarationContent = compiled[dtsFileName].split("\n").map((line) => {
			if (line.startsWith("import ")) return tsImportRows.shift();
			return line;
		}).join("\n");
		return [{
			fileName: input.fileName,
			fileExtension: typescript.Extension.Js,
			fileContent: await this.codeFormatter.formatCode(sourceContent)
		}, {
			fileName: input.fileName,
			fileExtension: typescript.Extension.Dts,
			fileContent: await this.codeFormatter.formatCode(declarationContent)
		}];
	};
};
//#endregion
//#region src/type-name-formatter.ts
var TypeNameFormatter = class {
	formattedModelNamesMap = /* @__PURE__ */ new Map();
	usedFormattedTypeNames = /* @__PURE__ */ new Map();
	config;
	constructor(config) {
		this.config = config;
	}
	/**
	* Return the TypeScript identifier for a raw OpenAPI name. Fast path is a
	* cache hit on names resolved by `precommit`. The fallback (for names
	* discovered after precommit, e.g. enum keys inside schemas or types added
	* dynamically by `extractEnums`/`extractResponses`) computes the identifier
	* inline WITHOUT collision handling — collision resolution is the sole
	* responsibility of `precommit`. Callers expecting collision safety must
	* list every raw name in the precommit input.
	*/
	format = (name, options = {}) => {
		const schemaType = options.type ?? "type-name";
		const { typePrefix, typeSuffix } = this.getAffixes(schemaType);
		const hashKey = `${typePrefix}_${name}_${typeSuffix}`;
		const cached = this.formattedModelNamesMap.get(hashKey);
		if (cached !== void 0) return cached;
		if (typeof name !== "string") {
			consola.consola.warn("wrong model name", name);
			return name;
		}
		const result = this.computeFormattedName(name, schemaType, typePrefix, typeSuffix);
		this.formattedModelNamesMap.set(hashKey, result);
		return result;
	};
	/**
	* Resolve the TypeScript identifier for every raw schema name passed in,
	* populating `formattedModelNamesMap` and `usedFormattedTypeNames`. Must be
	* called once, before any `format()` call from the schema parser, so that
	* every subsequent `format()` for these names is a cache hit and returns
	* the collision-resolved identifier.
	*
	* Two passes:
	*   1. Claim every raw name whose formatted output equals the raw name
	*      itself ("canonical"). User-declared identifiers like `FooBar` or
	*      `FooBar1` are preserved regardless of the source order in which
	*      the generator later visits them.
	*   2. For non-canonical names (e.g. `Foo_Bar` → `FooBar`), suffix with
	*      the smallest integer not already claimed, so collisions produce
	*      `FooBar`, `FooBar1`, `FooBar2`, … deterministically.
	*
	* Only type-names go through collision resolution here. Enum keys use the
	* fallback path in `format()` and are handled per-enum by the template.
	*/
	precommit = (rawNames) => {
		const schemaType = "type-name";
		const { typePrefix, typeSuffix } = this.getAffixes(schemaType);
		const seen = /* @__PURE__ */ new Set();
		const names = [];
		for (const name of rawNames) {
			if (typeof name !== "string") continue;
			if (seen.has(name)) continue;
			seen.add(name);
			names.push(name);
		}
		for (const name of names) {
			const formatted = this.computeFormattedName(name, schemaType, typePrefix, typeSuffix);
			if (name !== formatted) continue;
			if (this.usedFormattedTypeNames.has(formatted)) continue;
			this.usedFormattedTypeNames.set(formatted, name);
			const hashKey = `${typePrefix}_${name}_${typeSuffix}`;
			this.formattedModelNamesMap.set(hashKey, formatted);
		}
		for (const name of names) {
			const hashKey = `${typePrefix}_${name}_${typeSuffix}`;
			if (this.formattedModelNamesMap.has(hashKey)) continue;
			const formatted = this.computeFormattedName(name, schemaType, typePrefix, typeSuffix);
			let final = formatted;
			if (this.usedFormattedTypeNames.has(final)) {
				let suffix = 1;
				while (this.usedFormattedTypeNames.has(`${formatted}${suffix}`)) suffix += 1;
				final = `${formatted}${suffix}`;
			}
			this.usedFormattedTypeNames.set(final, name);
			this.formattedModelNamesMap.set(hashKey, final);
		}
	};
	isValidName = (name) => /^([A-Za-z$_]{1,})$/g.test(name);
	fixModelName = (name, options) => {
		if (!this.isValidName(name)) {
			if (!/^[a-zA-Z_$]/g.test(name)) return `${options.type === "enum-key" ? this.config.fixInvalidEnumKeyPrefix : this.config.fixInvalidTypeNamePrefix} ${name}`;
			if (name.includes(".")) return name.replace(/Exclude_keyof[A-Za-z]+/g, () => "ExcludeKeys").replace(/%22~AND~%22/g, "And").replace(/%22~OR~%22/g, "Or").replace(/(\.?%22)|\./g, "_").replace(/__+$/, "");
			if (name.includes("-")) return this.config.disableFormatTypeNames ? name.replace(/-/g, "_") : (0, es_toolkit_compat.startCase)(name).replace(/ /g, "");
		}
		return name;
	};
	getAffixes = (schemaType) => ({
		typePrefix: schemaType === "enum-key" ? this.config.enumKeyPrefix : this.config.typePrefix,
		typeSuffix: schemaType === "enum-key" ? this.config.enumKeySuffix : this.config.typeSuffix
	});
	computeFormattedName = (name, schemaType, typePrefix, typeSuffix) => {
		const typeNameSeparator = this.config.typeNameSeparator;
		let resultName = name;
		if (this.config.disableFormatTypeNames) resultName = (0, es_toolkit.compact)([
			typePrefix,
			resultName,
			typeSuffix
		]).join(typeNameSeparator);
		else if (/^(?!\d)([A-Z0-9_]{1,})$/g.test(resultName)) resultName = (0, es_toolkit.compact)([
			typePrefix,
			resultName,
			typeSuffix
		]).join(typeNameSeparator);
		else resultName = (0, es_toolkit_compat.startCase)((0, es_toolkit.compact)([
			typePrefix,
			this.fixModelName(resultName, { type: schemaType }),
			typeSuffix
		]).join(typeNameSeparator)).replace(/\s/g, "");
		return this.config.hooks.onFormatTypeName?.(resultName, name, schemaType) || resultName;
	};
};
//#endregion
//#region src/util/file-system.ts
var FileSystem = class {
	getFileContent = (path) => {
		return node_fs.readFileSync(path, { encoding: "utf8" });
	};
	readDir = (path) => {
		return node_fs.readdirSync(path);
	};
	pathIsDir = (path) => {
		if (!path) return false;
		try {
			return node_fs.statSync(path).isDirectory();
		} catch (e) {
			return false;
		}
	};
	cropExtension = (fileName) => {
		const fileNameParts = fileName.split(".");
		if (fileNameParts.length > 1) fileNameParts.pop();
		return fileNameParts.join(".");
	};
	removeDir = (path) => {
		try {
			if (typeof node_fs.rmSync === "function") node_fs.rmSync(path, { recursive: true });
			else node_fs.rmdirSync(path, { recursive: true });
		} catch (e) {
			consola.consola.debug("failed to remove dir", e);
		}
	};
	createDir = (path) => {
		try {
			node_fs.mkdirSync(path, { recursive: true });
		} catch (e) {
			consola.consola.debug("failed to create dir", e);
		}
	};
	cleanDir = (path) => {
		this.removeDir(path);
		this.createDir(path);
	};
	pathIsExist = (path) => {
		return !!path && node_fs.existsSync(path);
	};
	createFile = ({ path: path_, fileName, content, withPrefix }) => {
		const __dirname = node_path.dirname(node_url.fileURLToPath(require("url").pathToFileURL(__filename).href));
		const absolutePath = node_path.resolve(__dirname, path_, `./${fileName}`);
		const fileContent = `${withPrefix ? FILE_PREFIX : ""}${content}`;
		return node_fs.writeFileSync(absolutePath, fileContent);
	};
};
//#endregion
//#region src/util/lodash-compat.ts
function createLodashCompat() {
	return {
		...es_toolkit,
		...es_toolkit_compat
	};
}
//#endregion
//#region src/code-gen-process.ts
const PATCHABLE_INSTANCES = [
	"schemaWalker",
	"swaggerSchemaResolver",
	"schemaComponentsMap",
	"typeNameFormatter",
	"templatesWorker",
	"codeFormatter",
	"schemaParserFabric",
	"schemaRoutes",
	"javascriptTranslator"
];
var CodeGenProcess = class {
	config;
	swaggerSchemaResolver;
	schemaComponentsMap;
	typeNameFormatter;
	schemaParserFabric;
	schemaRoutes;
	fileSystem;
	codeFormatter;
	templatesWorker;
	javascriptTranslator;
	swaggerRefs;
	constructor(config) {
		this.config = new CodeGenConfig(config);
		this.fileSystem = new FileSystem();
		this.swaggerSchemaResolver = new SwaggerSchemaResolver(this.config, this.fileSystem);
		this.schemaComponentsMap = new SchemaComponentsMap(this.config);
		this.typeNameFormatter = new TypeNameFormatter(this.config);
		this.templatesWorker = new TemplatesWorker(this.config, this.fileSystem, this.getRenderTemplateData);
		this.codeFormatter = new CodeFormatter(this.config);
		this.schemaParserFabric = new SchemaParserFabric(this.config, this.templatesWorker, this.schemaComponentsMap, this.typeNameFormatter);
		this.schemaRoutes = new SchemaRoutes(this.config, this.schemaParserFabric, this.schemaComponentsMap, this.templatesWorker, this.typeNameFormatter);
		this.javascriptTranslator = new JavascriptTranslator(this.config, this.codeFormatter);
	}
	async start() {
		this.config.update({ templatePaths: this.templatesWorker.getTemplatePaths(this.config) });
		this.config.update({ templatesToRender: this.templatesWorker.getTemplates(this.config) });
		const resolvedSwaggerSchema = await this.swaggerSchemaResolver.create();
		this.config.update({
			resolvedSwaggerSchema,
			swaggerSchema: resolvedSwaggerSchema.usageSchema,
			originalSchema: resolvedSwaggerSchema.originalSchema
		});
		consola.consola.info("start generating your typescript api");
		this.config.update(this.config.hooks.onInit?.(this.config, this) || this.config);
		if (this.config.swaggerSchema) resolvedSwaggerSchema.usageSchema = this.config.swaggerSchema;
		if (this.config.originalSchema) resolvedSwaggerSchema.originalSchema = this.config.originalSchema;
		this.schemaComponentsMap.clear();
		for (const [componentName, component] of Object.entries(resolvedSwaggerSchema.usageSchema.components || {})) for (const [typeName, rawTypeData] of Object.entries(component)) this.schemaComponentsMap.createComponent(this.schemaComponentsMap.createRef([
			"components",
			componentName,
			typeName
		]), rawTypeData);
		this.schemaComponentsMap.discriminatorsFirst();
		this.schemaComponentsMap.enumsFirst();
		const componentsToParse = this.schemaComponentsMap.filter((0, es_toolkit.compact)(["schemas", this.config.extractResponses && "responses"]));
		this.typeNameFormatter.precommit(componentsToParse.map((c) => c.typeName));
		const parsedSchemas = componentsToParse.map((schemaComponent) => {
			const parsed = this.schemaParserFabric.parseSchema(schemaComponent.rawTypeData, schemaComponent.typeName);
			schemaComponent.typeData = parsed;
			return parsed;
		});
		this.schemaRoutes.attachSchema(resolvedSwaggerSchema, parsedSchemas);
		const rawConfiguration = {
			apiConfig: this.createApiConfig(resolvedSwaggerSchema.usageSchema),
			config: this.config,
			modelTypes: this.collectModelTypes(),
			hasSecurityRoutes: this.schemaRoutes.hasSecurityRoutes,
			hasQueryRoutes: this.schemaRoutes.hasQueryRoutes,
			hasFormDataRoutes: this.schemaRoutes.hasFormDataRoutes,
			generateResponses: this.config.generateResponses,
			routes: this.schemaRoutes.getGroupedRoutes(),
			extraTemplates: this.config.extraTemplates,
			fileName: this.config.fileName,
			translateToJavaScript: this.config.toJS,
			customTranslator: this.config.customTranslator ? new this.config.customTranslator() : null,
			utils: this.getRenderTemplateData().utils
		};
		const configuration = this.config.hooks.onPrepareConfig?.(rawConfiguration) || rawConfiguration;
		if (this.fileSystem.pathIsExist(this.config.output)) {
			if (this.config.cleanOutput) {
				consola.consola.debug("cleaning dir", this.config.output);
				this.fileSystem.cleanDir(this.config.output);
			}
		} else {
			consola.consola.debug(`path ${this.config.output} is not exist. creating dir by this path`);
			this.fileSystem.createDir(this.config.output);
		}
		const files = await this.generateOutputFiles({ configuration });
		if (this.fileSystem.pathIsDir(this.config.output)) for (const file of files) {
			this.fileSystem.createFile({
				path: this.config.output,
				fileName: `${file.fileName}${file.fileExtension}`,
				content: file.fileContent,
				withPrefix: true
			});
			consola.consola.success("api file", `"${file.fileName}${file.fileExtension}"`, `created in ${this.config.output}`);
		}
		return {
			files,
			configuration,
			getTemplate: this.templatesWorker.getTemplate,
			renderTemplate: this.templatesWorker.renderTemplate,
			createFile: this.fileSystem.createFile,
			formatTSContent: this.codeFormatter.formatCode
		};
	}
	getRenderTemplateData = () => {
		return {
			utils: {
				Ts: this.config.Ts,
				formatDescription: this.schemaParserFabric.schemaFormatters.formatDescription,
				escapeJSDocContent: this.schemaParserFabric.schemaFormatters.escapeJSDocContent,
				internalCase: es_toolkit_compat.camelCase,
				classNameCase: pascalCase,
				pascalCase,
				getInlineParseContent: this.schemaParserFabric.getInlineParseContent,
				getParseContent: this.schemaParserFabric.getParseContent,
				getComponentByRef: this.schemaComponentsMap.get,
				parseSchema: this.schemaParserFabric.parseSchema,
				checkAndAddNull: this.schemaParserFabric.schemaUtils.safeAddNullToType,
				safeAddNullToType: this.schemaParserFabric.schemaUtils.safeAddNullToType,
				isNeedToAddNull: this.schemaParserFabric.schemaUtils.isNullMissingInType,
				inlineExtraFormatters: this.schemaParserFabric.schemaFormatters.inline,
				formatters: this.schemaParserFabric.schemaFormatters.base,
				formatModelName: this.typeNameFormatter.format,
				fmtToJSDocLine: (line, { eol = true }) => {
					return ` * ${line}${eol ? "\n" : ""}`;
				},
				NameResolver,
				_: createLodashCompat(),
				require: this.templatesWorker.requireFnFromTemplate
			},
			config: this.config
		};
	};
	collectModelTypes = () => {
		const components = this.schemaComponentsMap.getComponents();
		let modelTypes = [];
		const modelTypeComponents = (0, es_toolkit.compact)(["schemas", this.config.extractResponses && "responses"]);
		const getSchemaComponentsCount = () => this.schemaComponentsMap.filter(...modelTypeComponents).length;
		let schemaComponentsCount = getSchemaComponentsCount();
		let processedCount = 0;
		while (processedCount < schemaComponentsCount) {
			modelTypes = [];
			processedCount = 0;
			for (const component of components) if (modelTypeComponents.includes(component.componentName)) {
				const modelType = this.prepareModelType(component);
				if (modelType) modelTypes.push(modelType);
				processedCount++;
			}
			schemaComponentsCount = getSchemaComponentsCount();
		}
		if (this.config.sortTypes) return modelTypes.sort(sortByProperty("name"));
		return modelTypes;
	};
	prepareModelType = (typeInfo) => {
		if (typeInfo.$prepared) return typeInfo.$prepared;
		if (!typeInfo.typeData) typeInfo.typeData = this.schemaParserFabric.parseSchema(typeInfo.rawTypeData, typeInfo.typeName);
		const rawTypeData = typeInfo.typeData;
		const typeData = this.schemaParserFabric.schemaFormatters.base[rawTypeData.type] ? this.schemaParserFabric.schemaFormatters.base[rawTypeData.type](rawTypeData) : rawTypeData;
		const { typeIdentifier, name: originalName, content, description } = typeData;
		const name = this.typeNameFormatter.format(originalName);
		if (name === null) return null;
		const preparedModelType = {
			...typeData,
			typeIdentifier,
			name,
			description,
			$content: rawTypeData.content,
			rawContent: rawTypeData.content,
			content,
			typeData
		};
		typeInfo.$prepared = preparedModelType;
		return preparedModelType;
	};
	generateOutputFiles = async ({ configuration }) => {
		const { modular, templatesToRender } = this.config;
		const output = modular ? await this.createMultipleFileInfos(templatesToRender, configuration) : await this.createSingleFileInfo(templatesToRender, configuration);
		if (configuration.extraTemplates?.length) for (const extraTemplate of configuration.extraTemplates) {
			const content = this.templatesWorker.renderTemplate(this.fileSystem.getFileContent(extraTemplate.path), configuration);
			output.push(...await this.createOutputFileInfo(configuration, extraTemplate.name, content));
		}
		return output.filter((fileInfo) => !!fileInfo && !!fileInfo.fileContent);
	};
	createMultipleFileInfos = async (templatesToRender, configuration) => {
		const { routes } = configuration;
		const { fileNames, generateRouteTypes, generateClient } = configuration.config;
		const modularApiFileInfos = [];
		if (routes.$outOfModule) {
			if (generateRouteTypes) {
				const outOfModuleRouteContent = this.templatesWorker.renderTemplate(templatesToRender.routeTypes, {
					...configuration,
					route: configuration.routes.$outOfModule
				});
				modularApiFileInfos.push(...await this.createOutputFileInfo(configuration, fileNames.outOfModuleApi, outOfModuleRouteContent));
			}
			if (generateClient) {
				const outOfModuleApiContent = this.templatesWorker.renderTemplate(templatesToRender.api, {
					...configuration,
					route: configuration.routes.$outOfModule
				});
				modularApiFileInfos.push(...await this.createOutputFileInfo(configuration, fileNames.outOfModuleApi, outOfModuleApiContent));
			}
		}
		if (routes.combined) for (const route of routes.combined) {
			if (generateRouteTypes) {
				const routeModuleContent = this.templatesWorker.renderTemplate(templatesToRender.routeTypes, {
					...configuration,
					route
				});
				modularApiFileInfos.push(...await this.createOutputFileInfo(configuration, pascalCase(`${route.moduleName}_Route`), routeModuleContent));
			}
			if (generateClient) {
				const apiModuleContent = this.templatesWorker.renderTemplate(templatesToRender.api, {
					...configuration,
					route
				});
				modularApiFileInfos.push(...await this.createOutputFileInfo(configuration, pascalCase(route.moduleName), apiModuleContent));
			}
		}
		return [
			...await this.createOutputFileInfo(configuration, fileNames.dataContracts, this.templatesWorker.renderTemplate(templatesToRender.dataContracts, configuration)),
			...generateClient ? await this.createOutputFileInfo(configuration, fileNames.httpClient, this.templatesWorker.renderTemplate(templatesToRender.httpClient, configuration)) : [],
			...modularApiFileInfos
		];
	};
	createSingleFileInfo = async (templatesToRender, configuration) => {
		const { generateRouteTypes, generateClient } = configuration.config;
		return await this.createOutputFileInfo(configuration, configuration.fileName, (0, es_toolkit.compact)([
			this.templatesWorker.renderTemplate(templatesToRender.dataContracts, configuration),
			generateRouteTypes && this.templatesWorker.renderTemplate(templatesToRender.routeTypes, configuration),
			generateClient && this.templatesWorker.renderTemplate(templatesToRender.httpClient, configuration),
			generateClient && this.templatesWorker.renderTemplate(templatesToRender.api, configuration)
		]).join("\n"));
	};
	createOutputFileInfo = async (configuration, fileNameFull, content) => {
		const fileName = this.fileSystem.cropExtension(fileNameFull);
		const fileExtension = typescript.Extension.Ts;
		if (configuration.translateToJavaScript) {
			consola.consola.debug("using js translator for", fileName);
			return await this.javascriptTranslator.translate({
				fileName,
				fileExtension,
				fileContent: content
			});
		}
		if (configuration.customTranslator) {
			consola.consola.debug("using custom translator for", fileName);
			return await configuration.customTranslator.translate({
				fileName,
				fileExtension,
				fileContent: content
			});
		}
		consola.consola.debug("generating output for", `${fileName}${fileExtension}`);
		return [{
			fileName,
			fileExtension,
			fileContent: await this.codeFormatter.formatCode(content)
		}];
	};
	createApiConfig = (swaggerSchema) => {
		const { info, servers, host, basePath, externalDocs, tags } = swaggerSchema;
		const server = servers?.[0] || { url: "" };
		const { title = "No title", version } = info || {};
		const { url: serverUrl } = server;
		return {
			info: info || {},
			servers: servers || [],
			basePath,
			host,
			externalDocs: (0, es_toolkit.merge)({
				url: "",
				description: ""
			}, externalDocs || {}),
			tags: (0, es_toolkit.compact)(tags || []),
			baseUrl: serverUrl,
			title,
			version
		};
	};
	injectClassInstance = (key, value) => {
		this[key] = value;
		for (const instanceKey of PATCHABLE_INSTANCES) if (instanceKey !== key && key in this[instanceKey]) this[instanceKey][key] = value;
	};
};
//#endregion
//#region types/index.ts
let RequestContentKind = /* @__PURE__ */ function(RequestContentKind) {
	RequestContentKind["JSON"] = "JSON";
	RequestContentKind["URL_ENCODED"] = "URL_ENCODED";
	RequestContentKind["FORM_DATA"] = "FORM_DATA";
	RequestContentKind["IMAGE"] = "IMAGE";
	RequestContentKind["OTHER"] = "OTHER";
	RequestContentKind["TEXT"] = "TEXT";
	return RequestContentKind;
}({});
let SCHEMA_TYPES = /* @__PURE__ */ function(SCHEMA_TYPES) {
	SCHEMA_TYPES["ARRAY"] = "array";
	SCHEMA_TYPES["OBJECT"] = "object";
	SCHEMA_TYPES["ENUM"] = "enum";
	SCHEMA_TYPES["REF"] = "$ref";
	SCHEMA_TYPES["PRIMITIVE"] = "primitive";
	SCHEMA_TYPES["COMPLEX"] = "complex";
	SCHEMA_TYPES["COMPLEX_ONE_OF"] = "oneOf";
	SCHEMA_TYPES["COMPLEX_ANY_OF"] = "anyOf";
	SCHEMA_TYPES["COMPLEX_ALL_OF"] = "allOf";
	SCHEMA_TYPES["COMPLEX_NOT"] = "not";
	SCHEMA_TYPES["COMPLEX_UNKNOWN"] = "__unknown";
	return SCHEMA_TYPES;
}({});
//#endregion
//#region src/commands/generate-templates/configuration.ts
var TemplatesGenConfig = class {
	cleanOutput = false;
	debug = false;
	httpClientType = HTTP_CLIENT.FETCH;
	modular = false;
	output = void 0;
	rewrite = false;
	silent = false;
	version = PROJECT_VERSION;
	constructor(config) {
		this.update(config);
	}
	update = (update) => {
		objectAssign(this, update);
	};
};
//#endregion
//#region src/commands/generate-templates/templates-gen-process.ts
const __dirname$1 = node_path.default.dirname(node_url.default.fileURLToPath(require("url").pathToFileURL(__filename).href));
var TemplatesGenProcess = class {
	config;
	fileSystem;
	rootDir = node_path.default.resolve(__dirname$1, "..");
	paths = {
		baseTemplates: "templates/base",
		httpClientTemplates: "templates/base/http-clients",
		moduleApiTemplates: "templates/modular",
		defaultApiTemplates: "templates/default"
	};
	importTemplatePrefixes = [
		"@base",
		"@modular",
		"@default"
	];
	constructor(config) {
		this.config = new TemplatesGenConfig(config);
		this.fileSystem = new FileSystem();
	}
	async start() {
		consola.consola.info("start generating source templates \".ejs\" for code generator");
		const templates = this.getTemplates();
		if (this.config.output) {
			consola.consola.info("preparing output directory for source templates");
			const outputPath = node_path.default.resolve(process.cwd(), this.config.output);
			if (this.fileSystem.pathIsExist(outputPath)) {
				if (this.config.cleanOutput) this.fileSystem.cleanDir(outputPath);
			} else this.fileSystem.createDir(outputPath);
			for (const template of templates) {
				const templateName = this.fileSystem.cropExtension(template.name);
				const templateEjsPath = node_path.default.resolve(outputPath, `${templateName}.ejs`);
				const templateEtaPath = node_path.default.resolve(outputPath, `${templateName}.eta`);
				const templateEjsPathExist = this.fileSystem.pathIsExist(templateEjsPath);
				const templateEtaPathExist = this.fileSystem.pathIsExist(templateEtaPath);
				if (!templateEjsPathExist && !templateEtaPathExist) this.fileSystem.createFile({
					path: outputPath,
					fileName: template.name,
					content: template.content,
					withPrefix: false
				});
				else if (this.config.rewrite) {
					if (templateEjsPathExist) this.fileSystem.createFile({
						path: outputPath,
						fileName: `${templateName}.ejs`,
						content: template.content,
						withPrefix: false
					});
					else if (templateEtaPathExist) this.fileSystem.createFile({
						path: outputPath,
						fileName: `${templateName}.eta`,
						content: template.content,
						withPrefix: false
					});
				}
			}
			consola.consola.success(`source templates has been successfully created in "${outputPath}"`);
		}
		return {
			files: templates,
			configuration: this.config,
			createFile: this.fileSystem.createFile
		};
	}
	getTemplates = () => {
		const outputFiles = [];
		const baseTemplates = this.getTemplateNamesFromDir(this.paths.baseTemplates);
		const httpClientTemplates = this.getTemplateNamesFromDir(this.paths.httpClientTemplates);
		const apiTemplatesPath = this.config.modular ? this.paths.moduleApiTemplates : this.paths.defaultApiTemplates;
		const apiTemplates = this.getTemplateNamesFromDir(apiTemplatesPath);
		const usingHttpClientTemplate = httpClientTemplates.find((template) => template.startsWith(`${this.config.httpClientType}-`));
		let httpClientTemplateContent = "";
		if (usingHttpClientTemplate) httpClientTemplateContent = this.fixTemplateContent(this.getTemplateContent(`${this.paths.httpClientTemplates}/${usingHttpClientTemplate}`));
		for (const fileName of baseTemplates) {
			const templateContent = fileName === "http-client.ejs" && httpClientTemplateContent || this.fixTemplateContent(this.getTemplateContent(`${this.paths.baseTemplates}/${fileName}`));
			outputFiles.push({
				name: fileName,
				content: templateContent
			});
		}
		for (const fileName of apiTemplates) outputFiles.push({
			name: fileName,
			content: this.fixTemplateContent(this.getTemplateContent(`${apiTemplatesPath}/${fileName}`))
		});
		return outputFiles;
	};
	fixTemplateContent = (content) => {
		const importsRegExp1 = new RegExp(`includeFile\\("(${this.importTemplatePrefixes.map((v) => `(${v})`).join("|")})/`, "g");
		const importsRegExp2 = new RegExp(`includeFile\\(\`(${this.importTemplatePrefixes.map((v) => `(${v})`).join("|")})/`, "g");
		const importsRegExp3 = new RegExp(`includeFile\\('(${this.importTemplatePrefixes.map((v) => `(${v})`).join("|")})/`, "g");
		return content.replace(importsRegExp1, "includeFile(\"./").replace(importsRegExp2, "includeFile(`./").replace(importsRegExp3, "includeFile('./");
	};
	getTemplateNamesFromDir = (dir) => {
		return this.fileSystem.readDir(node_path.default.resolve(this.rootDir, dir)).filter((file) => file.endsWith(".ejs"));
	};
	getTemplateContent = (pathToFile) => {
		return this.fileSystem.getFileContent(node_path.default.resolve(this.rootDir, pathToFile));
	};
};
//#endregion
//#region src/commands/generate-templates/index.ts
async function generateTemplates(config) {
	if (config.debug) consola.consola.level = Number.MAX_SAFE_INTEGER;
	if (config.silent) consola.consola.level = 0;
	return await new TemplatesGenProcess(config).start();
}
//#endregion
//#region src/index.ts
async function generateApi(config) {
	if (config.debug) consola.consola.level = Number.MAX_SAFE_INTEGER;
	if (config.silent) consola.consola.level = 0;
	return await new CodeGenProcess(config).start();
}
//#endregion
Object.defineProperty(exports, "CodeGenConfig", {
	enumerable: true,
	get: function() {
		return CodeGenConfig;
	}
});
Object.defineProperty(exports, "HTTP_CLIENT", {
	enumerable: true,
	get: function() {
		return HTTP_CLIENT;
	}
});
Object.defineProperty(exports, "RequestContentKind", {
	enumerable: true,
	get: function() {
		return RequestContentKind;
	}
});
Object.defineProperty(exports, "SCHEMA_TYPES", {
	enumerable: true,
	get: function() {
		return SCHEMA_TYPES;
	}
});
Object.defineProperty(exports, "TemplatesGenConfig", {
	enumerable: true,
	get: function() {
		return TemplatesGenConfig;
	}
});
Object.defineProperty(exports, "__toESM", {
	enumerable: true,
	get: function() {
		return __toESM;
	}
});
Object.defineProperty(exports, "constants_exports", {
	enumerable: true,
	get: function() {
		return constants_exports;
	}
});
Object.defineProperty(exports, "description", {
	enumerable: true,
	get: function() {
		return description;
	}
});
Object.defineProperty(exports, "generateApi", {
	enumerable: true,
	get: function() {
		return generateApi;
	}
});
Object.defineProperty(exports, "generateTemplates", {
	enumerable: true,
	get: function() {
		return generateTemplates;
	}
});
Object.defineProperty(exports, "name", {
	enumerable: true,
	get: function() {
		return name;
	}
});
Object.defineProperty(exports, "version", {
	enumerable: true,
	get: function() {
		return version;
	}
});
